{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3df36a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58114d66",
   "metadata": {},
   "source": [
    "### Check reference and download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ca8b2af3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 11:08:02 Start to check available reference files...\n",
      "2026/01/10 11:08:02  - Available keywords:  1kg_eas_hg19 1kg_eur_hg19 1kg_eas_hg38 1kg_eur_hg38 1kg_sas_hg19 1kg_amr_hg19 1kg_sas_hg38 1kg_amr_hg38 1kg_afr_hg19 1kg_pan_hg19 1kg_afr_hg38 1kg_pan_hg38 dbsnp_v151_hg19 dbsnp_v151_hg38 dbsnp_v157_hg19 dbsnp_v157_hg38 ucsc_genome_hg19 ucsc_genome_hg38 1kg_dbsnp151_hg19_auto 1kg_dbsnp151_hg38_auto recombination_hg19 recombination_hg38 ensembl_hg19_gtf ensembl_hg38_gtf refseq_hg19_gtf refseq_hg38_gtf testlink 19to38 38to19 1kg_hm3_hg38_eaf 1kg_hm3_hg19_eaf\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'1kg_eas_hg19': {'description': '1000 Genomes Project East Asian (1KG EAS) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EAS population'},\n",
       " '1kg_eur_hg19': {'description': '1000 Genomes Project European (1KG EUR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EUR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EUR population'},\n",
       " '1kg_eas_hg38': {'description': '1000 Genomes Project East Asian (1KG EAS) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EAS population'},\n",
       " '1kg_eur_hg38': {'description': '1000 Genomes Project European (1KG EUR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EUR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EUR population'},\n",
       " '1kg_sas_hg19': {'description': '1000 Genomes Project South Asian (1KG SAS) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the SAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for SAS population'},\n",
       " '1kg_amr_hg19': {'description': '1000 Genomes Project American (1KG AMR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AMR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AMR population'},\n",
       " '1kg_sas_hg38': {'description': '1000 Genomes Project South Asian (1KG SAS) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the SAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for SAS population'},\n",
       " '1kg_amr_hg38': {'description': '1000 Genomes Project American (1KG AMR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AMR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AMR population'},\n",
       " '1kg_afr_hg19': {'description': '1000 Genomes Project African (1KG AFR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AFR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AFR population'},\n",
       " '1kg_pan_hg19': {'description': '1000 Genomes Project Pan-ancestry (1KG PAN) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the PAN population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for PAN population'},\n",
       " '1kg_afr_hg38': {'description': '1000 Genomes Project African (1KG AFR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AFR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AFR population'},\n",
       " '1kg_pan_hg38': {'description': '1000 Genomes Project Pan-ancestry (1KG PAN) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the PAN population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for PAN population'},\n",
       " 'dbsnp_v151_hg19': {'description': 'dbSNP v151 hg19 is the 2018 release of the dbSNP database mapped to the human genome assembly GRCh37/hg19, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v151_hg38': {'description': 'dbSNP v151 hg38 is the 2018 release of the dbSNP database mapped to the human genome assembly GRCh38, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v157_hg19': {'description': 'dbSNP v157 hg19 is the release of the dbSNP database mapped to the human genome assembly GRCh37/hg19, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v157_hg38': {'description': 'dbSNP v157 hg38 is the release of the dbSNP database mapped to the human genome assembly GRCh38, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'ucsc_genome_hg19': {'description': 'UCSC Genome reference sequence on hg19 (GRCh37), providing the complete human genome assembly used for alignment and variant annotation.',\n",
       "  'suggested_use': 'Reference genome for alignment and annotation'},\n",
       " 'ucsc_genome_hg38': {'description': 'UCSC Genome reference sequence on hg38 (GRCh38), providing the complete human genome assembly used for alignment and variant annotation.',\n",
       "  'suggested_use': 'Reference genome for alignment and annotation'},\n",
       " '1kg_dbsnp151_hg19_auto': {'description': '1KG dbSNP151 hg19 auto variants table containing SNP ID, chromosome, position, effect allele, non-effect allele, and rsID mappings for autosomal variants.',\n",
       "  'suggested_use': 'SNP annotation and rsID mapping'},\n",
       " '1kg_dbsnp151_hg38_auto': {'description': '1KG dbSNP151 hg38 auto variants table containing SNP ID, chromosome, position, effect allele, non-effect allele, and rsID mappings for autosomal variants.',\n",
       "  'suggested_use': 'SNP annotation and rsID mapping'},\n",
       " 'recombination_hg19': {'description': 'Recombination map for hg19 (GRCh37), providing genetic distance estimates in centimorgans used for linkage disequilibrium and haplotype analysis.',\n",
       "  'suggested_use': 'Genetic mapping and LD calculations'},\n",
       " 'recombination_hg38': {'description': 'Recombination map for hg38 (GRCh38), providing genetic distance estimates in centimorgans used for linkage disequilibrium and haplotype analysis.',\n",
       "  'suggested_use': 'Genetic mapping and LD calculations'},\n",
       " 'ensembl_hg19_gtf': {'description': 'Ensembl annotation for hg19 (GRCh37) release 87, providing comprehensive gene models, transcript structures, and functional annotations for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'ensembl_hg38_gtf': {'description': 'Ensembl annotation for hg38 (GRCh38) release 109, providing comprehensive gene models, transcript structures, and functional annotations for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'refseq_hg19_gtf': {'description': 'RefSeq annotation for hg19 (GRCh37), providing curated gene models and transcript structures for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'refseq_hg38_gtf': {'description': 'RefSeq annotation for hg38 (GRCh38), providing curated gene models and transcript structures for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'testlink': {'description': 'Test link', 'suggested_use': ''},\n",
       " '19to38': {'description': 'LiftOver chain file for converting genomic coordinates from hg19 (GRCh37) to hg38 (GRCh38) reference builds, enabling cross-reference genome analysis.',\n",
       "  'suggested_use': 'Genome build conversion and coordinate remapping'},\n",
       " '38to19': {'description': 'LiftOver chain file for converting genomic coordinates from hg38 (GRCh38) to hg19 (GRCh37) reference builds, enabling cross-reference genome analysis.',\n",
       "  'suggested_use': 'Genome build conversion and coordinate remapping'},\n",
       " '1kg_hm3_hg38_eaf': {'description': '1KG HapMap3 hg38 EAF (GRCh38) provides allele frequency estimates for the HapMap3 populations on the hg38 reference genome, derived from 1000 Genomes Project data.',\n",
       "  'suggested_use': 'Ancestry inference and population structure analysis'},\n",
       " '1kg_hm3_hg19_eaf': {'description': '1KG HapMap3 hg19 EAF (GRCh37) provides allele frequency estimates for the HapMap3 populations on the hg19 reference genome, derived from 1000 Genomes Project data.',\n",
       "  'suggested_use': 'Ancestry inference and population structure analysis'}}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.check_available_ref()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aa0df3ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 11:12:06 Start to download  1kg_dbsnp151_hg38_auto  ...\n",
      "2026/01/10 11:12:06  -Downloading to: /home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz\n",
      "2026/01/10 11:12:49  - Updating record in config file...\n",
      "2026/01/10 11:12:49 Downloaded  1kg_dbsnp151_hg38_auto  successfully!\n"
     ]
    }
   ],
   "source": [
    "gl.download_ref(\"1kg_dbsnp151_hg38_auto\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e46a416e",
   "metadata": {},
   "source": [
    "### Load sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e799c0e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 11:19:17 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/10 11:19:17 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/10 11:19:17 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/10 11:19:17 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover.tsv.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 11:19:47  -Reading columns          : CHR,NEA,P,EAF,SNPID,BETA,POS,SE,EA\n",
      "2026/01/10 11:19:47  -Renaming columns to      : CHR,NEA,P,EAF,SNPID,BETA,POS,SE,EA\n",
      "2026/01/10 11:19:47  -Current Dataframe shape : 13640038  x  9\n",
      "2026/01/10 11:19:47  -Initiating a status column: STATUS ...\n",
      "2026/01/10 11:19:47 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/10 11:19:50 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 11:19:50  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/10 11:19:51 Finished reordering the columns.\n",
      "2026/01/10 11:19:51  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2026/01/10 11:19:53  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P      \n",
      "2026/01/10 11:19:53  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64\n",
      "2026/01/10 11:19:53  -Verified: T      T     T     T        T        T      T       T       T       T      \n",
      "2026/01/10 11:19:53  -Current Dataframe memory usage: 953.30 MB\n",
      "2026/01/10 11:19:53 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover.tsv.gz\",\n",
    "    snpid=\"SNPID\",\n",
    "    chrom=\"CHR\",\n",
    "    pos=\"POS\",\n",
    "    ea=\"EA\",            # In your data, it's now EA (was ALT)\n",
    "    nea=\"NEA\",          # In your data, it's now NEA (was REF)\n",
    "    eaf=\"EAF\",          # In your data, it's now EAF (was minor_AF/Frq)\n",
    "    beta=\"BETA\",\n",
    "    se=\"SE\",\n",
    "    p=\"P\",\n",
    "    # Note: 'Dir' and 'N' are not in your column list, \n",
    "    # so we should omit them or point to the correct ones.\n",
    "    # n=\"n_complete_samples\", # Optional: if you want to use sample size\n",
    "    verbose=True,\n",
    "    readargs={'sep': '\\t'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a71824a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>119016</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>119953</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>133743</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.097255</td>\n",
       "      <td>-0.010428</td>\n",
       "      <td>0.020804</td>\n",
       "      <td>0.616190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:717587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>144100</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.015679</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.049643</td>\n",
       "      <td>0.981132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:723329:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>149839</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.001732</td>\n",
       "      <td>-0.117102</td>\n",
       "      <td>0.146483</td>\n",
       "      <td>0.424044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640033</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>153165819</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640034</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>153166044</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640035</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>153166359</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640036</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>153166637</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640037</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>153166894</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13640038 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS       EAF  \\\n",
       "0             1:692794:CA:C    1     119016  C   CA  9999999  0.110590   \n",
       "1              1:693731:A:G    1     119953  G    A  9999999  0.115767   \n",
       "2              1:707522:G:C    1     133743  C    G  9999999  0.097255   \n",
       "3              1:717587:G:A    1     144100  A    G  9999999  0.015679   \n",
       "4              1:723329:A:T    1     149839  T    A  9999999  0.001732   \n",
       "...                     ...  ...        ... ..  ...      ...       ...   \n",
       "13640033    X:154929412:C:T   23  153165819  T    C  9999999  0.245473   \n",
       "13640034   X:154929637:CT:C   23  153166044  C   CT  9999999  0.229726   \n",
       "13640035  X:154929952:CAA:C   23  153166359  C  CAA  9999999  0.239430   \n",
       "13640036    X:154930230:A:G   23  153166637  G    A  9999999  0.245887   \n",
       "13640037    X:154930487:T:A   23  153166894  A    T  9999999  0.019559   \n",
       "\n",
       "              BETA        SE         P  \n",
       "0        -0.016436  0.019585  0.401342  \n",
       "1        -0.004255  0.018507  0.818155  \n",
       "2        -0.010428  0.020804  0.616190  \n",
       "3         0.001174  0.049643  0.981132  \n",
       "4        -0.117102  0.146483  0.424044  \n",
       "...            ...       ...       ...  \n",
       "13640033 -0.016260  0.010723  0.129427  \n",
       "13640034 -0.027098  0.011190  0.015456  \n",
       "13640035 -0.020494  0.011278  0.069202  \n",
       "13640036 -0.016347  0.010721  0.127334  \n",
       "13640037  0.007107  0.036783  0.846790  \n",
       "\n",
       "[13640038 rows x 10 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6922343d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 11:23:04 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/10 11:23:04  -Current Dataframe shape : 13640038 x 10 ; Memory usage: 953.30 MB\n",
      "2026/01/10 11:23:04  -Checking SNPID data type...\n",
      "2026/01/10 11:23:04  -Converted datatype for SNPID: object -> string\n",
      "2026/01/10 11:23:04  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/10 11:23:05  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/10 11:23:15 Finished checking SNPID/rsID.\n",
      "2026/01/10 11:23:15 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 11:23:15  -Checking CHR data type...\n",
      "2026/01/10 11:23:19  -Variants with standardized chromosome notation: 13640038\n",
      "2026/01/10 11:23:19  -All CHR are already fixed...\n",
      "2026/01/10 11:23:22 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 11:23:22 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 11:23:22  -Trying to convert datatype for POS: int64 -> Int64...\n",
      "2026/01/10 11:23:22  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 11:23:22  -No outlier variants were removed.\n",
      "2026/01/10 11:23:23 Finished fixing basepair positions (POS).\n",
      "2026/01/10 11:23:23 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/10 11:23:23  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/10 11:23:28  -All variants have valid alleles (EA and NEA are A/C/T/G, not NA, and different).\n",
      "2026/01/10 11:23:41 Finished fixing alleles (EA and NEA).\n",
      "2026/01/10 11:23:41 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/10 11:23:41  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/10 11:23:42  -Checking if any columns are empty...\n",
      "2026/01/10 11:23:42  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/10 11:23:43  -Checking if -1e-07 < P < 1.0000001 : removed 0 variants.\n",
      "2026/01/10 11:23:45  -Checking if -100.0000001 < BETA < 100.0000001 : removed 0 variants.\n",
      "2026/01/10 11:23:46  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/10 11:23:46  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/10 11:23:46  -No variants were removed with bad statistics in total.\n",
      "2026/01/10 11:23:46 Finished sanity check for statistics.\n",
      "2026/01/10 11:23:46 Start to check data consistency across columns ...(v4.0.4)\n",
      "2026/01/10 11:23:46  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "2026/01/10 11:23:46  -Checking if BETA/SE-derived-P is consistent with P...\n",
      "2026/01/10 11:23:47   -Variants with inconsistent values were not detected.\n",
      "2026/01/10 11:23:47   -Note: Minor differences are typically due to rounding.\n",
      "2026/01/10 11:23:47    -If the max difference is greater than expected, please check your original sumstats.\n",
      "2026/01/10 11:23:47 Finished checking data consistency across columns.\n",
      "2026/01/10 11:23:47 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/10 11:23:47  -Number of threads/cores to use: 1\n",
      "2026/01/10 11:23:47  -No available variants to normalize..\n",
      "2026/01/10 11:23:47 Finished normalizing indels.\n",
      "2026/01/10 11:23:47 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/10 11:24:01 Finished sorting coordinates.\n",
      "2026/01/10 11:24:01 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 11:24:01  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/10 11:24:01 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>119016</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>119953</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>133743</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.097255</td>\n",
       "      <td>-0.010428</td>\n",
       "      <td>0.020804</td>\n",
       "      <td>0.616190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:717587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>144100</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.015679</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.049643</td>\n",
       "      <td>0.981132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:723329:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>149839</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.001732</td>\n",
       "      <td>-0.117102</td>\n",
       "      <td>0.146483</td>\n",
       "      <td>0.424044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640033</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>153165819</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640034</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>153166044</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640035</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>153166359</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640036</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>153166637</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640037</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>153166894</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13640038 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x798ab8bd9d90>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.basic_check()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff842a81",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
