{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3df36a3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58114d66",
   "metadata": {},
   "source": [
    "### Check reference and download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ca8b2af3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 13:56:38 Start to check available reference files...\n",
      "2026/01/11 13:56:38  - Available keywords:  1kg_eas_hg19 1kg_eur_hg19 1kg_eas_hg38 1kg_eur_hg38 1kg_sas_hg19 1kg_amr_hg19 1kg_sas_hg38 1kg_amr_hg38 1kg_afr_hg19 1kg_pan_hg19 1kg_afr_hg38 1kg_pan_hg38 dbsnp_v151_hg19 dbsnp_v151_hg38 dbsnp_v157_hg19 dbsnp_v157_hg38 ucsc_genome_hg19 ucsc_genome_hg38 1kg_dbsnp151_hg19_auto 1kg_dbsnp151_hg38_auto recombination_hg19 recombination_hg38 ensembl_hg19_gtf ensembl_hg38_gtf refseq_hg19_gtf refseq_hg38_gtf testlink 19to38 38to19 1kg_hm3_hg38_eaf 1kg_hm3_hg19_eaf\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'1kg_eas_hg19': {'description': '1000 Genomes Project East Asian (1KG EAS) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EAS population'},\n",
       " '1kg_eur_hg19': {'description': '1000 Genomes Project European (1KG EUR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EUR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EUR population'},\n",
       " '1kg_eas_hg38': {'description': '1000 Genomes Project East Asian (1KG EAS) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EAS population'},\n",
       " '1kg_eur_hg38': {'description': '1000 Genomes Project European (1KG EUR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EUR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EUR population'},\n",
       " '1kg_sas_hg19': {'description': '1000 Genomes Project South Asian (1KG SAS) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the SAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for SAS population'},\n",
       " '1kg_amr_hg19': {'description': '1000 Genomes Project American (1KG AMR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AMR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AMR population'},\n",
       " '1kg_sas_hg38': {'description': '1000 Genomes Project South Asian (1KG SAS) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the SAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for SAS population'},\n",
       " '1kg_amr_hg38': {'description': '1000 Genomes Project American (1KG AMR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AMR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AMR population'},\n",
       " '1kg_afr_hg19': {'description': '1000 Genomes Project African (1KG AFR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AFR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AFR population'},\n",
       " '1kg_pan_hg19': {'description': '1000 Genomes Project Pan-ancestry (1KG PAN) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the PAN population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for PAN population'},\n",
       " '1kg_afr_hg38': {'description': '1000 Genomes Project African (1KG AFR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AFR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AFR population'},\n",
       " '1kg_pan_hg38': {'description': '1000 Genomes Project Pan-ancestry (1KG PAN) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the PAN population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for PAN population'},\n",
       " 'dbsnp_v151_hg19': {'description': 'dbSNP v151 hg19 is the 2018 release of the dbSNP database mapped to the human genome assembly GRCh37/hg19, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v151_hg38': {'description': 'dbSNP v151 hg38 is the 2018 release of the dbSNP database mapped to the human genome assembly GRCh38, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v157_hg19': {'description': 'dbSNP v157 hg19 is the release of the dbSNP database mapped to the human genome assembly GRCh37/hg19, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v157_hg38': {'description': 'dbSNP v157 hg38 is the release of the dbSNP database mapped to the human genome assembly GRCh38, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'ucsc_genome_hg19': {'description': 'UCSC Genome reference sequence on hg19 (GRCh37), providing the complete human genome assembly used for alignment and variant annotation.',\n",
       "  'suggested_use': 'Reference genome for alignment and annotation'},\n",
       " 'ucsc_genome_hg38': {'description': 'UCSC Genome reference sequence on hg38 (GRCh38), providing the complete human genome assembly used for alignment and variant annotation.',\n",
       "  'suggested_use': 'Reference genome for alignment and annotation'},\n",
       " '1kg_dbsnp151_hg19_auto': {'description': '1KG dbSNP151 hg19 auto variants table containing SNP ID, chromosome, position, effect allele, non-effect allele, and rsID mappings for autosomal variants.',\n",
       "  'suggested_use': 'SNP annotation and rsID mapping'},\n",
       " '1kg_dbsnp151_hg38_auto': {'description': '1KG dbSNP151 hg38 auto variants table containing SNP ID, chromosome, position, effect allele, non-effect allele, and rsID mappings for autosomal variants.',\n",
       "  'suggested_use': 'SNP annotation and rsID mapping'},\n",
       " 'recombination_hg19': {'description': 'Recombination map for hg19 (GRCh37), providing genetic distance estimates in centimorgans used for linkage disequilibrium and haplotype analysis.',\n",
       "  'suggested_use': 'Genetic mapping and LD calculations'},\n",
       " 'recombination_hg38': {'description': 'Recombination map for hg38 (GRCh38), providing genetic distance estimates in centimorgans used for linkage disequilibrium and haplotype analysis.',\n",
       "  'suggested_use': 'Genetic mapping and LD calculations'},\n",
       " 'ensembl_hg19_gtf': {'description': 'Ensembl annotation for hg19 (GRCh37) release 87, providing comprehensive gene models, transcript structures, and functional annotations for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'ensembl_hg38_gtf': {'description': 'Ensembl annotation for hg38 (GRCh38) release 109, providing comprehensive gene models, transcript structures, and functional annotations for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'refseq_hg19_gtf': {'description': 'RefSeq annotation for hg19 (GRCh37), providing curated gene models and transcript structures for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'refseq_hg38_gtf': {'description': 'RefSeq annotation for hg38 (GRCh38), providing curated gene models and transcript structures for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'testlink': {'description': 'Test link', 'suggested_use': ''},\n",
       " '19to38': {'description': 'LiftOver chain file for converting genomic coordinates from hg19 (GRCh37) to hg38 (GRCh38) reference builds, enabling cross-reference genome analysis.',\n",
       "  'suggested_use': 'Genome build conversion and coordinate remapping'},\n",
       " '38to19': {'description': 'LiftOver chain file for converting genomic coordinates from hg38 (GRCh38) to hg19 (GRCh37) reference builds, enabling cross-reference genome analysis.',\n",
       "  'suggested_use': 'Genome build conversion and coordinate remapping'},\n",
       " '1kg_hm3_hg38_eaf': {'description': '1KG HapMap3 hg38 EAF (GRCh38) provides allele frequency estimates for the HapMap3 populations on the hg38 reference genome, derived from 1000 Genomes Project data.',\n",
       "  'suggested_use': 'Ancestry inference and population structure analysis'},\n",
       " '1kg_hm3_hg19_eaf': {'description': '1KG HapMap3 hg19 EAF (GRCh37) provides allele frequency estimates for the HapMap3 populations on the hg19 reference genome, derived from 1000 Genomes Project data.',\n",
       "  'suggested_use': 'Ancestry inference and population structure analysis'}}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.check_available_ref()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aa0df3ed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 13:56:43 Start to download  1kg_dbsnp151_hg38_auto  ...\n",
      "2026/01/11 13:56:43  -Downloading to: /home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz\n",
      "2026/01/11 13:56:43  -File /home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz exists.\n",
      "2026/01/11 13:56:43  - Updating record in config file...\n",
      "2026/01/11 13:56:43 Downloaded  1kg_dbsnp151_hg38_auto  successfully!\n"
     ]
    }
   ],
   "source": [
    "gl.download_ref(\"1kg_dbsnp151_hg38_auto\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e46a416e",
   "metadata": {},
   "source": [
    "### Load sample data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "e799c0e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 13:56:58 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/11 13:56:58 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/11 13:56:58 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/11 13:56:58 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover.tsv.gz\n",
      "2026/01/11 13:57:22  -Reading columns          : SE,P,POS,BETA,SNPID,EAF,NEA,CHR,EA\n",
      "2026/01/11 13:57:22  -Renaming columns to      : SE,P,POS,BETA,SNPID,EAF,NEA,CHR,EA\n",
      "2026/01/11 13:57:22  -Current Dataframe shape : 13791467  x  9\n",
      "2026/01/11 13:57:23  -Initiating a status column: STATUS ...\n",
      "2026/01/11 13:57:23 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/11 13:57:26 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/11 13:57:26  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/11 13:57:26 Finished reordering the columns.\n",
      "2026/01/11 13:57:26  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2026/01/11 13:57:28  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P      \n",
      "2026/01/11 13:57:28  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64\n",
      "2026/01/11 13:57:28  -Verified: T      T     T     T        T        T      T       T       T       T      \n",
      "2026/01/11 13:57:28  -Current Dataframe memory usage: 963.86 MB\n",
      "2026/01/11 13:57:28 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover.tsv.gz\",\n",
    "    snpid=\"SNPID\",\n",
    "    chrom=\"CHR\",\n",
    "    pos=\"POS\",\n",
    "    ea=\"EA\",            # In your data, it's now EA (was ALT)\n",
    "    nea=\"NEA\",          # In your data, it's now NEA (was REF)\n",
    "    eaf=\"EAF\",          # In your data, it's now EAF (was minor_AF/Frq)\n",
    "    beta=\"BETA\",\n",
    "    se=\"SE\",\n",
    "    p=\"P\",\n",
    "    # Note: 'Dir' and 'N' are not in your column list, \n",
    "    # so we should omit them or point to the correct ones.\n",
    "    # n=\"n_complete_samples\", # Optional: if you want to use sample size\n",
    "    verbose=True,\n",
    "    readargs={'sep': '\\t'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a71824a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>894.616000</td>\n",
       "      <td>1204.870000</td>\n",
       "      <td>0.457786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS       EAF  \\\n",
       "0               1:15791:C:T    1      15791  T    C  9999999  1.000000   \n",
       "1               1:69487:G:A    1      69487  A    G  9999999  0.999994   \n",
       "2               1:69569:T:C    1      69569  C    T  9999999  0.999812   \n",
       "3              1:139853:C:T    1     139853  T    C  9999999  0.999994   \n",
       "4             1:692794:CA:C    1     692794  C   CA  9999999  0.889410   \n",
       "...                     ...  ...        ... ..  ...      ...       ...   \n",
       "13791462    X:154929412:C:T   23  154929412  T    C  9999999  0.754527   \n",
       "13791463   X:154929637:CT:C   23  154929637  C   CT  9999999  0.770274   \n",
       "13791464  X:154929952:CAA:C   23  154929952  C  CAA  9999999  0.760570   \n",
       "13791465    X:154930230:A:G   23  154930230  G    A  9999999  0.754113   \n",
       "13791466    X:154930487:T:A   23  154930487  A    T  9999999  0.980441   \n",
       "\n",
       "                BETA           SE         P  \n",
       "0         894.616000  1204.870000  0.457786  \n",
       "1          -2.715450     2.360060  0.249902  \n",
       "2          -0.484284     0.423462  0.252778  \n",
       "3          -2.703560     2.360130  0.251997  \n",
       "4          -0.016436     0.019585  0.401342  \n",
       "...              ...          ...       ...  \n",
       "13791462   -0.016260     0.010723  0.129427  \n",
       "13791463   -0.027098     0.011190  0.015456  \n",
       "13791464   -0.020494     0.011278  0.069202  \n",
       "13791465   -0.016347     0.010721  0.127334  \n",
       "13791466    0.007107     0.036783  0.846790  \n",
       "\n",
       "[13791467 rows x 10 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6922343d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 13:57:38 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/11 13:57:38  -Current Dataframe shape : 13791467 x 10 ; Memory usage: 963.86 MB\n",
      "2026/01/11 13:57:38  -Checking SNPID data type...\n",
      "2026/01/11 13:57:38  -Converted datatype for SNPID: object -> string\n",
      "2026/01/11 13:57:38  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/11 13:57:39  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/11 13:57:49 Finished checking SNPID/rsID.\n",
      "2026/01/11 13:57:49 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/11 13:57:49  -Checking CHR data type...\n",
      "2026/01/11 13:57:53  -Variants with standardized chromosome notation: 13791467\n",
      "2026/01/11 13:57:53  -All CHR are already fixed...\n",
      "2026/01/11 13:57:56 Finished fixing chromosome notation (CHR).\n",
      "2026/01/11 13:57:56 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/11 13:57:56  -Trying to convert datatype for POS: int64 -> Int64...\n",
      "2026/01/11 13:57:57  -Position bound:(0 , 250,000,000)\n",
      "2026/01/11 13:57:57  -No outlier variants were removed.\n",
      "2026/01/11 13:57:58 Finished fixing basepair positions (POS).\n",
      "2026/01/11 13:57:58 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/11 13:57:58  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/11 13:58:03  -Variants with bad EA: 0\n",
      "2026/01/11 13:58:03  -Variants with bad NEA: 0\n",
      "2026/01/11 13:58:03  -Variants with NA for EA or NEA: 0\n",
      "2026/01/11 13:58:03  -Variants with same EA and NEA: 1950604\n",
      "2026/01/11 13:58:16 Finished fixing alleles (EA and NEA).\n",
      "2026/01/11 13:58:16 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/11 13:58:16  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/11 13:58:17  -Checking if any columns are empty...\n",
      "2026/01/11 13:58:18  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/11 13:58:18   -Examples of invalid variants(SNPID): 1:776572:C:T,1:819974:G:C,1:871173:C:T,1:871198:C:T,1:874451:C:T ...\n",
      "2026/01/11 13:58:18   -Examples of invalid values (P): NA,NA,NA,NA,NA ...\n",
      "2026/01/11 13:58:19  -Checking if -1e-07 < P < 1.0000001 : removed 1676 variants.\n",
      "2026/01/11 13:58:19   -Examples of invalid variants(SNPID): 1:15791:C:T,1:877608:C:G,1:897032:A:G,1:899336:G:A,1:907676:A:T ...\n",
      "2026/01/11 13:58:19   -Examples of invalid values (BETA): 894.616,279.71,-108.064,270.068,116.14 ...\n",
      "2026/01/11 13:58:20  -Checking if -100.0000001 < BETA < 100.0000001 : removed 9906 variants.\n",
      "2026/01/11 13:58:21  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/11 13:58:21  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/11 13:58:21  -Removed 11582 variants with bad statistics in total.\n",
      "2026/01/11 13:58:21 Finished sanity check for statistics.\n",
      "2026/01/11 13:58:21 Start to check data consistency across columns ...(v4.0.4)\n",
      "2026/01/11 13:58:21  -Current Dataframe shape : 13779885 x 10 ; Memory usage: 1043.38 MB\n",
      "2026/01/11 13:58:21  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "2026/01/11 13:58:21  -Checking if BETA/SE-derived-P is consistent with P...\n",
      "2026/01/11 13:58:22   -Variants with inconsistent values were not detected.\n",
      "2026/01/11 13:58:22   -Note: Minor differences are typically due to rounding.\n",
      "2026/01/11 13:58:22    -If the max difference is greater than expected, please check your original sumstats.\n",
      "2026/01/11 13:58:22 Finished checking data consistency across columns.\n",
      "2026/01/11 13:58:22 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/11 13:58:22  -Number of threads/cores to use: 1\n",
      "2026/01/11 13:58:23  -Number of variants to check:558\n",
      "2026/01/11 13:58:23  -Chunk size:3000000\n",
      "2026/01/11 13:58:23  -Processing in chunks:0 \n",
      "2026/01/11 13:58:25 Finished normalizing indels.\n",
      "2026/01/11 13:58:25 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/11 13:58:38 Finished sorting coordinates.\n",
      "2026/01/11 13:58:38 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/11 13:58:38  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/11 13:58:38 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7759d836e3f0>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.basic_check()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ff842a81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 13:59:35 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/11 13:59:35  -Checking SNPID data type...\n",
      "2026/01/11 13:59:35  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/11 13:59:36  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/11 13:59:46  -Replacing separators in SNPID with \":\" ...\n",
      "2026/01/11 13:59:52 Finished checking SNPID/rsID.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7759d836e3f0>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.fix_id(fixsep=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0660d0d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS       EAF  \\\n",
       "0               1:69487:G:A    1      69487  A    G  9960099  0.999994   \n",
       "1               1:69569:T:C    1      69569  C    T  9960099  0.999812   \n",
       "2              1:139853:C:T    1     139853  T    C  9960099  0.999994   \n",
       "3             1:692794:CA:C    1     692794  C   CA  9960399  0.889410   \n",
       "4              1:693731:A:G    1     693731  G    A  9960099  0.884233   \n",
       "...                     ...  ...        ... ..  ...      ...       ...   \n",
       "13779880    X:154929412:C:T   23  154929412  T    C  9960099  0.754527   \n",
       "13779881   X:154929637:CT:C   23  154929637  C   CT  9960399  0.770274   \n",
       "13779882  X:154929952:CAA:C   23  154929952  C  CAA  9960399  0.760570   \n",
       "13779883    X:154930230:A:G   23  154930230  G    A  9960099  0.754113   \n",
       "13779884    X:154930487:T:A   23  154930487  A    T  9960099  0.980441   \n",
       "\n",
       "              BETA        SE         P  \n",
       "0        -2.715450  2.360060  0.249902  \n",
       "1        -0.484284  0.423462  0.252778  \n",
       "2        -2.703560  2.360130  0.251997  \n",
       "3        -0.016436  0.019585  0.401342  \n",
       "4        -0.004255  0.018507  0.818155  \n",
       "...            ...       ...       ...  \n",
       "13779880 -0.016260  0.010723  0.129427  \n",
       "13779881 -0.027098  0.011190  0.015456  \n",
       "13779882 -0.020494  0.011278  0.069202  \n",
       "13779883 -0.016347  0.010721  0.127334  \n",
       "13779884  0.007107  0.036783  0.846790  \n",
       "\n",
       "[13779885 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72fc4d30",
   "metadata": {},
   "source": [
    "### Assign rsID"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de4110a9",
   "metadata": {},
   "source": [
    "#### assign rsID using a SNPID-rsID table (variants in 1KG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a72f27fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 14:01:10 Start to assign rsID using reference file ...(v4.0.4)\n",
      "2026/01/11 14:01:10  -Number of threads/cores to use: 1\n",
      "2026/01/11 14:01:11  -13779885 rsID could be possibly fixed...\n",
      "2026/01/11 14:01:12  -Setting block size:  5000000\n",
      "2026/01/11 14:01:12  -Loading block: 0   1   2   3   "
     ]
    }
   ],
   "source": [
    "mysumstats.assign_rsid(ref_rsid_tsv=gl.get_path(\"1kg_dbsnp151_hg38_auto\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd3c7d1",
   "metadata": {},
   "source": [
    "#### Note: you may need a dictionay to match the chromosome with your VCF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "75187f89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: 'NC_000001.11',\n",
       " 2: 'NC_000002.12',\n",
       " 3: 'NC_000003.12',\n",
       " 4: 'NC_000004.12',\n",
       " 5: 'NC_000005.10',\n",
       " 6: 'NC_000006.12',\n",
       " 7: 'NC_000007.14',\n",
       " 8: 'NC_000008.11',\n",
       " 9: 'NC_000009.12',\n",
       " 10: 'NC_000010.11',\n",
       " 11: 'NC_000011.10',\n",
       " 12: 'NC_000012.12',\n",
       " 13: 'NC_000013.11',\n",
       " 14: 'NC_000014.9',\n",
       " 15: 'NC_000015.10',\n",
       " 16: 'NC_000016.10',\n",
       " 17: 'NC_000017.11',\n",
       " 18: 'NC_000018.10',\n",
       " 19: 'NC_000019.10',\n",
       " 20: 'NC_000020.11',\n",
       " 21: 'NC_000021.9',\n",
       " 22: 'NC_000022.11',\n",
       " 23: 'NC_000023.11',\n",
       " 24: 'NC_000024.10',\n",
       " 25: 'NC_012920.1'}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.get_number_to_NC(build=\"38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bcf47063",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: 'NC_000001.10',\n",
       " 2: 'NC_000002.11',\n",
       " 3: 'NC_000003.11',\n",
       " 4: 'NC_000004.11',\n",
       " 5: 'NC_000005.9',\n",
       " 6: 'NC_000006.11',\n",
       " 7: 'NC_000007.13',\n",
       " 8: 'NC_000008.10',\n",
       " 9: 'NC_000009.11',\n",
       " 10: 'NC_000010.10',\n",
       " 11: 'NC_000011.9',\n",
       " 12: 'NC_000012.11',\n",
       " 13: 'NC_000013.10',\n",
       " 14: 'NC_000014.8',\n",
       " 15: 'NC_000015.9',\n",
       " 16: 'NC_000016.9',\n",
       " 17: 'NC_000017.10',\n",
       " 18: 'NC_000018.9',\n",
       " 19: 'NC_000019.9',\n",
       " 20: 'NC_000020.10',\n",
       " 21: 'NC_000021.8',\n",
       " 22: 'NC_000022.10',\n",
       " 23: 'NC_000023.10',\n",
       " 24: 'NC_000024.9',\n",
       " 25: 'NC_012920.1'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.get_number_to_NC(build=\"19\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47db4eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to a gzipped TSV file\n",
    "mysumstats.to_csv(\"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\", \n",
    "                  sep=\"\\t\", \n",
    "                  index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
