{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7d48475f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0,\"/home/yunye/work/gwaslab/src\")\n",
    "import gwaslab as gl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "623ca292",
   "metadata": {},
   "source": [
    "### loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1217ef70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 13:31:45 Start to check downloaded reference files...\n",
      "2026/01/10 13:31:45  -Checking the config file:/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/data/config.json\n",
      "2026/01/10 13:31:45  -Config file exists.\n",
      "2026/01/10 13:31:45  -Updating config.json...\n",
      "2026/01/10 13:31:45  - Local file keywords:  1kg_dbsnp151_hg38_auto\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'1kg_dbsnp151_hg38_auto': {'local_path': '/home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz',\n",
       "  'description': '1KG dbSNP151 hg38 auto variants table containing SNP ID, chromosome, position, effect allele, non-effect allele, and rsID mappings for autosomal variants.',\n",
       "  'suggested_use': 'SNP annotation and rsID mapping'}}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.check_downloaded_ref()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d56b7c03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 12:17:58 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/10 12:17:58 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/10 12:17:58 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/10 12:17:58 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\n",
      "2026/01/10 12:18:22  -Reading columns          : BETA,SNPID,SE,NEA,EA,P,CHR,EAF,POS\n",
      "2026/01/10 12:18:22  -Renaming columns to      : BETA,SNPID,SE,NEA,EA,P,CHR,EAF,POS\n",
      "2026/01/10 12:18:22  -Current Dataframe shape : 13640038  x  9\n",
      "2026/01/10 12:18:23  -Initiating a status column: STATUS ...\n",
      "2026/01/10 12:18:23 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/10 12:18:26 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 12:18:26  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/10 12:18:26 Finished reordering the columns.\n",
      "2026/01/10 12:18:27  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2026/01/10 12:18:28  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P      \n",
      "2026/01/10 12:18:28  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64\n",
      "2026/01/10 12:18:28  -Verified: T      T     T     T        T        T      T       T       T       T      \n",
      "2026/01/10 12:18:29  -Current Dataframe memory usage: 953.30 MB\n",
      "2026/01/10 12:18:29 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\",\n",
    "    snpid=\"SNPID\",\n",
    "    chrom=\"CHR\",\n",
    "    pos=\"POS\",\n",
    "    ea=\"EA\",            \n",
    "    nea=\"NEA\",          \n",
    "    eaf=\"EAF\",          \n",
    "    beta=\"BETA\",\n",
    "    se=\"SE\",\n",
    "    p=\"P\",\n",
    "    # Note: 'Dir' and 'N' are not in your column list, \n",
    "    # so we should omit them or point to the correct ones.\n",
    "    # n=\"n_complete_samples\", # Optional: if you want to use sample size\n",
    "    verbose=True,\n",
    "    readargs={'sep': '\\t'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a978d0ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 12:19:08 Start to infer strand for palindromic SNPs/align indistinguishable indels ...(v4.0.4)\n",
      "2026/01/10 12:19:08  -Number of threads/cores to use: 4\n",
      "2026/01/10 12:19:08  -Current Dataframe shape : 13640038 x 10 ; Memory usage: 953.30 MB\n",
      "2026/01/10 12:19:08 #WARNING! Reference file not found: /mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr%CHR.vcf.gz. Using default format.\n",
      "2026/01/10 12:19:08  -Field for alternative allele frequency in VCF INFO: AF\n",
      "2026/01/10 12:19:09  -Identified  0  palindromic SNPs...\n",
      "2026/01/10 12:19:09  -After filtering by MAF<= 0.4 , 0 palindromic SNPs with unknown strand will be inferred...\n",
      "2026/01/10 12:19:09 #WARNING! No palindromic variants available for checking.\n",
      "2026/01/10 12:19:11   -Non-palindromic :  0\n",
      "2026/01/10 12:19:12   -Palindromic SNPs on + strand:  0\n",
      "2026/01/10 12:19:13   -Palindromic SNPs on - strand and needed to be flipped: 0\n",
      "2026/01/10 12:19:14   -Palindromic SNPs with MAF not available to infer :  0\n",
      "2026/01/10 12:19:16   -Palindromic SNPs lacking reference matches, or with reference ALT unavailable or above MAF threshold for inference :  0\n",
      "2026/01/10 12:19:16  -Identified  0  indistinguishable Indels...\n",
      "2026/01/10 12:19:16 #WARNING! No indistinguishable indels available for checking.\n",
      "2026/01/10 12:19:16 Finished inferring strand for palindromic SNPs/align indistinguishable indels.\n",
      "2026/01/10 12:19:16 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/10 12:19:17  -No statistics have been changed.\n",
      "2026/01/10 12:19:17 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/10 12:19:17 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/10 12:19:17  -Trying to convert datatype for POS: int64 -> Int64...\n",
      "2026/01/10 12:19:32 Finished sorting coordinates.\n",
      "2026/01/10 12:19:32 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 12:19:32  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/10 12:19:33 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>119016</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>119953</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>133743</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.097255</td>\n",
       "      <td>-0.010428</td>\n",
       "      <td>0.020804</td>\n",
       "      <td>0.616190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:717587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>144100</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.015679</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.049643</td>\n",
       "      <td>0.981132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:723329:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>149839</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.001732</td>\n",
       "      <td>-0.117102</td>\n",
       "      <td>0.146483</td>\n",
       "      <td>0.424044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640033</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>153165819</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640034</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>153166044</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640035</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>153166359</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640036</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>153166637</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13640037</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>153166894</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13640038 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x706f5557ffe0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Wildcard path for your hg38 VCFs\n",
    "ref_pattern = \"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr%CHR.vcf.gz\"\n",
    "\n",
    "mysumstats.harmonize(\n",
    "    ref_infer=ref_pattern,    # Path to your VCF reference\n",
    "    ref_alt_freq=\"AF\",        # Field in VCF for freq comparison\n",
    "    basic_check=False,        # Skipping as you already ran it\n",
    "    threads=4                 # Parallel processing\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "402d4f9e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
