{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a1174ec2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cda9d7ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/02 22:25:46 GWASLab v4.0.2 https://cloufield.github.io/gwaslab/\n",
      "2026/01/02 22:25:46 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/02 22:25:46 Python version: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "gl.show_version()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "06708bf8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/02 22:28:30 GWASLab v4.0.2 https://cloufield.github.io/gwaslab/\n",
      "2026/01/02 22:28:30 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/02 22:28:30 Python version: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]\n",
      "2026/01/02 22:28:30 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxytool/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/02 22:28:59  -Reading columns          : beta,minor_allele,variant,minor_AF,se,pval,n_complete_samples\n",
      "2026/01/02 22:28:59  -Renaming columns to      : BETA,EA,SNPID,MAF,SE,P,N\n",
      "2026/01/02 22:28:59  -Current Dataframe shape : 13791467  x  7\n",
      "2026/01/02 22:29:00  -Initiating a status column: STATUS ...\n",
      "2026/01/02 22:29:00 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/02 22:29:01 Start to reorder the columns ...(v4.0.2)\n",
      "2026/01/02 22:29:01  -Reordering columns to    : SNPID,EA,STATUS,MAF,BETA,SE,P,N\n",
      "2026/01/02 22:29:05 Finished reordering the columns.\n",
      "2026/01/02 22:29:05  -Column  : SNPID  EA       STATUS MAF     BETA    SE      P       N    \n",
      "2026/01/02 22:29:05  -DType   : object category int64  float64 float64 float64 float64 int64\n",
      "2026/01/02 22:29:05  -Verified: T      T        T      T       T       T       T       T    \n",
      "2026/01/02 22:29:05  -Current Dataframe memory usage: 790.42 MB\n",
      "2026/01/02 22:29:05 Finished loading data successfully!\n",
      "2026/01/02 22:29:05 Start to check SNPID/rsID ...(v4.0.2)\n",
      "2026/01/02 22:29:05  -Current Dataframe shape : 13791467 x 8 ; Memory usage: 790.42 MB\n",
      "2026/01/02 22:29:05  -Checking SNPID data type...\n",
      "2026/01/02 22:29:05  -Converted datatype for SNPID: object -> string\n",
      "2026/01/02 22:29:06  -Checking NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,\n",
      "2026/01/02 22:29:06  -Checking if SNPID contains NA strings...\n",
      "2026/01/02 22:29:06  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/02 22:29:15 Finished checking SNPID/rsID.\n",
      "2026/01/02 22:29:15 Start to fix chromosome notation (CHR) ...(v4.0.2)\n",
      "2026/01/02 22:29:15 #WARNING! Necessary columns for .fix_chr() were not detected:CHR\n",
      "2026/01/02 22:29:15 Start to fix basepair positions (POS) ...(v4.0.2)\n",
      "2026/01/02 22:29:15 #WARNING! Necessary columns for .fix_pos() were not detected:POS\n",
      "2026/01/02 22:29:15 Start to fix alleles (EA and NEA) ...(v4.0.2)\n",
      "2026/01/02 22:29:15 #WARNING! Necessary columns for .fix_allele() were not detected:NEA\n",
      "2026/01/02 22:29:15 Start to perform sanity check for statistics ...(v4.0.2)\n",
      "2026/01/02 22:29:15  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/02 22:29:17  -Checking if any columns are empty...\n",
      "2026/01/02 22:29:17  -Checking if 0 <= N <= 2147483647 ...\n",
      "2026/01/02 22:29:24  -Removed 0 variants with bad/na N.\n",
      "2026/01/02 22:29:24  -Checking if -1e-07 < P < 1.0000001 ...\n",
      "2026/01/02 22:29:24   -Examples of invalid variants(SNPID): 1:776572:C:T,1:819974:G:C,1:871173:C:T,1:871198:C:T,1:874451:C:T ...\n",
      "2026/01/02 22:29:24   -Examples of invalid values (P): NA,NA,NA,NA,NA ...\n",
      "2026/01/02 22:29:27  -Removed 1676 variants with bad/na P.\n",
      "2026/01/02 22:29:27  -Checking if -100.0000001 < BETA < 100.0000001 ...\n",
      "2026/01/02 22:29:27   -Examples of invalid variants(SNPID): 1:15791:C:T,1:877608:C:G,1:897032:A:G,1:899336:G:A,1:907676:A:T ...\n",
      "2026/01/02 22:29:27   -Examples of invalid values (BETA): 894.616,279.71,-108.064,270.068,116.14 ...\n",
      "2026/01/02 22:29:31  -Removed 9906 variants with bad/na BETA.\n",
      "2026/01/02 22:29:31  -Checking if -1e-07 < SE < inf ...\n",
      "2026/01/02 22:29:34  -Removed 0 variants with bad/na SE.\n",
      "2026/01/02 22:29:34  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/02 22:29:34  -Removed 11582 variants with bad statistics in total.\n",
      "2026/01/02 22:29:34  -Data types for each column:\n",
      "2026/01/02 22:29:34 Finished sanity check for statistics.\n",
      "2026/01/02 22:29:34 Start to check data consistency across columns ...(v4.0.2)\n",
      "2026/01/02 22:29:34  -Current Dataframe shape : 13779885 x 8 ; Memory usage: 921.17 MB\n",
      "2026/01/02 22:29:34  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "2026/01/02 22:29:34  -Checking if BETA/SE-derived-P is consistent with P...\n",
      "2026/01/02 22:29:56   -Variants with inconsistent values were not detected.\n",
      "2026/01/02 22:29:56  -Note: Minor differences are typically due to rounding.\n",
      "2026/01/02 22:29:56   -If the max difference is greater than expected, please check your original sumstats.\n",
      "2026/01/02 22:29:56 Finished checking data consistency across columns.\n",
      "2026/01/02 22:29:56  -Number of threads/cores to use: 1\n",
      "2026/01/02 22:29:56 Start to normalize indels ...(v4.0.2)\n",
      "2026/01/02 22:29:56 #WARNING! Necessary columns for .normalize() were not detected:NEA\n",
      "2026/01/02 22:29:56 Start to sort the genome coordinates ...(v4.0.2)\n",
      "2026/01/02 22:29:56 #WARNING! Necessary columns for .sort_coordinate() were not detected:CHR,POS\n",
      "2026/01/02 22:29:56 Start to reorder the columns ...(v4.0.2)\n",
      "2026/01/02 22:29:56  -Reordering columns to    : SNPID,EA,STATUS,MAF,BETA,SE,P,N\n",
      "2026/01/02 22:29:57 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>EA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>MAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7f0ce065e960>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxytool/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "    snpid=\"variant\",\n",
    "    ea=\"minor_allele\",\n",
    "    beta=\"beta\",   # <-- use your beta column\n",
    "    se=\"se\",       # <-- use your SE column\n",
    "    p=\"pval\",\n",
    "    n=\"n_complete_samples\",\n",
    "    maf=\"minor_AF\",\n",
    "    readargs={\"compression\":\"gzip\"},\n",
    "    verbose=True\n",
    ")\n",
    "mysumstats.basic_check()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "55a5d38b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>EA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>MAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>N</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>359983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID EA   STATUS       MAF      BETA        SE  \\\n",
       "1               1:69487:G:A  A  9969999  0.000006 -2.715450  2.360060   \n",
       "2               1:69569:T:C  C  9969999  0.000188 -0.484284  0.423462   \n",
       "3              1:139853:C:T  T  9969999  0.000006 -2.703560  2.360130   \n",
       "4             1:692794:CA:C  C  9969999  0.110590 -0.016436  0.019585   \n",
       "5              1:693731:A:G  G  9969999  0.115767 -0.004255  0.018507   \n",
       "...                     ... ..      ...       ...       ...       ...   \n",
       "13791462    X:154929412:C:T  T  9969999  0.245473 -0.016260  0.010723   \n",
       "13791463   X:154929637:CT:C  C  9969999  0.229726 -0.027098  0.011190   \n",
       "13791464  X:154929952:CAA:C  C  9969999  0.239430 -0.020494  0.011278   \n",
       "13791465    X:154930230:A:G  G  9969999  0.245887 -0.016347  0.010721   \n",
       "13791466    X:154930487:T:A  A  9969999  0.019559  0.007107  0.036783   \n",
       "\n",
       "                 P       N  \n",
       "1         0.249902  359983  \n",
       "2         0.252778  359983  \n",
       "3         0.251997  359983  \n",
       "4         0.401342  359983  \n",
       "5         0.818155  359983  \n",
       "...            ...     ...  \n",
       "13791462  0.129427  359983  \n",
       "13791463  0.015456  359983  \n",
       "13791464  0.069202  359983  \n",
       "13791465  0.127334  359983  \n",
       "13791466  0.846790  359983  \n",
       "\n",
       "[13779885 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f0c550",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
