{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9a367f5d",
   "metadata": {},
   "source": [
    "### Liftover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d5621adb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "86ffd018",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 11:29:39 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/11 11:29:39 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/11 11:29:39 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n"
     ]
    }
   ],
   "source": [
    "gl.show_version()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1ad06f48",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "read_table() got an unexpected keyword argument 'vcfid'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mTypeError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[4]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# Load your UKBB data\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m mysumstats = \u001b[43mgl\u001b[49m\u001b[43m.\u001b[49m\u001b[43mSumstats\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      3\u001b[39m \u001b[43m                         \u001b[49m\u001b[43mvcfid\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mvariant\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# Parses into SNPID, CHR, POS, EA, NEA\u001b[39;49;00m\n\u001b[32m      4\u001b[39m \u001b[43m                         \u001b[49m\u001b[43meaf\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mminor_AF\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      5\u001b[39m \u001b[43m                         \u001b[49m\u001b[43mbeta\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mbeta\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      6\u001b[39m \u001b[43m                         \u001b[49m\u001b[43mse\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mse\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      7\u001b[39m \u001b[43m                         \u001b[49m\u001b[43mp\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mpval\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      8\u001b[39m \u001b[43m                         \u001b[49m\u001b[43mn\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mn_complete_samples\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      9\u001b[39m \u001b[43m                         \u001b[49m\u001b[43mother\u001b[49m\u001b[43m=\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mlow_confidence_variant\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# Track this for filtering\u001b[39;49;00m\n\u001b[32m     10\u001b[39m \u001b[43m                        \u001b[49m\u001b[38;5;66;43;03m#  nrows=5000,  # For demo; remove for full file\u001b[39;49;00m\n\u001b[32m     11\u001b[39m \u001b[43m                         \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/g_Sumstats.py:251\u001b[39m, in \u001b[36mSumstats.__init__\u001b[39m\u001b[34m(self, sumstats, fmt, tab_fmt, snpid, rsid, chrom, pos, ea, nea, ref, alt, eaf, neaf, maf, n, beta, se, chisq, z, f, t, p, q, mlog10p, test, info, OR, OR_95L, OR_95U, beta_95L, beta_95U, HR, HR_95L, HR_95U, ncase, ncontrol, neff, i2, phet, dof, snpr2, status, other, exclude, include, chrom_pat, snpid_pat, direction, verbose, study, trait, build, species, readargs, **kwreadargs)\u001b[39m\n\u001b[32m    248\u001b[39m \u001b[38;5;28mself\u001b[39m._last_shape = \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m    250\u001b[39m \u001b[38;5;66;03m#preformat the data\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m251\u001b[39m \u001b[38;5;28mself\u001b[39m.data  = \u001b[43m_preformat\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    252\u001b[39m \u001b[43m  \u001b[49m\u001b[43msumstats\u001b[49m\u001b[43m=\u001b[49m\u001b[43msumstats\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    253\u001b[39m \u001b[43m  \u001b[49m\u001b[43mfmt\u001b[49m\u001b[43m=\u001b[49m\u001b[43mfmt\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    254\u001b[39m \u001b[43m  \u001b[49m\u001b[43mtab_fmt\u001b[49m\u001b[43m \u001b[49m\u001b[43m=\u001b[49m\u001b[43m \u001b[49m\u001b[43mtab_fmt\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    255\u001b[39m \u001b[43m  \u001b[49m\u001b[43msnpid\u001b[49m\u001b[43m=\u001b[49m\u001b[43msnpid\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    256\u001b[39m \u001b[43m  \u001b[49m\u001b[43mrsid\u001b[49m\u001b[43m=\u001b[49m\u001b[43mrsid\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    257\u001b[39m \u001b[43m  \u001b[49m\u001b[43mchrom\u001b[49m\u001b[43m=\u001b[49m\u001b[43mchrom\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    258\u001b[39m \u001b[43m  \u001b[49m\u001b[43mpos\u001b[49m\u001b[43m=\u001b[49m\u001b[43mpos\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    259\u001b[39m \u001b[43m  \u001b[49m\u001b[43mea\u001b[49m\u001b[43m=\u001b[49m\u001b[43mea\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    260\u001b[39m \u001b[43m  \u001b[49m\u001b[43mnea\u001b[49m\u001b[43m=\u001b[49m\u001b[43mnea\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    261\u001b[39m \u001b[43m  \u001b[49m\u001b[43mref\u001b[49m\u001b[43m=\u001b[49m\u001b[43mref\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    262\u001b[39m \u001b[43m  \u001b[49m\u001b[43malt\u001b[49m\u001b[43m=\u001b[49m\u001b[43malt\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    263\u001b[39m \u001b[43m  \u001b[49m\u001b[43meaf\u001b[49m\u001b[43m=\u001b[49m\u001b[43meaf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    264\u001b[39m \u001b[43m  \u001b[49m\u001b[43mneaf\u001b[49m\u001b[43m=\u001b[49m\u001b[43mneaf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    265\u001b[39m \u001b[43m  \u001b[49m\u001b[43mmaf\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmaf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    266\u001b[39m \u001b[43m  \u001b[49m\u001b[43mn\u001b[49m\u001b[43m=\u001b[49m\u001b[43mn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    267\u001b[39m \u001b[43m  \u001b[49m\u001b[43mbeta\u001b[49m\u001b[43m=\u001b[49m\u001b[43mbeta\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    268\u001b[39m \u001b[43m  \u001b[49m\u001b[43mse\u001b[49m\u001b[43m=\u001b[49m\u001b[43mse\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    269\u001b[39m \u001b[43m  \u001b[49m\u001b[43mchisq\u001b[49m\u001b[43m=\u001b[49m\u001b[43mchisq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    270\u001b[39m \u001b[43m  \u001b[49m\u001b[43mz\u001b[49m\u001b[43m=\u001b[49m\u001b[43mz\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    271\u001b[39m \u001b[43m  \u001b[49m\u001b[43mf\u001b[49m\u001b[43m=\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    272\u001b[39m \u001b[43m  \u001b[49m\u001b[43mt\u001b[49m\u001b[43m=\u001b[49m\u001b[43mt\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    273\u001b[39m \u001b[43m  \u001b[49m\u001b[43mp\u001b[49m\u001b[43m=\u001b[49m\u001b[43mp\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    274\u001b[39m \u001b[43m  \u001b[49m\u001b[43mq\u001b[49m\u001b[43m=\u001b[49m\u001b[43mq\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    275\u001b[39m \u001b[43m  \u001b[49m\u001b[43mmlog10p\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmlog10p\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    276\u001b[39m \u001b[43m  \u001b[49m\u001b[43mtest\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtest\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    277\u001b[39m \u001b[43m  \u001b[49m\u001b[43minfo\u001b[49m\u001b[43m=\u001b[49m\u001b[43minfo\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    278\u001b[39m \u001b[43m  \u001b[49m\u001b[43mOR\u001b[49m\u001b[43m=\u001b[49m\u001b[43mOR\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    279\u001b[39m \u001b[43m  \u001b[49m\u001b[43mOR_95L\u001b[49m\u001b[43m=\u001b[49m\u001b[43mOR_95L\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    280\u001b[39m \u001b[43m  \u001b[49m\u001b[43mOR_95U\u001b[49m\u001b[43m=\u001b[49m\u001b[43mOR_95U\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    281\u001b[39m \u001b[43m  \u001b[49m\u001b[43mbeta_95L\u001b[49m\u001b[43m=\u001b[49m\u001b[43mbeta_95L\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    282\u001b[39m \u001b[43m  \u001b[49m\u001b[43mbeta_95U\u001b[49m\u001b[43m=\u001b[49m\u001b[43mbeta_95U\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    283\u001b[39m \u001b[43m  \u001b[49m\u001b[43mHR\u001b[49m\u001b[43m=\u001b[49m\u001b[43mHR\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    284\u001b[39m \u001b[43m  \u001b[49m\u001b[43mHR_95L\u001b[49m\u001b[43m=\u001b[49m\u001b[43mHR_95L\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    285\u001b[39m \u001b[43m  \u001b[49m\u001b[43mHR_95U\u001b[49m\u001b[43m=\u001b[49m\u001b[43mHR_95U\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    286\u001b[39m \u001b[43m  \u001b[49m\u001b[43mi2\u001b[49m\u001b[43m=\u001b[49m\u001b[43mi2\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    287\u001b[39m \u001b[43m  \u001b[49m\u001b[43mphet\u001b[49m\u001b[43m=\u001b[49m\u001b[43mphet\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    288\u001b[39m \u001b[43m  \u001b[49m\u001b[43mdof\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdof\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    289\u001b[39m \u001b[43m  \u001b[49m\u001b[43msnpr2\u001b[49m\u001b[43m=\u001b[49m\u001b[43msnpr2\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    290\u001b[39m \u001b[43m  \u001b[49m\u001b[43mncase\u001b[49m\u001b[43m=\u001b[49m\u001b[43mncase\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    291\u001b[39m \u001b[43m  \u001b[49m\u001b[43mncontrol\u001b[49m\u001b[43m=\u001b[49m\u001b[43mncontrol\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    292\u001b[39m \u001b[43m  \u001b[49m\u001b[43mneff\u001b[49m\u001b[43m=\u001b[49m\u001b[43mneff\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    293\u001b[39m \u001b[43m  \u001b[49m\u001b[43mdirection\u001b[49m\u001b[43m=\u001b[49m\u001b[43mdirection\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    294\u001b[39m \u001b[43m  \u001b[49m\u001b[43mstudy\u001b[49m\u001b[43m=\u001b[49m\u001b[43mstudy\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    295\u001b[39m \u001b[43m  \u001b[49m\u001b[43mbuild\u001b[49m\u001b[43m=\u001b[49m\u001b[43mbuild\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    296\u001b[39m \u001b[43m  \u001b[49m\u001b[43mtrait\u001b[49m\u001b[43m=\u001b[49m\u001b[43mtrait\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    297\u001b[39m \u001b[43m  \u001b[49m\u001b[43mstatus\u001b[49m\u001b[43m=\u001b[49m\u001b[43mstatus\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    298\u001b[39m \u001b[43m  \u001b[49m\u001b[43mother\u001b[49m\u001b[43m=\u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    299\u001b[39m \u001b[43m  \u001b[49m\u001b[43mexclude\u001b[49m\u001b[43m=\u001b[49m\u001b[43mexclude\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    300\u001b[39m \u001b[43m  \u001b[49m\u001b[43minclude\u001b[49m\u001b[43m=\u001b[49m\u001b[43minclude\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    301\u001b[39m \u001b[43m  \u001b[49m\u001b[43mchrom_pat\u001b[49m\u001b[43m=\u001b[49m\u001b[43mchrom_pat\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    302\u001b[39m \u001b[43m  \u001b[49m\u001b[43msnpid_pat\u001b[49m\u001b[43m=\u001b[49m\u001b[43msnpid_pat\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    303\u001b[39m \u001b[43m  \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m=\u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    304\u001b[39m \u001b[43m  \u001b[49m\u001b[43mlog\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mlog\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[32m    305\u001b[39m \u001b[43m  \u001b[49m\u001b[43mreadargs\u001b[49m\u001b[43m=\u001b[49m\u001b[43mreadargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    306\u001b[39m \u001b[43m  \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwreadargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    308\u001b[39m \u001b[38;5;66;03m# meta information\u001b[39;00m\n\u001b[32m    309\u001b[39m \u001b[38;5;28mself\u001b[39m.meta = _init_meta() \n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/io/io_preformat_input.py:620\u001b[39m, in \u001b[36m_preformat\u001b[39m\u001b[34m(sumstats, fmt, tab_fmt, snpid, rsid, chrom, pos, ea, nea, ref, alt, eaf, neaf, maf, n, beta, se, chisq, z, f, t, p, q, mlog10p, test, info, OR, OR_95L, OR_95U, beta_95L, beta_95U, HR, HR_95L, HR_95U, i2, snpr2, phet, dof, ncase, ncontrol, neff, direction, status, study, trait, build, other, exclude, include, chrom_pat, snpid_pat, verbose, log, readargs, **kwreadargs)\u001b[39m\n\u001b[32m    617\u001b[39m dtype_dictionary = {}\n\u001b[32m    619\u001b[39m \u001b[38;5;66;03m# Step 4: Check path and header, get available columns\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m620\u001b[39m inpath, inpath_chr_list, inpath_chr_num_list, format_cols, raw_cols, usecols, dtype_dictionary = \u001b[43m_check_path_and_header\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    621\u001b[39m \u001b[43m    \u001b[49m\u001b[43msumstats\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfmt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmeta_data\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mreadargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43musecols\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype_dictionary\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrename_dictionary\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlog\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\n\u001b[32m    622\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    624\u001b[39m \u001b[38;5;66;03m# Step 5: Build column mappings from user parameters\u001b[39;00m\n\u001b[32m    625\u001b[39m column_params = {\n\u001b[32m    626\u001b[39m     \u001b[33m'\u001b[39m\u001b[33msnpid\u001b[39m\u001b[33m'\u001b[39m: snpid, \u001b[33m'\u001b[39m\u001b[33mrsid\u001b[39m\u001b[33m'\u001b[39m: rsid, \u001b[33m'\u001b[39m\u001b[33mchrom\u001b[39m\u001b[33m'\u001b[39m: chrom, \u001b[33m'\u001b[39m\u001b[33mpos\u001b[39m\u001b[33m'\u001b[39m: pos, \u001b[33m'\u001b[39m\u001b[33mea\u001b[39m\u001b[33m'\u001b[39m: ea, \u001b[33m'\u001b[39m\u001b[33mnea\u001b[39m\u001b[33m'\u001b[39m: nea,\n\u001b[32m    627\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mref\u001b[39m\u001b[33m'\u001b[39m: ref, \u001b[33m'\u001b[39m\u001b[33malt\u001b[39m\u001b[33m'\u001b[39m: alt, \u001b[33m'\u001b[39m\u001b[33meaf\u001b[39m\u001b[33m'\u001b[39m: eaf, \u001b[33m'\u001b[39m\u001b[33mneaf\u001b[39m\u001b[33m'\u001b[39m: neaf, \u001b[33m'\u001b[39m\u001b[33mmaf\u001b[39m\u001b[33m'\u001b[39m: maf, \u001b[33m'\u001b[39m\u001b[33mn\u001b[39m\u001b[33m'\u001b[39m: n,\n\u001b[32m   (...)\u001b[39m\u001b[32m    633\u001b[39m     \u001b[33m'\u001b[39m\u001b[33mdof\u001b[39m\u001b[33m'\u001b[39m: dof, \u001b[33m'\u001b[39m\u001b[33mdirection\u001b[39m\u001b[33m'\u001b[39m: direction, \u001b[33m'\u001b[39m\u001b[33mstatus\u001b[39m\u001b[33m'\u001b[39m: status, \u001b[33m'\u001b[39m\u001b[33mother\u001b[39m\u001b[33m'\u001b[39m: other\n\u001b[32m    634\u001b[39m }\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/io/io_preformat_input.py:904\u001b[39m, in \u001b[36m_check_path_and_header\u001b[39m\u001b[34m(sumstats, fmt, meta_data, readargs, usecols, dtype_dictionary, rename_dictionary, log, verbose)\u001b[39m\n\u001b[32m    901\u001b[39m inpath = sumstats\n\u001b[32m    903\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m904\u001b[39m     format_cols, raw_cols, inpath_chr_list, inpath_chr_num_list = \u001b[43m_process_inpath_and_load_header\u001b[49m\u001b[43m(\u001b[49m\u001b[43minpath\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfmt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmeta_data\u001b[49m\u001b[43m,\u001b[49m\u001b[43m  \u001b[49m\u001b[43mreadargs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlog\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    906\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mFileNotFoundError\u001b[39;00m, \u001b[38;5;167;01mIndexError\u001b[39;00m):\n\u001b[32m    907\u001b[39m     log.warning(\u001b[33m\"\u001b[39m\u001b[33mLoading \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[33m failed...Tesing if compressed/uncompressed...\u001b[39m\u001b[33m\"\u001b[39m.format(inpath),verbose=verbose)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/io/io_preformat_input.py:986\u001b[39m, in \u001b[36m_process_inpath_and_load_header\u001b[39m\u001b[34m(inpath, fmt, meta_data, readargs, log, verbose)\u001b[39m\n\u001b[32m    982\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    983\u001b[39m \u001b[38;5;66;03m##### loading data from tabular file#################################################\u001b[39;00m\n\u001b[32m    984\u001b[39m \u001b[38;5;66;03m#if file not found, FileNotFoundError\u001b[39;00m\n\u001b[32m    985\u001b[39m     readargs_header = _get_readargs_header(inpath = inpath, readargs = readargs)\n\u001b[32m--> \u001b[39m\u001b[32m986\u001b[39m     row_one = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_table\u001b[49m\u001b[43m(\u001b[49m\u001b[43minpath\u001b[49m\u001b[43m,\u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mreadargs_header\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    987\u001b[39m     raw_cols = row_one.columns\n\u001b[32m    989\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m fmt==\u001b[33m\"\u001b[39m\u001b[33mvcf\u001b[39m\u001b[33m\"\u001b[39m:\n\u001b[32m    990\u001b[39m     \u001b[38;5;66;03m# expanded\u001b[39;00m\n",
      "\u001b[31mTypeError\u001b[39m: read_table() got an unexpected keyword argument 'vcfid'"
     ]
    }
   ],
   "source": [
    "# Load your UKBB data\n",
    "mysumstats = gl.Sumstats(\"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "                         vcfid=\"variant\",  # Parses into SNPID, CHR, POS, EA, NEA\n",
    "                         eaf=\"minor_AF\",\n",
    "                         beta=\"beta\",\n",
    "                         se=\"se\",\n",
    "                         p=\"pval\",\n",
    "                         n=\"n_complete_samples\",\n",
    "                         other=[\"low_confidence_variant\"],  # Track this for filtering\n",
    "                        #  nrows=5000,  # For demo; remove for full file\n",
    "                         verbose=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7e4765dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 11:35:18 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/11 11:35:18 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/11 11:35:18 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/11 11:35:18 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 11:35:48  -Reading columns          : minor_allele,minor_AF,variant,beta,se,pval\n",
      "2026/01/11 11:35:48  -Renaming columns to      : EA,EAF,SNPID,BETA,SE,P\n",
      "2026/01/11 11:35:48  -Current Dataframe shape : 13791467  x  6\n",
      "2026/01/11 11:35:48  -Initiating a status column: STATUS ...\n",
      "2026/01/11 11:35:48 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/11 11:35:50  -NEAF is specified...\n",
      "2026/01/11 11:35:50  -Checking if 0<= NEAF <=1 ...\n",
      "2026/01/11 11:35:52  -Converted NEAF to EAF.\n",
      "2026/01/11 11:35:52 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/11 11:35:52  -Reordering columns to    : SNPID,EA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/11 11:35:52 Finished reordering the columns.\n",
      "2026/01/11 11:35:52  -Column  : SNPID  EA       STATUS EAF     BETA    SE      P      \n",
      "2026/01/11 11:35:52  -DType   : object category int64  float64 float64 float64 float64\n",
      "2026/01/11 11:35:52  -Verified: T      T        T      T       T       T       T      \n",
      "2026/01/11 11:35:53  -Current Dataframe memory usage: 790.42 MB\n",
      "2026/01/11 11:35:53 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "# Load the dataset using the confirmed headers\n",
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "    snpid=\"variant\",           \n",
    "    ea=\"minor_allele\",         \n",
    "    neaf=\"minor_AF\",           \n",
    "    beta=\"beta\",\n",
    "    se=\"se\",\n",
    "    p=\"pval\",\n",
    "    verbose=True,\n",
    "    readargs={\n",
    "        \"compression\": \"gzip\",\n",
    "        \"sep\": \"\\t\"\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "65dd817f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>EA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>894.616000</td>\n",
       "      <td>1204.870000</td>\n",
       "      <td>0.457786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID EA   STATUS       EAF        BETA           SE  \\\n",
       "0               1:15791:C:T  T  9999999  1.000000  894.616000  1204.870000   \n",
       "1               1:69487:G:A  A  9999999  0.999994   -2.715450     2.360060   \n",
       "2               1:69569:T:C  C  9999999  0.999812   -0.484284     0.423462   \n",
       "3              1:139853:C:T  T  9999999  0.999994   -2.703560     2.360130   \n",
       "4             1:692794:CA:C  C  9999999  0.889410   -0.016436     0.019585   \n",
       "...                     ... ..      ...       ...         ...          ...   \n",
       "13791462    X:154929412:C:T  T  9999999  0.754527   -0.016260     0.010723   \n",
       "13791463   X:154929637:CT:C  C  9999999  0.770274   -0.027098     0.011190   \n",
       "13791464  X:154929952:CAA:C  C  9999999  0.760570   -0.020494     0.011278   \n",
       "13791465    X:154930230:A:G  G  9999999  0.754113   -0.016347     0.010721   \n",
       "13791466    X:154930487:T:A  A  9999999  0.980441    0.007107     0.036783   \n",
       "\n",
       "                 P  \n",
       "0         0.457786  \n",
       "1         0.249902  \n",
       "2         0.252778  \n",
       "3         0.251997  \n",
       "4         0.401342  \n",
       "...            ...  \n",
       "13791462  0.129427  \n",
       "13791463  0.015456  \n",
       "13791464  0.069202  \n",
       "13791465  0.127334  \n",
       "13791466  0.846790  \n",
       "\n",
       "[13791467 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5189394e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 11:36:16 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/11 11:36:16  -Current Dataframe shape : 13791467 x 7 ; Memory usage: 790.42 MB\n",
      "2026/01/11 11:36:16  -Checking SNPID data type...\n",
      "2026/01/11 11:36:16  -Converted datatype for SNPID: object -> string\n",
      "2026/01/11 11:36:17  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/11 11:36:18  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/11 11:36:28 Finished checking SNPID/rsID.\n",
      "2026/01/11 11:36:28 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/11 11:36:28 #WARNING! Necessary columns for .fix_chr() were not detected:CHR\n",
      "2026/01/11 11:36:28 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/11 11:36:28 #WARNING! Necessary columns for .fix_pos() were not detected:POS\n",
      "2026/01/11 11:36:28 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/11 11:36:28 #WARNING! Necessary columns for .fix_allele() were not detected:NEA\n",
      "2026/01/11 11:36:28 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/11 11:36:28  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/11 11:36:29  -Checking if any columns are empty...\n",
      "2026/01/11 11:36:30  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/11 11:36:30   -Examples of invalid variants(SNPID): 1:776572:C:T,1:819974:G:C,1:871173:C:T,1:871198:C:T,1:874451:C:T ...\n",
      "2026/01/11 11:36:30   -Examples of invalid values (P): NA,NA,NA,NA,NA ...\n",
      "2026/01/11 11:36:31  -Checking if -1e-07 < P < 1.0000001 : removed 1676 variants.\n",
      "2026/01/11 11:36:32   -Examples of invalid variants(SNPID): 1:15791:C:T,1:877608:C:G,1:897032:A:G,1:899336:G:A,1:907676:A:T ...\n",
      "2026/01/11 11:36:32   -Examples of invalid values (BETA): 894.616,279.71,-108.064,270.068,116.14 ...\n",
      "2026/01/11 11:36:32  -Checking if -100.0000001 < BETA < 100.0000001 : removed 9906 variants.\n",
      "2026/01/11 11:36:33  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/11 11:36:33  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/11 11:36:33  -Removed 11582 variants with bad statistics in total.\n",
      "2026/01/11 11:36:33 Finished sanity check for statistics.\n",
      "2026/01/11 11:36:33 Start to check data consistency across columns ...(v4.0.4)\n",
      "2026/01/11 11:36:33  -Current Dataframe shape : 13779885 x 7 ; Memory usage: 750.33 MB\n",
      "2026/01/11 11:36:33  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "2026/01/11 11:36:33  -Checking if BETA/SE-derived-P is consistent with P...\n",
      "2026/01/11 11:36:35   -Variants with inconsistent values were not detected.\n",
      "2026/01/11 11:36:35   -Note: Minor differences are typically due to rounding.\n",
      "2026/01/11 11:36:35    -If the max difference is greater than expected, please check your original sumstats.\n",
      "2026/01/11 11:36:35 Finished checking data consistency across columns.\n",
      "2026/01/11 11:36:35 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/11 11:36:35  -Number of threads/cores to use: 1\n",
      "2026/01/11 11:36:35 #WARNING! Necessary columns for .normalize() were not detected:NEA\n",
      "2026/01/11 11:36:35 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/11 11:36:35 #WARNING! Necessary columns for .sort_coordinate() were not detected:CHR,POS\n",
      "2026/01/11 11:36:35 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/11 11:36:35  -Reordering columns to    : SNPID,EA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/11 11:36:36 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>EA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>9969999</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7740511921b0>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. Basic quality check\n",
    "mysumstats.basic_check()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0644cad2",
   "metadata": {},
   "source": [
    "### Liftover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0c9fb290",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 11:10:13 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/11 11:10:13 #WARNING! Necessary columns for .liftover() were not detected:CHR,POS\n"
     ]
    }
   ],
   "source": [
    "# 3. Perform Liftover from HG19 (Build 37) to HG38\n",
    "# UKB v3 imputed data is natively in GRCh37/hg19\n",
    "mysumstats.liftover(from_build=\"19\", to_build=\"38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "344a017f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Manually split the SNPID string into separate columns\n",
    "# This splits \"1:69487:G:A\" into 4 columns\n",
    "split_cols = mysumstats.data['SNPID'].str.split(':', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Assign the split values to the expected column names\n",
    "mysumstats.data['CHR'] = split_cols[0]\n",
    "mysumstats.data['POS'] = split_cols[1].astype(int)  # Liftover needs POS as an integer\n",
    "mysumstats.data['NEA'] = split_cols[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e4b96b95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Actual columns in dataframe: Index(['SNPID', 'EA', 'STATUS', 'EAF', 'BETA', 'SE', 'P', 'CHR', 'POS', 'NEA'], dtype='object')\n",
      "           SNPID CHR     POS\n",
      "1    1:69487:G:A   1   69487\n",
      "2    1:69569:T:C   1   69569\n",
      "3   1:139853:C:T   1  139853\n",
      "4  1:692794:CA:C   1  692794\n",
      "5   1:693731:A:G   1  693731\n"
     ]
    }
   ],
   "source": [
    "# 3. Verify the columns are actually there now\n",
    "print(\"Actual columns in dataframe:\", mysumstats.data.columns)\n",
    "print(mysumstats.data[['SNPID', 'CHR', 'POS']].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "49912658",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 11:38:22 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/11 11:38:22  -Current Dataframe shape : 13779885 x 10 ; Memory usage: 1065.73 MB\n",
      "2026/01/11 11:38:22 #WARNING! Data types were not fixed for : CHR \n",
      "2026/01/11 11:38:22 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/11 11:38:22  -Checking CHR data type...\n",
      "2026/01/11 11:38:23  -Variants with standardized chromosome notation: 13352828\n",
      "2026/01/11 11:38:23  -Variants with fixable chromosome notations: 427057\n",
      "2026/01/11 11:38:23  -No unrecognized chromosome notations...\n",
      "2026/01/11 11:38:24  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/11 11:38:24  -Identified 427057 variants on sex chromosomes...\n",
      "2026/01/11 11:38:24  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/11 11:38:27 Finished fixing chromosome notation (CHR).\n",
      "2026/01/11 11:38:27  -Dtype fixes applied successfully for requested columns.\n",
      "2026/01/11 11:38:27  -Using built-in chain file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/data/chains/hg19ToHg38.over.chain.gz\n",
      "2026/01/11 11:38:27  -No variants with status code xxx0xxx to liftover\n",
      "2026/01/11 11:38:28 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "# 4. Now run Liftover\n",
    "# Now that 'CHR' and 'POS' are physically in the index, liftover will find them.\n",
    "mysumstats.liftover(from_build=\"19\", to_build=\"38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9fafeec2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      SNPID EA   STATUS       EAF      BETA        SE  \\\n",
      "1               1:69487:G:A  A  9965999  0.999994 -2.715450  2.360060   \n",
      "2               1:69569:T:C  C  9965999  0.999812 -0.484284  0.423462   \n",
      "3              1:139853:C:T  T  9965999  0.999994 -2.703560  2.360130   \n",
      "4             1:692794:CA:C  C  9965999  0.889410 -0.016436  0.019585   \n",
      "5              1:693731:A:G  G  9965999  0.884233 -0.004255  0.018507   \n",
      "...                     ... ..      ...       ...       ...       ...   \n",
      "13791462    X:154929412:C:T  T  9965999  0.754527 -0.016260  0.010723   \n",
      "13791463   X:154929637:CT:C  C  9965999  0.770274 -0.027098  0.011190   \n",
      "13791464  X:154929952:CAA:C  C  9965999  0.760570 -0.020494  0.011278   \n",
      "13791465    X:154930230:A:G  G  9965999  0.754113 -0.016347  0.010721   \n",
      "13791466    X:154930487:T:A  A  9965999  0.980441  0.007107  0.036783   \n",
      "\n",
      "                 P  CHR        POS  NEA  \n",
      "1         0.249902    1      69487    G  \n",
      "2         0.252778    1      69569    T  \n",
      "3         0.251997    1     139853    C  \n",
      "4         0.401342    1     692794   CA  \n",
      "5         0.818155    1     693731    A  \n",
      "...            ...  ...        ...  ...  \n",
      "13791462  0.129427   23  154929412    C  \n",
      "13791463  0.015456   23  154929637   CT  \n",
      "13791464  0.069202   23  154929952  CAA  \n",
      "13791465  0.127334   23  154930230    A  \n",
      "13791466  0.846790   23  154930487    T  \n",
      "\n",
      "[13779885 rows x 10 columns]\n"
     ]
    }
   ],
   "source": [
    "print(mysumstats.data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "75db8cf3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/11 11:39:05 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/11 11:39:05  -Using provided chain file: ./grch38-chm13v2.chain\n",
      "2026/01/11 11:39:05  -No variants with status code xxx0xxx to liftover\n",
      "2026/01/11 11:39:05 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "# Liftover from hg38 to CHM13\n",
    "mysumstats.liftover(from_build=\"38\", to_build=\"13\", chain_path=\"./grch38-chm13v2.chain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b87aab5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['variant', 'minor_allele', 'minor_AF', 'low_confidence_variant', 'n_complete_samples', 'AC', 'ytx', 'beta', 'se', 'tstat', 'pval']\n"
     ]
    }
   ],
   "source": [
    "# import gzip\n",
    "# with gzip.open(\"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\", 'rt') as f:\n",
    "#     header = f.readline()\n",
    "#     print(header.split())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e63bd5d9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68e7b371",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ae2b8acc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "5b259464",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "46f99bda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gzipped GWAS summary stats\n",
    "df = pd.read_csv(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\"\n",
    ")\n",
    "\n",
    "# Split variant into CHR, POS, REF, ALT\n",
    "df[['CHR', 'POS', 'REF', 'ALT']] = df['variant'].str.split(\":\", expand=True)\n",
    "df['POS'] = df['POS'].astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f5aad49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['variant', 'minor_allele', 'minor_AF', 'low_confidence_variant',\n",
       "       'n_complete_samples', 'AC', 'ytx', 'beta', 'se', 'tstat', 'pval', 'CHR',\n",
       "       'POS', 'REF', 'ALT'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6dd3e7fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Create a copy and rename columns to GWASLab standard names\n",
    "df_standard = df.rename(columns={\n",
    "    'variant': 'SNPID',\n",
    "    'CHR': 'CHR',\n",
    "    'POS': 'POS',\n",
    "    'ALT': 'EA',\n",
    "    'REF': 'NEA',\n",
    "    'minor_AF': 'EAF',  # Map your frequency directly to EAF\n",
    "    'beta': 'BETA',\n",
    "    'se': 'SE',\n",
    "    'pval': 'P'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8314a3f9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 10:36:40 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/10 10:36:40 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/10 10:36:40 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/10 10:36:40 Start to initialize gl.Sumstats from pandas DataFrame ...\n",
      "2026/01/10 10:36:41  -Reading columns          : \n",
      "2026/01/10 10:36:41  -Renaming columns to      : \n",
      "2026/01/10 10:36:41  -Current Dataframe shape : 13791467  x  15\n",
      "2026/01/10 10:36:42  -Initiating a status column: STATUS ...\n",
      "2026/01/10 10:36:42 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/10 10:36:44 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 10:36:44  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 10:36:45 Finished reordering the columns.\n",
      "2026/01/10 10:36:45  -Trying to convert datatype for CHR: object -> Int64...Failed...\n",
      "2026/01/10 10:36:48  -Column  : SNPID  CHR    POS   EA       NEA      STATUS EAF     BETA    SE      P       minor_allele low_confidence_variant n_complete_samples AC      ytx     tstat  \n",
      "2026/01/10 10:36:48  -DType   : object object int64 category category int64  float64 float64 float64 float64 object       bool                   int64              float64 float64 float64\n",
      "2026/01/10 10:36:48  -Verified: T      F      T     T        T        T      T       T       T       T       NA           NA                     NA                 NA      NA      NA     \n",
      "2026/01/10 10:36:48 #WARNING! Columns with possibly incompatible dtypes: CHR\n",
      "2026/01/10 10:36:48 #WARNING! Consider using Sumstats.fix_chr() to fix CHR dtype\n",
      "2026/01/10 10:36:48  -Current Dataframe memory usage: 1489.96 MB\n",
      "2026/01/10 10:36:49 Finished loading data successfully!\n",
      "2026/01/10 10:36:50 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/10 10:36:50  -Current Dataframe shape : 13791467 x 16 ; Memory usage: 1489.96 MB\n",
      "2026/01/10 10:36:50  -Checking SNPID data type...\n",
      "2026/01/10 10:36:50  -Converted datatype for SNPID: object -> string\n",
      "2026/01/10 10:36:50  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/10 10:36:51  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/10 10:37:00 Finished checking SNPID/rsID.\n",
      "2026/01/10 10:37:00 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 10:37:00  -Checking CHR data type...\n",
      "2026/01/10 10:37:02  -Variants with standardized chromosome notation: 13364303\n",
      "2026/01/10 10:37:03  -Variants with fixable chromosome notations: 427164\n",
      "2026/01/10 10:37:03  -No unrecognized chromosome notations...\n",
      "2026/01/10 10:37:03  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/10 10:37:03  -Identified 427164 variants on sex chromosomes...\n",
      "2026/01/10 10:37:03  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/10 10:37:06 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 10:37:06 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 10:37:06  -Trying to convert datatype for POS: int64 -> Int64...\n",
      "2026/01/10 10:37:06  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 10:37:06  -No outlier variants were removed.\n",
      "2026/01/10 10:37:08 Finished fixing basepair positions (POS).\n",
      "2026/01/10 10:37:08 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/10 10:37:08  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/10 10:37:13  -All variants have valid alleles (EA and NEA are A/C/T/G, not NA, and different).\n",
      "2026/01/10 10:37:26 Finished fixing alleles (EA and NEA).\n",
      "2026/01/10 10:37:26 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/10 10:37:26  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/10 10:37:27  -Checking if any columns are empty...\n",
      "2026/01/10 10:37:28  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/10 10:37:28   -Examples of invalid variants(SNPID): 1:776572:C:T,1:819974:G:C,1:871173:C:T,1:871198:C:T,1:874451:C:T ...\n",
      "2026/01/10 10:37:28   -Examples of invalid values (P): NA,NA,NA,NA,NA ...\n",
      "2026/01/10 10:37:29  -Checking if -1e-07 < P < 1.0000001 : removed 1676 variants.\n",
      "2026/01/10 10:37:29   -Examples of invalid variants(SNPID): 1:15791:C:T,1:877608:C:G,1:897032:A:G,1:899336:G:A,1:907676:A:T ...\n",
      "2026/01/10 10:37:29   -Examples of invalid values (BETA): 894.616,279.71,-108.064,270.068,116.14 ...\n",
      "2026/01/10 10:37:31  -Checking if -100.0000001 < BETA < 100.0000001 : removed 9906 variants.\n",
      "2026/01/10 10:37:32  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/10 10:37:32  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/10 10:37:32  -Removed 11582 variants with bad statistics in total.\n",
      "2026/01/10 10:37:32 Finished sanity check for statistics.\n",
      "2026/01/10 10:37:32 Start to check data consistency across columns ...(v4.0.4)\n",
      "2026/01/10 10:37:32  -Current Dataframe shape : 13779885 x 16 ; Memory usage: 1582.28 MB\n",
      "2026/01/10 10:37:32  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "2026/01/10 10:37:32  -Checking if BETA/SE-derived-P is consistent with P...\n",
      "2026/01/10 10:37:34   -Variants with inconsistent values were not detected.\n",
      "2026/01/10 10:37:34   -Note: Minor differences are typically due to rounding.\n",
      "2026/01/10 10:37:34    -If the max difference is greater than expected, please check your original sumstats.\n",
      "2026/01/10 10:37:34 Finished checking data consistency across columns.\n",
      "2026/01/10 10:37:34 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/10 10:37:34  -Number of threads/cores to use: 1\n",
      "2026/01/10 10:37:34  -Number of variants to check:728\n",
      "2026/01/10 10:37:34  -Chunk size:3000000\n",
      "2026/01/10 10:37:34  -Processing in chunks:0 \n",
      "2026/01/10 10:37:36 Finished normalizing indels.\n",
      "2026/01/10 10:37:36 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/10 10:37:49 Finished sorting coordinates.\n",
      "2026/01/10 10:37:49 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 10:37:49  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 10:37:50 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.15294</td>\n",
       "      <td>103.515</td>\n",
       "      <td>-1.150590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.27800</td>\n",
       "      <td>3644.560</td>\n",
       "      <td>-1.143630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.09020</td>\n",
       "      <td>101.840</td>\n",
       "      <td>-1.145510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10000</td>\n",
       "      <td>2179000.000</td>\n",
       "      <td>-0.839228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00000</td>\n",
       "      <td>2281760.000</td>\n",
       "      <td>-0.229918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00000</td>\n",
       "      <td>4837480.000</td>\n",
       "      <td>-1.516370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00000</td>\n",
       "      <td>4525550.000</td>\n",
       "      <td>-2.421510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00000</td>\n",
       "      <td>4717910.000</td>\n",
       "      <td>-1.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00000</td>\n",
       "      <td>4845680.000</td>\n",
       "      <td>-1.524700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80000</td>\n",
       "      <td>385739.000</td>\n",
       "      <td>0.193216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7384af2f4b00>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. Load the DataFrame without specifying column mappings \n",
    "# (GWASLab will recognize SNPID, CHR, POS, EA, NEA, EAF, BETA, SE, P automatically)\n",
    "mysumstats = gl.Sumstats(df_standard, \n",
    "                        #  nrows=5000, \n",
    "                         verbose=True)\n",
    "\n",
    "mysumstats.basic_check(verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "da6396b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.15294</td>\n",
       "      <td>103.515</td>\n",
       "      <td>-1.150590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.27800</td>\n",
       "      <td>3644.560</td>\n",
       "      <td>-1.143630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.09020</td>\n",
       "      <td>101.840</td>\n",
       "      <td>-1.145510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10000</td>\n",
       "      <td>2179000.000</td>\n",
       "      <td>-0.839228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00000</td>\n",
       "      <td>2281760.000</td>\n",
       "      <td>-0.229918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00000</td>\n",
       "      <td>4837480.000</td>\n",
       "      <td>-1.516370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00000</td>\n",
       "      <td>4525550.000</td>\n",
       "      <td>-2.421510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00000</td>\n",
       "      <td>4717910.000</td>\n",
       "      <td>-1.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00000</td>\n",
       "      <td>4845680.000</td>\n",
       "      <td>-1.524700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80000</td>\n",
       "      <td>385739.000</td>\n",
       "      <td>0.193216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS       EAF  \\\n",
       "0               1:69487:G:A    1      69487  A    G  9960099  0.000006   \n",
       "1               1:69569:T:C    1      69569  C    T  9960099  0.000188   \n",
       "2              1:139853:C:T    1     139853  T    C  9960099  0.000006   \n",
       "3             1:692794:CA:C    1     692794  C   CA  9960399  0.110590   \n",
       "4              1:693731:A:G    1     693731  G    A  9960099  0.115767   \n",
       "...                     ...  ...        ... ..  ...      ...       ...   \n",
       "13779880    X:154929412:C:T   23  154929412  T    C  9960099  0.245473   \n",
       "13779881   X:154929637:CT:C   23  154929637  C   CT  9960399  0.229726   \n",
       "13779882  X:154929952:CAA:C   23  154929952  C  CAA  9960399  0.239430   \n",
       "13779883    X:154930230:A:G   23  154930230  G    A  9960099  0.245887   \n",
       "13779884    X:154930487:T:A   23  154930487  A    T  9960099  0.019559   \n",
       "\n",
       "              BETA        SE         P minor_allele  low_confidence_variant  \\\n",
       "0        -2.715450  2.360060  0.249902            A                    True   \n",
       "1        -0.484284  0.423462  0.252778            C                    True   \n",
       "2        -2.703560  2.360130  0.251997            T                    True   \n",
       "3        -0.016436  0.019585  0.401342            C                   False   \n",
       "4        -0.004255  0.018507  0.818155            G                   False   \n",
       "...            ...       ...       ...          ...                     ...   \n",
       "13779880 -0.016260  0.010723  0.129427            T                   False   \n",
       "13779881 -0.027098  0.011190  0.015456            C                   False   \n",
       "13779882 -0.020494  0.011278  0.069202            C                   False   \n",
       "13779883 -0.016347  0.010721  0.127334            G                   False   \n",
       "13779884  0.007107  0.036783  0.846790            A                   False   \n",
       "\n",
       "          n_complete_samples            AC          ytx     tstat  \n",
       "0                     359983       4.15294      103.515 -1.150590  \n",
       "1                     359983     135.27800     3644.560 -1.143630  \n",
       "2                     359983       4.09020      101.840 -1.145510  \n",
       "3                     359983   79621.10000  2179000.000 -0.839228  \n",
       "4                     359983   83348.00000  2281760.000 -0.229918  \n",
       "...                      ...           ...          ...       ...  \n",
       "13779880              359983  176732.00000  4837480.000 -1.516370  \n",
       "13779881              359983  165395.00000  4525550.000 -2.421510  \n",
       "13779882              359983  172381.00000  4717910.000 -1.817100  \n",
       "13779883              359983  177030.00000  4845680.000 -1.524700  \n",
       "13779884              359983   14081.80000   385739.000  0.193216  \n",
       "\n",
       "[13779885 rows x 16 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6ecd2fa",
   "metadata": {},
   "source": [
    "### Liftover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76a18065",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>SNPID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>894.616000</td>\n",
       "      <td>1204.870000</td>\n",
       "      <td>0.457786</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        NEA   STATUS       EAF        BETA           SE  \\\n",
       "0               1:15791:C:T  9999999  1.000000  894.616000  1204.870000   \n",
       "1               1:69487:G:A  9999999  0.999994   -2.715450     2.360060   \n",
       "2               1:69569:T:C  9999999  0.999812   -0.484284     0.423462   \n",
       "3              1:139853:C:T  9999999  0.999994   -2.703560     2.360130   \n",
       "4             1:692794:CA:C  9999999  0.889410   -0.016436     0.019585   \n",
       "...                     ...      ...       ...         ...          ...   \n",
       "13791462    X:154929412:C:T  9999999  0.754527   -0.016260     0.010723   \n",
       "13791463   X:154929637:CT:C  9999999  0.770274   -0.027098     0.011190   \n",
       "13791464  X:154929952:CAA:C  9999999  0.760570   -0.020494     0.011278   \n",
       "13791465    X:154930230:A:G  9999999  0.754113   -0.016347     0.010721   \n",
       "13791466    X:154930487:T:A  9999999  0.980441    0.007107     0.036783   \n",
       "\n",
       "                 P SNPID  \n",
       "0         0.457786  <NA>  \n",
       "1         0.249902  <NA>  \n",
       "2         0.252778  <NA>  \n",
       "3         0.251997  <NA>  \n",
       "4         0.401342  <NA>  \n",
       "...            ...   ...  \n",
       "13791462  0.129427  <NA>  \n",
       "13791463  0.015456  <NA>  \n",
       "13791464  0.069202  <NA>  \n",
       "13791465  0.127334  <NA>  \n",
       "13791466  0.846790  <NA>  \n",
       "\n",
       "[13791467 rows x 7 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.liftover(from_build=\"19\", to_build=\"38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5f2d31db",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.15294</td>\n",
       "      <td>103.515</td>\n",
       "      <td>-1.150590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.27800</td>\n",
       "      <td>3644.560</td>\n",
       "      <td>-1.143630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.09020</td>\n",
       "      <td>101.840</td>\n",
       "      <td>-1.145510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10000</td>\n",
       "      <td>2179000.000</td>\n",
       "      <td>-0.839228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00000</td>\n",
       "      <td>2281760.000</td>\n",
       "      <td>-0.229918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00000</td>\n",
       "      <td>4837480.000</td>\n",
       "      <td>-1.516370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00000</td>\n",
       "      <td>4525550.000</td>\n",
       "      <td>-2.421510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00000</td>\n",
       "      <td>4717910.000</td>\n",
       "      <td>-1.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00000</td>\n",
       "      <td>4845680.000</td>\n",
       "      <td>-1.524700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80000</td>\n",
       "      <td>385739.000</td>\n",
       "      <td>0.193216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS       EAF  \\\n",
       "0               1:69487:G:A    1      69487  A    G  9960099  0.000006   \n",
       "1               1:69569:T:C    1      69569  C    T  9960099  0.000188   \n",
       "2              1:139853:C:T    1     139853  T    C  9960099  0.000006   \n",
       "3             1:692794:CA:C    1     692794  C   CA  9960399  0.110590   \n",
       "4              1:693731:A:G    1     693731  G    A  9960099  0.115767   \n",
       "...                     ...  ...        ... ..  ...      ...       ...   \n",
       "13779880    X:154929412:C:T   23  154929412  T    C  9960099  0.245473   \n",
       "13779881   X:154929637:CT:C   23  154929637  C   CT  9960399  0.229726   \n",
       "13779882  X:154929952:CAA:C   23  154929952  C  CAA  9960399  0.239430   \n",
       "13779883    X:154930230:A:G   23  154930230  G    A  9960099  0.245887   \n",
       "13779884    X:154930487:T:A   23  154930487  A    T  9960099  0.019559   \n",
       "\n",
       "              BETA        SE         P minor_allele  low_confidence_variant  \\\n",
       "0        -2.715450  2.360060  0.249902            A                    True   \n",
       "1        -0.484284  0.423462  0.252778            C                    True   \n",
       "2        -2.703560  2.360130  0.251997            T                    True   \n",
       "3        -0.016436  0.019585  0.401342            C                   False   \n",
       "4        -0.004255  0.018507  0.818155            G                   False   \n",
       "...            ...       ...       ...          ...                     ...   \n",
       "13779880 -0.016260  0.010723  0.129427            T                   False   \n",
       "13779881 -0.027098  0.011190  0.015456            C                   False   \n",
       "13779882 -0.020494  0.011278  0.069202            C                   False   \n",
       "13779883 -0.016347  0.010721  0.127334            G                   False   \n",
       "13779884  0.007107  0.036783  0.846790            A                   False   \n",
       "\n",
       "          n_complete_samples            AC          ytx     tstat  \n",
       "0                     359983       4.15294      103.515 -1.150590  \n",
       "1                     359983     135.27800     3644.560 -1.143630  \n",
       "2                     359983       4.09020      101.840 -1.145510  \n",
       "3                     359983   79621.10000  2179000.000 -0.839228  \n",
       "4                     359983   83348.00000  2281760.000 -0.229918  \n",
       "...                      ...           ...          ...       ...  \n",
       "13779880              359983  176732.00000  4837480.000 -1.516370  \n",
       "13779881              359983  165395.00000  4525550.000 -2.421510  \n",
       "13779882              359983  172381.00000  4717910.000 -1.817100  \n",
       "13779883              359983  177030.00000  4845680.000 -1.524700  \n",
       "13779884              359983   14081.80000   385739.000  0.193216  \n",
       "\n",
       "[13779885 rows x 16 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "498eed59",
   "metadata": {},
   "source": [
    "### Liftover using user-provided chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d31335e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2026-01-10 10:39:53--  https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/CHM13/assemblies/chain/v1_nflo/grch38-chm13v2.chain\n",
      "Resolving s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)... 3.5.79.179, 52.218.246.64, 52.92.240.144, ...\n",
      "Connecting to s3-us-west-2.amazonaws.com (s3-us-west-2.amazonaws.com)|3.5.79.179|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 6288201 (6.0M) [binary/octet-stream]\n",
      "Saving to: ‘grch38-chm13v2.chain’\n",
      "\n",
      "grch38-chm13v2.chai 100%[===================>]   6.00M  3.74MB/s    in 1.6s    \n",
      "\n",
      "2026-01-10 10:39:55 (3.74 MB/s) - ‘grch38-chm13v2.chain’ saved [6288201/6288201]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! wget https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/CHM13/assemblies/chain/v1_nflo/grch38-chm13v2.chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4a7614f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.15294</td>\n",
       "      <td>103.515</td>\n",
       "      <td>-1.150590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.27800</td>\n",
       "      <td>3644.560</td>\n",
       "      <td>-1.143630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.09020</td>\n",
       "      <td>101.840</td>\n",
       "      <td>-1.145510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10000</td>\n",
       "      <td>2179000.000</td>\n",
       "      <td>-0.839228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00000</td>\n",
       "      <td>2281760.000</td>\n",
       "      <td>-0.229918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00000</td>\n",
       "      <td>4837480.000</td>\n",
       "      <td>-1.516370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00000</td>\n",
       "      <td>4525550.000</td>\n",
       "      <td>-2.421510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00000</td>\n",
       "      <td>4717910.000</td>\n",
       "      <td>-1.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00000</td>\n",
       "      <td>4845680.000</td>\n",
       "      <td>-1.524700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80000</td>\n",
       "      <td>385739.000</td>\n",
       "      <td>0.193216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS       EAF  \\\n",
       "0               1:69487:G:A    1      69487  A    G  9960099  0.000006   \n",
       "1               1:69569:T:C    1      69569  C    T  9960099  0.000188   \n",
       "2              1:139853:C:T    1     139853  T    C  9960099  0.000006   \n",
       "3             1:692794:CA:C    1     692794  C   CA  9960399  0.110590   \n",
       "4              1:693731:A:G    1     693731  G    A  9960099  0.115767   \n",
       "...                     ...  ...        ... ..  ...      ...       ...   \n",
       "13779880    X:154929412:C:T   23  154929412  T    C  9960099  0.245473   \n",
       "13779881   X:154929637:CT:C   23  154929637  C   CT  9960399  0.229726   \n",
       "13779882  X:154929952:CAA:C   23  154929952  C  CAA  9960399  0.239430   \n",
       "13779883    X:154930230:A:G   23  154930230  G    A  9960099  0.245887   \n",
       "13779884    X:154930487:T:A   23  154930487  A    T  9960099  0.019559   \n",
       "\n",
       "              BETA        SE         P minor_allele  low_confidence_variant  \\\n",
       "0        -2.715450  2.360060  0.249902            A                    True   \n",
       "1        -0.484284  0.423462  0.252778            C                    True   \n",
       "2        -2.703560  2.360130  0.251997            T                    True   \n",
       "3        -0.016436  0.019585  0.401342            C                   False   \n",
       "4        -0.004255  0.018507  0.818155            G                   False   \n",
       "...            ...       ...       ...          ...                     ...   \n",
       "13779880 -0.016260  0.010723  0.129427            T                   False   \n",
       "13779881 -0.027098  0.011190  0.015456            C                   False   \n",
       "13779882 -0.020494  0.011278  0.069202            C                   False   \n",
       "13779883 -0.016347  0.010721  0.127334            G                   False   \n",
       "13779884  0.007107  0.036783  0.846790            A                   False   \n",
       "\n",
       "          n_complete_samples            AC          ytx     tstat  \n",
       "0                     359983       4.15294      103.515 -1.150590  \n",
       "1                     359983     135.27800     3644.560 -1.143630  \n",
       "2                     359983       4.09020      101.840 -1.145510  \n",
       "3                     359983   79621.10000  2179000.000 -0.839228  \n",
       "4                     359983   83348.00000  2281760.000 -0.229918  \n",
       "...                      ...           ...          ...       ...  \n",
       "13779880              359983  176732.00000  4837480.000 -1.516370  \n",
       "13779881              359983  165395.00000  4525550.000 -2.421510  \n",
       "13779882              359983  172381.00000  4717910.000 -1.817100  \n",
       "13779883              359983  177030.00000  4845680.000 -1.524700  \n",
       "13779884              359983   14081.80000   385739.000  0.193216  \n",
       "\n",
       "[13779885 rows x 16 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a664ae25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 10:40:08 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/10 10:40:08  -Using provided chain file: ./grch38-chm13v2.chain\n",
      "2026/01/10 10:40:10  -Converting variants with status code xxx0xxx: 13,779,885\n",
      "2026/01/10 10:40:11  -Target build: 13\n",
      "2026/01/10 10:40:11  -Input positions are 1-based\n",
      "2026/01/10 10:40:11  -Output positions will be 1-based\n",
      "2026/01/10 10:41:41  -Mapped: 13640038 variants\n",
      "2026/01/10 10:41:41  -Unmapped: 139847 variants\n",
      "2026/01/10 10:41:41  -Examples of unmapped variants:\n",
      "2026/01/10 10:41:41    SNPID=1:69487:G:A | CHR=1 | POS=69487 | STATUS=9960099\n",
      "2026/01/10 10:41:41    SNPID=1:69569:T:C | CHR=1 | POS=69569 | STATUS=9960099\n",
      "2026/01/10 10:41:41    SNPID=1:139853:C:T | CHR=1 | POS=139853 | STATUS=9960099\n",
      "2026/01/10 10:41:41    SNPID=1:905017:T:C | CHR=1 | POS=905017 | STATUS=9960099\n",
      "2026/01/10 10:41:41    SNPID=1:909221:T:C | CHR=1 | POS=909221 | STATUS=9960099\n",
      "2026/01/10 10:41:49  -Removed 139847 unmapped variants\n",
      "2026/01/10 10:41:49 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 10:41:49  -Current Dataframe shape : 13640038 x 16 ; Memory usage: 1553.27 MB\n",
      "2026/01/10 10:41:49  -Checking CHR data type...\n",
      "2026/01/10 10:41:50  -Variants with standardized chromosome notation: 13217569\n",
      "2026/01/10 10:41:52  -Variants with fixable chromosome notations: 422469\n",
      "2026/01/10 10:41:52  -No unrecognized chromosome notations...\n",
      "2026/01/10 10:41:53  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/10 10:41:53  -Identified 422469 variants on sex chromosomes...\n",
      "2026/01/10 10:41:53  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/10 10:41:56 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 10:41:56 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 10:41:56  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/10 10:41:57  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 10:41:57  -No outlier variants were removed.\n",
      "2026/01/10 10:42:00  -Removed variants with bad positions: 0\n",
      "2026/01/10 10:42:00 Finished fixing basepair positions (POS).\n",
      "2026/01/10 10:42:00 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "mysumstats.liftover(from_build=\"38\", to_build=\"13\",chain_path=\"./grch38-chm13v2.chain\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cd9b0aca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>119016</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>1360399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10</td>\n",
       "      <td>2179000.0</td>\n",
       "      <td>-0.839228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>119953</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1360099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00</td>\n",
       "      <td>2281760.0</td>\n",
       "      <td>-0.229918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>133743</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1360099</td>\n",
       "      <td>0.097255</td>\n",
       "      <td>-0.010428</td>\n",
       "      <td>0.020804</td>\n",
       "      <td>0.616190</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>70020.60</td>\n",
       "      <td>1916410.0</td>\n",
       "      <td>-0.501258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1:717587:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>144100</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1360099</td>\n",
       "      <td>0.015679</td>\n",
       "      <td>0.001174</td>\n",
       "      <td>0.049643</td>\n",
       "      <td>0.981132</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>11288.70</td>\n",
       "      <td>309182.0</td>\n",
       "      <td>0.023649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1:723329:A:T</td>\n",
       "      <td>1</td>\n",
       "      <td>149839</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1360099</td>\n",
       "      <td>0.001732</td>\n",
       "      <td>-0.117102</td>\n",
       "      <td>0.146483</td>\n",
       "      <td>0.424044</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>1247.02</td>\n",
       "      <td>34009.5</td>\n",
       "      <td>-0.799426</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>153165819</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1360099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00</td>\n",
       "      <td>4837480.0</td>\n",
       "      <td>-1.516370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>153166044</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>1360399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00</td>\n",
       "      <td>4525550.0</td>\n",
       "      <td>-2.421510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>153166359</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>1360399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00</td>\n",
       "      <td>4717910.0</td>\n",
       "      <td>-1.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>153166637</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1360099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00</td>\n",
       "      <td>4845680.0</td>\n",
       "      <td>-1.524700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>153166894</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1360099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80</td>\n",
       "      <td>385739.0</td>\n",
       "      <td>0.193216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13640038 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS       EAF  \\\n",
       "3             1:692794:CA:C    1     119016  C   CA  1360399  0.110590   \n",
       "4              1:693731:A:G    1     119953  G    A  1360099  0.115767   \n",
       "5              1:707522:G:C    1     133743  C    G  1360099  0.097255   \n",
       "6              1:717587:G:A    1     144100  A    G  1360099  0.015679   \n",
       "7              1:723329:A:T    1     149839  T    A  1360099  0.001732   \n",
       "...                     ...  ...        ... ..  ...      ...       ...   \n",
       "13779880    X:154929412:C:T   23  153165819  T    C  1360099  0.245473   \n",
       "13779881   X:154929637:CT:C   23  153166044  C   CT  1360399  0.229726   \n",
       "13779882  X:154929952:CAA:C   23  153166359  C  CAA  1360399  0.239430   \n",
       "13779883    X:154930230:A:G   23  153166637  G    A  1360099  0.245887   \n",
       "13779884    X:154930487:T:A   23  153166894  A    T  1360099  0.019559   \n",
       "\n",
       "              BETA        SE         P minor_allele  low_confidence_variant  \\\n",
       "3        -0.016436  0.019585  0.401342            C                   False   \n",
       "4        -0.004255  0.018507  0.818155            G                   False   \n",
       "5        -0.010428  0.020804  0.616190            C                   False   \n",
       "6         0.001174  0.049643  0.981132            A                   False   \n",
       "7        -0.117102  0.146483  0.424044            T                   False   \n",
       "...            ...       ...       ...          ...                     ...   \n",
       "13779880 -0.016260  0.010723  0.129427            T                   False   \n",
       "13779881 -0.027098  0.011190  0.015456            C                   False   \n",
       "13779882 -0.020494  0.011278  0.069202            C                   False   \n",
       "13779883 -0.016347  0.010721  0.127334            G                   False   \n",
       "13779884  0.007107  0.036783  0.846790            A                   False   \n",
       "\n",
       "          n_complete_samples         AC        ytx     tstat  \n",
       "3                     359983   79621.10  2179000.0 -0.839228  \n",
       "4                     359983   83348.00  2281760.0 -0.229918  \n",
       "5                     359983   70020.60  1916410.0 -0.501258  \n",
       "6                     359983   11288.70   309182.0  0.023649  \n",
       "7                     359983    1247.02    34009.5 -0.799426  \n",
       "...                      ...        ...        ...       ...  \n",
       "13779880              359983  176732.00  4837480.0 -1.516370  \n",
       "13779881              359983  165395.00  4525550.0 -2.421510  \n",
       "13779882              359983  172381.00  4717910.0 -1.817100  \n",
       "13779883              359983  177030.00  4845680.0 -1.524700  \n",
       "13779884              359983   14081.80   385739.0  0.193216  \n",
       "\n",
       "[13640038 rows x 16 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e5f867be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save to a gzipped TSV file\n",
    "mysumstats.to_csv(\"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover.tsv.gz\", \n",
    "                  sep=\"\\t\", \n",
    "                  index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
