{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c40ea94b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c8695a92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:09:08 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/14 08:09:08 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/14 08:09:08 Python version: 3.12.3 (main, Jan  8 2026, 11:30:50) [GCC 13.3.0]\n",
      "2026/01/14 08:09:08 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\n",
      "2026/01/14 08:09:33  -Reading columns          : BETA,P,CHR,POS,SNPID,EA,NEA,EAF,SE\n",
      "2026/01/14 08:09:33  -Renaming columns to      : BETA,P,CHR,POS,SNPID,EA,NEA,EAF,SE\n",
      "2026/01/14 08:09:33  -Current Dataframe shape : 13779885  x  9\n",
      "2026/01/14 08:09:34  -Initiating a status column: STATUS ...\n",
      "2026/01/14 08:09:34 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/14 08:09:37 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 08:09:37  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 08:09:37 Finished reordering the columns.\n",
      "2026/01/14 08:09:38  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2026/01/14 08:09:39  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P      \n",
      "2026/01/14 08:09:39  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64\n",
      "2026/01/14 08:09:39  -Verified: T      T     T     T        T        T      T       T       T       T      \n",
      "2026/01/14 08:09:39  -Current Dataframe memory usage: 963.05 MB\n",
      "2026/01/14 08:09:40 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\",\n",
    "    snpid=\"SNPID\",\n",
    "    chrom=\"CHR\",\n",
    "    pos=\"POS\",\n",
    "    ea=\"EA\",            # In your data, it's now EA (was ALT)\n",
    "    nea=\"NEA\",          # In your data, it's now NEA (was REF)\n",
    "    eaf=\"EAF\",          # In your data, it's now EAF (was minor_AF/Frq)\n",
    "    beta=\"BETA\",\n",
    "    se=\"SE\",\n",
    "    p=\"P\",\n",
    "    # Note: 'Dir' and 'N' are not in your column list, \n",
    "    # so we should omit them or point to the correct ones.\n",
    "    # n=\"n_complete_samples\", # Optional: if you want to use sample size\n",
    "    verbose=True,\n",
    "    readargs={'sep': '\\t'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5f255f90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SNPID', 'CHR', 'POS', 'EA', 'NEA', 'STATUS', 'EAF', 'BETA', 'SE', 'P']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "01eb78b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:11:19 Start to download  1kg_dbsnp151_hg38_auto  ...\n",
      "2026/01/14 08:11:19  -Downloading to: /home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz\n",
      "2026/01/14 08:11:19  -File /home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz exists.\n",
      "2026/01/14 08:11:19  - Updating record in config file...\n",
      "2026/01/14 08:11:19 Downloaded  1kg_dbsnp151_hg38_auto  successfully!\n"
     ]
    }
   ],
   "source": [
    "# import gwaslab as gl\n",
    "\n",
    "# GWASLab has a built-in helper to fetch some of these\n",
    "# This will download the auto-chromosome dbSNP reference for hg38 which includes frequencies\n",
    "gl.download_ref(\"1kg_dbsnp151_hg38_auto\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b5e7ca1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:13:52 Start to check if NEA is aligned with reference sequence ...(v4.0.4)\n",
      "2026/01/14 08:13:52  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\n"
     ]
    }
   ],
   "source": [
    "# 1. Path to your local hg38 FASTA\n",
    "ref_fasta = \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\"\n",
    "\n",
    "# 2. Path to your local hg38 VCFs (using the {chrom} wildcard)\n",
    "# GWASLab will replace {chrom} with 1, 2, 3... as it iterates\n",
    "ref_vcf = \"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz\"\n",
    "\n",
    "# 3. Perform Harmonization\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=ref_fasta,      # Checks alleles against the hg38 genome sequence\n",
    "    ref_infer=ref_vcf,      # Checks strand and frequency against the VCFs\n",
    "    ref_alt_freq=\"AF\",      # Standard dbSNP frequency column name\n",
    "    threads=6,              # Optimized for your 6 threads\n",
    "    verbose=False,          # Keeps the console clean\n",
    "    # log=False               # Prevents giant log files for 13M rows\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "20ff1309",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:11:31 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/14 08:11:31  -Current Dataframe shape : 13779885 x 10 ; Memory usage: 963.05 MB\n",
      "2026/01/14 08:11:31  -Checking SNPID data type...\n",
      "2026/01/14 08:11:31  -Converted datatype for SNPID: object -> string\n",
      "2026/01/14 08:11:32  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/14 08:11:33  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 9\u001b[39m\n\u001b[32m      6\u001b[39m ref_vcf = \u001b[33m\"\u001b[39m\u001b[33m/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr\u001b[39m\u001b[38;5;132;01m{chrom}\u001b[39;00m\u001b[33m.parquet\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m      8\u001b[39m \u001b[38;5;66;03m# 3. Perform Harmonization\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m9\u001b[39m \u001b[43mmysumstats\u001b[49m\u001b[43m.\u001b[49m\u001b[43mharmonize\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m     10\u001b[39m \u001b[43m    \u001b[49m\u001b[43mref_seq\u001b[49m\u001b[43m=\u001b[49m\u001b[43mref_fasta\u001b[49m\u001b[43m,\u001b[49m\u001b[43m      \u001b[49m\u001b[38;5;66;43;03m# Used to check if alleles match the human genome\u001b[39;49;00m\n\u001b[32m     11\u001b[39m \u001b[43m    \u001b[49m\u001b[43mref_infer\u001b[49m\u001b[43m=\u001b[49m\u001b[43mref_vcf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m      \u001b[49m\u001b[38;5;66;43;03m# Used to check AF (Allele Frequency) and strand alignment\u001b[39;49;00m\n\u001b[32m     12\u001b[39m \u001b[43m    \u001b[49m\u001b[43mref_alt_freq\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mAF\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m      \u001b[49m\u001b[38;5;66;43;03m# Note: Check your parquet files, it's usually just \"AF\" or \"EAF\"\u001b[39;49;00m\n\u001b[32m     13\u001b[39m \u001b[43m    \u001b[49m\u001b[43mthreads\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m6\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m              \u001b[49m\u001b[38;5;66;43;03m# Using your 6 threads\u001b[39;49;00m\n\u001b[32m     14\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# build=\"38\" is assumed since you lifted over\u001b[39;49;00m\n\u001b[32m     15\u001b[39m \u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/g_Sumstats.py:750\u001b[39m, in \u001b[36mSumstats.harmonize\u001b[39m\u001b[34m(self, basic_check, ref_seq, ref_rsid_tsv, ref_rsid_vcf, ref_infer, ref_alt_freq, ref_maf_threshold, maf_threshold, threads, remove, check_ref_kwargs, remove_dup_kwargs, assign_rsid_kwargs, infer_strand_kwargs, flip_allele_stats_kwargs, liftover_kwargs, fix_id_kwargs, fix_chr_kwargs, fix_pos_kwargs, fix_allele_kwargs, sanity_check_stats_kwargs, normalize_allele_kwargs, verbose, sweep_mode)\u001b[39m\n\u001b[32m    748\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m basic_check \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[32m    749\u001b[39m     fix_id_kwargs = remove_overlapping_kwargs(fix_id_kwargs,{\u001b[33m\"\u001b[39m\u001b[33mlog\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mremove\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mverbose\u001b[39m\u001b[33m\"\u001b[39m})\n\u001b[32m--> \u001b[39m\u001b[32m750\u001b[39m     \u001b[38;5;28mself\u001b[39m.data = \u001b[43m_fix_ID\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43mlog\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43mlog\u001b[49m\u001b[43m,\u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m=\u001b[49m\u001b[43mverbose\u001b[49m\u001b[43m,\u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mfix_id_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    751\u001b[39m     fix_chr_kwargs = remove_overlapping_kwargs(fix_chr_kwargs,{\u001b[33m\"\u001b[39m\u001b[33mlog\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mremove\u001b[39m\u001b[33m\"\u001b[39m, \u001b[33m\"\u001b[39m\u001b[33mverbose\u001b[39m\u001b[33m\"\u001b[39m})\n\u001b[32m    752\u001b[39m     \u001b[38;5;28mself\u001b[39m.data = _fix_chr(\u001b[38;5;28mself\u001b[39m,remove=remove,log=\u001b[38;5;28mself\u001b[39m.log,verbose=verbose,**fix_chr_kwargs)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/qc/qc_decorator.py:206\u001b[39m, in \u001b[36mwith_logging.<locals>.decorator.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    202\u001b[39m         \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[32m    204\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m is_enough_col==\u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[32m    205\u001b[39m     \u001b[38;5;66;03m# Execute the original function\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m206\u001b[39m     result = \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    207\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    208\u001b[39m     \u001b[38;5;66;03m# not enough cols, return sumstats\u001b[39;00m\n\u001b[32m    209\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m sumstats\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/qc/qc_fix_sumstats.py:141\u001b[39m, in \u001b[36m_fix_ID\u001b[39m\u001b[34m(sumstats_obj, snpid, rsid, chrom, pos, nea, ea, status, fixprefix, fixchrpos, fixid, fixeanea, fixeanea_flip, fixsep, reversea, overwrite, verbose, forcefixid, log)\u001b[39m\n\u001b[32m    139\u001b[39m log.write(\u001b[33m\"\u001b[39m\u001b[33m -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\u001b[39m\u001b[33m\"\u001b[39m,verbose=verbose)\n\u001b[32m    140\u001b[39m \u001b[38;5;66;03m# check if SNPID is CHR:POS:EA:NEA\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m141\u001b[39m is_chrposrefalt = \u001b[43msumstats\u001b[49m\u001b[43m[\u001b[49m\u001b[43msnpid\u001b[49m\u001b[43m]\u001b[49m\u001b[43m.\u001b[49m\u001b[43mstr\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmatch\u001b[49m\u001b[43m(\u001b[49m\u001b[43mSNPID_PATTERN_EXTRACT\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m=\u001b[49m\u001b[43mFLAGS\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[32m    142\u001b[39m \u001b[38;5;66;03m# check if SNPID is NA\u001b[39;00m\n\u001b[32m    143\u001b[39m is_snpid_na = sumstats[snpid].isna()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/core/strings/accessor.py:140\u001b[39m, in \u001b[36mforbid_nonstring_types.<locals>._forbid_nonstring_types.<locals>.wrapper\u001b[39m\u001b[34m(self, *args, **kwargs)\u001b[39m\n\u001b[32m    135\u001b[39m     msg = (\n\u001b[32m    136\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33mCannot use .str.\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfunc_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m with values of \u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    137\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33minferred dtype \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m._inferred_dtype\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    138\u001b[39m     )\n\u001b[32m    139\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(msg)\n\u001b[32m--> \u001b[39m\u001b[32m140\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/core/strings/accessor.py:1388\u001b[39m, in \u001b[36mStringMethods.match\u001b[39m\u001b[34m(self, pat, case, flags, na)\u001b[39m\n\u001b[32m   1349\u001b[39m \u001b[38;5;129m@forbid_nonstring_types\u001b[39m([\u001b[33m\"\u001b[39m\u001b[33mbytes\u001b[39m\u001b[33m\"\u001b[39m])\n\u001b[32m   1350\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mmatch\u001b[39m(\u001b[38;5;28mself\u001b[39m, pat: \u001b[38;5;28mstr\u001b[39m, case: \u001b[38;5;28mbool\u001b[39m = \u001b[38;5;28;01mTrue\u001b[39;00m, flags: \u001b[38;5;28mint\u001b[39m = \u001b[32m0\u001b[39m, na=lib.no_default):\n\u001b[32m   1351\u001b[39m \u001b[38;5;250m    \u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m   1352\u001b[39m \u001b[33;03m    Determine if each string starts with a match of a regular expression.\u001b[39;00m\n\u001b[32m   1353\u001b[39m \n\u001b[32m   (...)\u001b[39m\u001b[32m   1386\u001b[39m \u001b[33;03m    dtype: bool\u001b[39;00m\n\u001b[32m   1387\u001b[39m \u001b[33;03m    \"\"\"\u001b[39;00m\n\u001b[32m-> \u001b[39m\u001b[32m1388\u001b[39m     result = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_data\u001b[49m\u001b[43m.\u001b[49m\u001b[43marray\u001b[49m\u001b[43m.\u001b[49m\u001b[43m_str_match\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpat\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcase\u001b[49m\u001b[43m=\u001b[49m\u001b[43mcase\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mflags\u001b[49m\u001b[43m=\u001b[49m\u001b[43mflags\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna\u001b[49m\u001b[43m=\u001b[49m\u001b[43mna\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1389\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._wrap_result(result, fill_value=na, returns_string=\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/core/strings/object_array.py:259\u001b[39m, in \u001b[36mObjectStringArrayMixin._str_match\u001b[39m\u001b[34m(self, pat, case, flags, na)\u001b[39m\n\u001b[32m    256\u001b[39m regex = re.compile(pat, flags=flags)\n\u001b[32m    258\u001b[39m f = \u001b[38;5;28;01mlambda\u001b[39;00m x: regex.match(x) \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m259\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_str_map\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mna_value\u001b[49m\u001b[43m=\u001b[49m\u001b[43mna\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mnp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mbool\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/core/arrays/string_.py:466\u001b[39m, in \u001b[36mBaseStringArray._str_map\u001b[39m\u001b[34m(self, f, na_value, dtype, convert)\u001b[39m\n\u001b[32m    463\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m dtype == np.dtype(\u001b[33m\"\u001b[39m\u001b[33mbool\u001b[39m\u001b[33m\"\u001b[39m):\n\u001b[32m    464\u001b[39m     \u001b[38;5;66;03m# GH#55736\u001b[39;00m\n\u001b[32m    465\u001b[39m     na_value = \u001b[38;5;28mbool\u001b[39m(na_value)\n\u001b[32m--> \u001b[39m\u001b[32m466\u001b[39m result = \u001b[43mlib\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmap_infer_mask\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m    467\u001b[39m \u001b[43m    \u001b[49m\u001b[43marr\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    468\u001b[39m \u001b[43m    \u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    469\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmask\u001b[49m\u001b[43m.\u001b[49m\u001b[43mview\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43muint8\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    470\u001b[39m \u001b[43m    \u001b[49m\u001b[43mconvert\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[32m    471\u001b[39m \u001b[43m    \u001b[49m\u001b[43mna_value\u001b[49m\u001b[43m=\u001b[49m\u001b[43mna_value\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    472\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# error: Argument 1 to \"dtype\" has incompatible type\u001b[39;49;00m\n\u001b[32m    473\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# \"Union[ExtensionDtype, str, dtype[Any], Type[object]]\"; expected\u001b[39;49;00m\n\u001b[32m    474\u001b[39m \u001b[43m    \u001b[49m\u001b[38;5;66;43;03m# \"Type[object]\"\u001b[39;49;00m\n\u001b[32m    475\u001b[39m \u001b[43m    \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m=\u001b[49m\u001b[43mnp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcast\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mtype\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdtype\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    476\u001b[39m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    478\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m na_value_is_na:\n\u001b[32m    479\u001b[39m     mask[:] = \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/lib.pyx:2922\u001b[39m, in \u001b[36mpandas._libs.lib.map_infer_mask\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/lib.pyx:2959\u001b[39m, in \u001b[36mpandas._libs.lib._map_infer_mask\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/core/strings/object_array.py:258\u001b[39m, in \u001b[36mObjectStringArrayMixin._str_match.<locals>.<lambda>\u001b[39m\u001b[34m(x)\u001b[39m\n\u001b[32m    254\u001b[39m     flags |= re.IGNORECASE\n\u001b[32m    256\u001b[39m regex = re.compile(pat, flags=flags)\n\u001b[32m--> \u001b[39m\u001b[32m258\u001b[39m f = \u001b[38;5;28;01mlambda\u001b[39;00m x: \u001b[43mregex\u001b[49m\u001b[43m.\u001b[49m\u001b[43mmatch\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m)\u001b[49m \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m    259\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._str_map(f, na_value=na, dtype=np.dtype(\u001b[38;5;28mbool\u001b[39m))\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "# 1. Path to your local FASTA (Reference Sequence)\n",
    "ref_fasta = \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\"\n",
    "\n",
    "# 2. Path to your local Reference (using the {chrom} wildcard)\n",
    "# We point to the .parquet files because they are much faster for frequency lookups\n",
    "ref_vcf = \"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.parquet\"\n",
    "\n",
    "# 3. Perform Harmonization\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=ref_fasta,      # Used to check if alleles match the human genome\n",
    "    ref_infer=ref_vcf,      # Used to check AF (Allele Frequency) and strand alignment\n",
    "    ref_alt_freq=\"AF\",      # Note: Check your parquet files, it's usually just \"AF\" or \"EAF\"\n",
    "    threads=6,              # Using your 6 threads\n",
    "    # build=\"38\" is assumed since you lifted over\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "00c6d44b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to the FASTA you just successfully downloaded\n",
    "ref_fasta = \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\"\n",
    "\n",
    "# Path to the VCF (Update filename if you used gl.download_ref)\n",
    "ref_vcf = \"/home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz\"\n",
    "\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=ref_fasta,\n",
    "    ref_infer=ref_vcf,\n",
    "    ref_alt_freq=\"EUR_AF\", \n",
    "    basic_check=True,\n",
    "    threads=4\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26e84088",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd3e48e",
   "metadata": {},
   "source": [
    "### Check summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "742ac24e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mysumstats.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "946f78c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use ref_rsid_tsv because the file is a text table, not a VCF\n",
    "mysumstats.assign_rsid(\n",
    "    ref_rsid_tsv=\"/home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz\",\n",
    "    threads=4\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6efb77ac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8d9fb93",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A. Manually fix Chromosomes to avoid the \"target\" error\n",
    "mysumstats.data['CHR'] = mysumstats.data['CHR'].astype(str).str.replace('chr', '', case=False)\n",
    "mysumstats.data['CHR'] = mysumstats.data['CHR'].replace({'X': '23', 'Y': '24', 'MT': '25'})\n",
    "\n",
    "# B. Set the build globally for this object\n",
    "mysumstats.set_build(\"38\")\n",
    "\n",
    "# C. Run harmonization (removing the 'build' keyword that caused the error)\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=\"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/hg38.fa\",\n",
    "    basic_check=False,\n",
    "    threads=4\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b300662b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
