{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "36639ec4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7e64f42f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import gwaslab as gl\n",
    "\n",
    "# Load gzipped GWAS summary stats\n",
    "df = pd.read_csv(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxytool/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\"\n",
    ")\n",
    "\n",
    "# Split variant into CHR, POS, REF, ALT\n",
    "df[['CHR', 'POS', 'REF', 'ALT']] = df['variant'].str.split(\":\", expand=True)\n",
    "df['POS'] = df['POS'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1d75697",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/03 03:18:42 GWASLab v4.0.2 https://cloufield.github.io/gwaslab/\n",
      "2026/01/03 03:18:42 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/03 03:18:42 Python version: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]\n",
      "2026/01/03 03:18:42 Start to load format from formatbook....\n",
      "2026/01/03 03:18:42  -gwaslab format meta info:\n",
      "2026/01/03 03:18:42   - format_name  : gwaslab\n",
      "2026/01/03 03:18:42   - format_source  : https://cloufield.github.io/gwaslab/\n",
      "2026/01/03 03:18:42   - format_version  : 20231220_v4\n",
      "2026/01/03 03:18:42 Start to initialize gl.Sumstats from pandas DataFrame ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/03 03:18:55  -Reading columns          : REF,CHR,POS,ALT,NOTE\n",
      "2026/01/03 03:18:55  -Renaming columns to      : REF,CHR,POS,ALT,NOTE\n",
      "2026/01/03 03:18:55  -Current Dataframe shape : 13791467  x  15\n",
      "2026/01/03 03:19:03  -Initiating a status column: STATUS ...\n",
      "2026/01/03 03:19:03 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/03 03:19:04  -Converting REF and ALT to NEA and EA ...\n",
      "2026/01/03 03:19:16 Start to reorder the columns ...(v4.0.2)\n",
      "2026/01/03 03:19:16  -Reordering columns to    : CHR,POS,EA,NEA,STATUS,variant,minor_allele,minor_AF,low_confidence_variant,n_complete_samples,AC,ytx,beta,se,tstat,pval\n",
      "2026/01/03 03:19:23 Finished reordering the columns.\n",
      "2026/01/03 03:19:23  -Trying to convert datatype for CHR: string -> Int64...Failed...\n",
      "2026/01/03 03:19:45  -No rsID/SNPID found; created SNPID from CHR:POS[:NEA:EA]\n",
      "2026/01/03 03:19:45  -Column  : CHR    POS   EA       NEA      STATUS variant minor_allele minor_AF low_confidence_variant n_complete_samples AC      ytx     beta    se      tstat   pval    SNPID \n",
      "2026/01/03 03:19:45  -DType   : string int64 category category int64  object  object       float64  bool                   int64              float64 float64 float64 float64 float64 float64 string\n",
      "2026/01/03 03:19:45  -Verified: F      T     T        T        T      NA      NA           NA       NA                     NA                 NA      NA      NA      NA      NA      NA      T     \n",
      "2026/01/03 03:19:45 #WARNING! Columns with possibly incompatible dtypes: CHR\n",
      "2026/01/03 03:19:45 #WARNING! Consider using Sumstats.fix_chr() to fix CHR dtype\n",
      "2026/01/03 03:19:45  -Current Dataframe memory usage: 1595.18 MB\n",
      "2026/01/03 03:19:45 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(df,fmt=\"gwaslab\",other=[\"NOTE\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "47e4105c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>minor_AF</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>SNPID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.446880e-09</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>1.238950e-01</td>\n",
       "      <td>894.616000</td>\n",
       "      <td>1204.870000</td>\n",
       "      <td>0.742499</td>\n",
       "      <td>0.457786</td>\n",
       "      <td>1:15791:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9999999</td>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>5.768250e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>1.035150e+02</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>-1.150590</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>1:69487:G:A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.878960e-04</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.278000</td>\n",
       "      <td>3.644560e+03</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>-1.143630</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>1:69569:T:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>5.681100e-06</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>1.018400e+02</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>-1.145510</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>1:139853:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9999999</td>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>1.105900e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.100000</td>\n",
       "      <td>2.179000e+06</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>-0.839228</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>1:692794:CA:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9999999</td>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>2.454730e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.000000</td>\n",
       "      <td>4.837480e+06</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>-1.516370</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>X:154929412:C:T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9999999</td>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.297260e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.000000</td>\n",
       "      <td>4.525550e+06</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>-2.421510</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9999999</td>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>2.394300e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.000000</td>\n",
       "      <td>4.717910e+06</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>-1.817100</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9999999</td>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>2.458870e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.000000</td>\n",
       "      <td>4.845680e+06</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>-1.524700</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>X:154930230:A:G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9999999</td>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>1.955900e-02</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.800000</td>\n",
       "      <td>3.857390e+05</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.193216</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>X:154930487:T:A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         CHR        POS EA  NEA   STATUS            variant minor_allele  \\\n",
       "0          1      15791  T    C  9999999        1:15791:C:T            T   \n",
       "1          1      69487  A    G  9999999        1:69487:G:A            A   \n",
       "2          1      69569  C    T  9999999        1:69569:T:C            C   \n",
       "3          1     139853  T    C  9999999       1:139853:C:T            T   \n",
       "4          1     692794  C   CA  9999999      1:692794:CA:C            C   \n",
       "...       ..        ... ..  ...      ...                ...          ...   \n",
       "13791462   X  154929412  T    C  9999999    X:154929412:C:T            T   \n",
       "13791463   X  154929637  C   CT  9999999   X:154929637:CT:C            C   \n",
       "13791464   X  154929952  C  CAA  9999999  X:154929952:CAA:C            C   \n",
       "13791465   X  154930230  G    A  9999999    X:154930230:A:G            G   \n",
       "13791466   X  154930487  A    T  9999999    X:154930487:T:A            A   \n",
       "\n",
       "              minor_AF  low_confidence_variant  n_complete_samples  \\\n",
       "0         5.446880e-09                    True              359983   \n",
       "1         5.768250e-06                    True              359983   \n",
       "2         1.878960e-04                    True              359983   \n",
       "3         5.681100e-06                    True              359983   \n",
       "4         1.105900e-01                   False              359983   \n",
       "...                ...                     ...                 ...   \n",
       "13791462  2.454730e-01                   False              359983   \n",
       "13791463  2.297260e-01                   False              359983   \n",
       "13791464  2.394300e-01                   False              359983   \n",
       "13791465  2.458870e-01                   False              359983   \n",
       "13791466  1.955900e-02                   False              359983   \n",
       "\n",
       "                     AC           ytx        beta           se     tstat  \\\n",
       "0              0.003922  1.238950e-01  894.616000  1204.870000  0.742499   \n",
       "1              4.152940  1.035150e+02   -2.715450     2.360060 -1.150590   \n",
       "2            135.278000  3.644560e+03   -0.484284     0.423462 -1.143630   \n",
       "3              4.090200  1.018400e+02   -2.703560     2.360130 -1.145510   \n",
       "4          79621.100000  2.179000e+06   -0.016436     0.019585 -0.839228   \n",
       "...                 ...           ...         ...          ...       ...   \n",
       "13791462  176732.000000  4.837480e+06   -0.016260     0.010723 -1.516370   \n",
       "13791463  165395.000000  4.525550e+06   -0.027098     0.011190 -2.421510   \n",
       "13791464  172381.000000  4.717910e+06   -0.020494     0.011278 -1.817100   \n",
       "13791465  177030.000000  4.845680e+06   -0.016347     0.010721 -1.524700   \n",
       "13791466   14081.800000  3.857390e+05    0.007107     0.036783  0.193216   \n",
       "\n",
       "              pval              SNPID  \n",
       "0         0.457786        1:15791:C:T  \n",
       "1         0.249902        1:69487:G:A  \n",
       "2         0.252778        1:69569:T:C  \n",
       "3         0.251997       1:139853:C:T  \n",
       "4         0.401342      1:692794:CA:C  \n",
       "...            ...                ...  \n",
       "13791462  0.129427    X:154929412:C:T  \n",
       "13791463  0.015456   X:154929637:CT:C  \n",
       "13791464  0.069202  X:154929952:CAA:C  \n",
       "13791465  0.127334    X:154930230:A:G  \n",
       "13791466  0.846790    X:154930487:T:A  \n",
       "\n",
       "[13791467 rows x 17 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f0de526a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/03 03:19:56 Start to check available reference files...\n",
      "2026/01/03 03:19:56  - Available keywords:  1kg_eas_hg19 1kg_eur_hg19 1kg_eas_hg38 1kg_eur_hg38 1kg_sas_hg19 1kg_amr_hg19 1kg_sas_hg38 1kg_amr_hg38 1kg_afr_hg19 1kg_pan_hg19 1kg_afr_hg38 1kg_pan_hg38 dbsnp_v151_hg19 dbsnp_v151_hg38 dbsnp_v157_hg19 dbsnp_v157_hg38 ucsc_genome_hg19 ucsc_genome_hg38 1kg_dbsnp151_hg19_auto 1kg_dbsnp151_hg38_auto recombination_hg19 recombination_hg38 ensembl_hg19_gtf ensembl_hg38_gtf refseq_hg19_gtf refseq_hg38_gtf testlink 19to38 38to19 1kg_hm3_hg38_eaf 1kg_hm3_hg19_eaf\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'1kg_eas_hg19': {'description': '1000 Genomes Project East Asian (1KG EAS) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EAS population'},\n",
       " '1kg_eur_hg19': {'description': '1000 Genomes Project European (1KG EUR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EUR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EUR population'},\n",
       " '1kg_eas_hg38': {'description': '1000 Genomes Project East Asian (1KG EAS) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EAS population'},\n",
       " '1kg_eur_hg38': {'description': '1000 Genomes Project European (1KG EUR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the EUR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for EUR population'},\n",
       " '1kg_sas_hg19': {'description': '1000 Genomes Project South Asian (1KG SAS) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the SAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for SAS population'},\n",
       " '1kg_amr_hg19': {'description': '1000 Genomes Project American (1KG AMR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AMR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AMR population'},\n",
       " '1kg_sas_hg38': {'description': '1000 Genomes Project South Asian (1KG SAS) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the SAS population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for SAS population'},\n",
       " '1kg_amr_hg38': {'description': '1000 Genomes Project American (1KG AMR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AMR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AMR population'},\n",
       " '1kg_afr_hg19': {'description': '1000 Genomes Project African (1KG AFR) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AFR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AFR population'},\n",
       " '1kg_pan_hg19': {'description': '1000 Genomes Project Pan-ancestry (1KG PAN) VCF on the hg19 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the PAN population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for PAN population'},\n",
       " '1kg_afr_hg38': {'description': '1000 Genomes Project African (1KG AFR) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the AFR population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for AFR population'},\n",
       " '1kg_pan_hg38': {'description': '1000 Genomes Project Pan-ancestry (1KG PAN) VCF on the hg38 reference. Multi-allelic variants were decomposed. Variants were normalized. The INFO field includes the AF annotation, representing allele frequency in the PAN population.',\n",
       "  'suggested_use': 'LD reference panel for creating region plot; infer strand for PAN population'},\n",
       " 'dbsnp_v151_hg19': {'description': 'dbSNP v151 hg19 is the 2018 release of the dbSNP database mapped to the human genome assembly GRCh37/hg19, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v151_hg38': {'description': 'dbSNP v151 hg38 is the 2018 release of the dbSNP database mapped to the human genome assembly GRCh38, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v157_hg19': {'description': 'dbSNP v157 hg19 is the release of the dbSNP database mapped to the human genome assembly GRCh37/hg19, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'dbsnp_v157_hg38': {'description': 'dbSNP v157 hg38 is the release of the dbSNP database mapped to the human genome assembly GRCh38, providing standardized rsID identifiers and allele information for known genetic variants.',\n",
       "  'suggested_use': 'rsID assignment'},\n",
       " 'ucsc_genome_hg19': {'description': 'UCSC Genome reference sequence on hg19 (GRCh37), providing the complete human genome assembly used for alignment and variant annotation.',\n",
       "  'suggested_use': 'Reference genome for alignment and annotation'},\n",
       " 'ucsc_genome_hg38': {'description': 'UCSC Genome reference sequence on hg38 (GRCh38), providing the complete human genome assembly used for alignment and variant annotation.',\n",
       "  'suggested_use': 'Reference genome for alignment and annotation'},\n",
       " '1kg_dbsnp151_hg19_auto': {'description': '1KG dbSNP151 hg19 auto variants table containing SNP ID, chromosome, position, effect allele, non-effect allele, and rsID mappings for autosomal variants.',\n",
       "  'suggested_use': 'SNP annotation and rsID mapping'},\n",
       " '1kg_dbsnp151_hg38_auto': {'description': '1KG dbSNP151 hg38 auto variants table containing SNP ID, chromosome, position, effect allele, non-effect allele, and rsID mappings for autosomal variants.',\n",
       "  'suggested_use': 'SNP annotation and rsID mapping'},\n",
       " 'recombination_hg19': {'description': 'Recombination map for hg19 (GRCh37), providing genetic distance estimates in centimorgans used for linkage disequilibrium and haplotype analysis.',\n",
       "  'suggested_use': 'Genetic mapping and LD calculations'},\n",
       " 'recombination_hg38': {'description': 'Recombination map for hg38 (GRCh38), providing genetic distance estimates in centimorgans used for linkage disequilibrium and haplotype analysis.',\n",
       "  'suggested_use': 'Genetic mapping and LD calculations'},\n",
       " 'ensembl_hg19_gtf': {'description': 'Ensembl annotation for hg19 (GRCh37) release 87, providing comprehensive gene models, transcript structures, and functional annotations for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'ensembl_hg38_gtf': {'description': 'Ensembl annotation for hg38 (GRCh38) release 109, providing comprehensive gene models, transcript structures, and functional annotations for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'refseq_hg19_gtf': {'description': 'RefSeq annotation for hg19 (GRCh37), providing curated gene models and transcript structures for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'refseq_hg38_gtf': {'description': 'RefSeq annotation for hg38 (GRCh38), providing curated gene models and transcript structures for the human genome.',\n",
       "  'suggested_use': 'Gene annotation and functional analysis'},\n",
       " 'testlink': {'description': 'Test link', 'suggested_use': ''},\n",
       " '19to38': {'description': 'LiftOver chain file for converting genomic coordinates from hg19 (GRCh37) to hg38 (GRCh38) reference builds, enabling cross-reference genome analysis.',\n",
       "  'suggested_use': 'Genome build conversion and coordinate remapping'},\n",
       " '38to19': {'description': 'LiftOver chain file for converting genomic coordinates from hg38 (GRCh38) to hg19 (GRCh37) reference builds, enabling cross-reference genome analysis.',\n",
       "  'suggested_use': 'Genome build conversion and coordinate remapping'},\n",
       " '1kg_hm3_hg38_eaf': {'description': '1KG HapMap3 hg38 EAF (GRCh38) provides allele frequency estimates for the HapMap3 populations on the hg38 reference genome, derived from 1000 Genomes Project data.',\n",
       "  'suggested_use': 'Ancestry inference and population structure analysis'},\n",
       " '1kg_hm3_hg19_eaf': {'description': '1KG HapMap3 hg19 EAF (GRCh37) provides allele frequency estimates for the HapMap3 populations on the hg19 reference genome, derived from 1000 Genomes Project data.',\n",
       "  'suggested_use': 'Ancestry inference and population structure analysis'}}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.check_available_ref()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d3e6d6af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The history saving thread hit an unexpected error (OperationalError('database or disk is full')).History will not be written to the database.\n",
      "2026/01/03 03:20:12 Start to download  1kg_dbsnp151_hg19_auto  ...\n",
      "2026/01/03 03:20:12  -Downloading to: /home/icog-bioai2/.gwaslab/1kg_dbsnp151_hg19_auto.txt.gz\n",
      "2026/01/03 03:20:12  -File /home/icog-bioai2/.gwaslab/1kg_dbsnp151_hg19_auto.txt.gz exists.\n",
      "2026/01/03 03:20:12  - Updating record in config file...\n",
      "2026/01/03 03:20:12 Downloaded  1kg_dbsnp151_hg19_auto  successfully!\n"
     ]
    }
   ],
   "source": [
    "gl.download_ref(\"1kg_dbsnp151_hg19_auto\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4eecb854",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/03 03:43:07 Start to check SNPID/rsID ...(v4.0.2)\n",
      "2026/01/03 03:43:07  -Checking SNPID data type...\n",
      "2026/01/03 03:43:07  -Checking NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,\n",
      "2026/01/03 03:43:07  -Checking if SNPID contains NA strings...\n",
      "2026/01/03 03:43:07  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/03 03:43:26 Finished checking SNPID/rsID.\n",
      "2026/01/03 03:43:26 Start to fix chromosome notation (CHR) ...(v4.0.2)\n",
      "2026/01/03 03:43:26  -Checking CHR data type...\n",
      "2026/01/03 03:43:29  -Variants with standardized chromosome notation: 13791467\n",
      "2026/01/03 03:43:29  -All CHR are already fixed...\n",
      "2026/01/03 03:43:41 Finished fixing chromosome notation (CHR).\n",
      "2026/01/03 03:43:41 Start to fix basepair positions (POS) ...(v4.0.2)\n",
      "2026/01/03 03:43:41  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/03 03:43:53  -Position bound:(0 , 250,000,000)\n",
      "2026/01/03 03:43:53  -Removed variants outliers: 0\n",
      "2026/01/03 03:44:01 Finished fixing basepair positions (POS).\n",
      "2026/01/03 03:44:01 Start to fix alleles (EA and NEA) ...(v4.0.2)\n",
      "2026/01/03 03:44:01  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/03 03:44:09  -Variants with bad EA: 0\n",
      "2026/01/03 03:44:09  -Variants with bad NEA: 0\n",
      "2026/01/03 03:44:09  -Variants with NA for EA or NEA: 0\n",
      "2026/01/03 03:44:09  -Variants with same EA and NEA: 0\n",
      "2026/01/03 03:44:11  -Detected 0 variants with alleles that contain bases other than A/C/T/G .\n",
      "2026/01/03 03:44:35 Finished fixing alleles (EA and NEA).\n",
      "2026/01/03 03:44:35 Start to perform sanity check for statistics ...(v4.0.2)\n",
      "2026/01/03 03:44:35  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/03 03:44:42  -Checking if any columns are empty...\n",
      "2026/01/03 03:44:42  -Checking if -1e-07 < EAF < 1.0000001 ...\n",
      "2026/01/03 03:44:53  -Removed 0 variants with bad/na EAF.\n",
      "2026/01/03 03:44:53  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/03 03:44:53  -Removed 0 variants with bad statistics in total.\n",
      "2026/01/03 03:44:53  -Data types for each column:\n",
      "2026/01/03 03:44:53 Finished sanity check for statistics.\n",
      "2026/01/03 03:44:54  -Number of threads/cores to use: 3\n",
      "2026/01/03 03:44:54 Start to normalize indels ...(v4.0.2)\n",
      "2026/01/03 03:44:55  -No available variants to normalize..\n",
      "2026/01/03 03:44:55 Finished normalizing indels.\n",
      "2026/01/03 03:44:55 Start to reorder the columns ...(v4.0.2)\n",
      "2026/01/03 03:44:55  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,variant,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,beta,se,tstat,pval\n",
      "2026/01/03 03:45:03 Finished reordering the columns.\n",
      "2026/01/03 03:45:03 Start to sort the genome coordinates ...(v4.0.2)\n",
      "2026/01/03 03:45:27 Finished sorting coordinates.\n",
      "2026/01/03 03:45:28 Start to check if NEA is aligned with reference sequence ...(v4.0.2)\n",
      "2026/01/03 03:45:30  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/galaxytool/Data/ref/hg19.fa\n",
      "2026/01/03 03:45:52    -Loading and building numpy fasta records:\n",
      "2026/01/03 03:45:59  -Variants allele on given reference sequence :  0\n",
      "2026/01/03 03:45:59  -Variants flipped :  0\n",
      "2026/01/03 03:45:59   -Raw Matching rate :  0.00%\n",
      "2026/01/03 03:45:59 #WARNING! Matching rate is low, please check if the right reference genome is used.\n",
      "2026/01/03 03:45:59  -Variants inferred reverse_complement :  0\n",
      "2026/01/03 03:45:59  -Variants inferred reverse_complement_flipped :  0\n",
      "2026/01/03 03:45:59  -Both allele on genome + unable to distinguish :  0\n",
      "2026/01/03 03:45:59  -Variants not on given reference sequence :  0\n",
      "2026/01/03 03:46:00 Finished checking if NEA is aligned with reference sequence.\n",
      "2026/01/03 03:46:00 Start to adjust statistics based on STATUS code ...(v4.0.2)\n",
      "2026/01/03 03:46:10  -No statistics have been changed.\n",
      "2026/01/03 03:46:10 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/03 03:46:10  -Number of threads/cores to use: 3\n",
      "2026/01/03 03:46:10 Start to infer strand for palindromic SNPs/align indistinguishable indels ...(v4.0.2)\n",
      "2026/01/03 03:46:10  -Field for alternative allele frequency in VCF INFO: AF\n",
      "2026/01/03 03:46:23  -Identified  1929225  palindromic SNPs...\n",
      "2026/01/03 03:46:27  -After filtering by MAF<= 0.4 , 0 palindromic SNPs with unknown strand will be inferred...\n",
      "2026/01/03 03:46:27 #WARNING! No palindromic variants available for checking.\n",
      "2026/01/03 03:46:29   -Non-palindromic :  10608049\n",
      "2026/01/03 03:46:30   -Palindromic SNPs on + strand:  0\n",
      "2026/01/03 03:46:31   -Palindromic SNPs on - strand and needed to be flipped: 0\n",
      "2026/01/03 03:46:31   -Palindromic SNPs with MAF not available to infer :  156873\n",
      "2026/01/03 03:46:32   -Palindromic SNPs lacking reference matches, or with reference ALT unavailable or above MAF threshold for inference :  0\n",
      "2026/01/03 03:46:36  -Identified  0  indistinguishable Indels...\n",
      "2026/01/03 03:46:36 #WARNING! No indistinguishable indels available for checking.\n",
      "2026/01/03 03:46:36 Finished inferring strand for palindromic SNPs/align indistinguishable indels.\n",
      "2026/01/03 03:46:36 Start to adjust statistics based on STATUS code ...(v4.0.2)\n",
      "2026/01/03 03:46:45  -No statistics have been changed.\n",
      "2026/01/03 03:46:45 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/03 03:46:45 Start to sort the genome coordinates ...(v4.0.2)\n",
      "2026/01/03 03:47:10 Finished sorting coordinates.\n",
      "2026/01/03 03:47:10 Start to reorder the columns ...(v4.0.2)\n",
      "2026/01/03 03:47:10  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,variant,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,beta,se,tstat,pval\n",
      "2026/01/03 03:47:18 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.446880e-09</td>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>1.238950e-01</td>\n",
       "      <td>894.616000</td>\n",
       "      <td>1204.870000</td>\n",
       "      <td>0.742499</td>\n",
       "      <td>0.457786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.768250e-06</td>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>1.035150e+02</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>-1.150590</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>1.878960e-04</td>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.278000</td>\n",
       "      <td>3.644560e+03</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>-1.143630</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.681100e-06</td>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>1.018400e+02</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>-1.145510</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1.105900e-01</td>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.100000</td>\n",
       "      <td>2.179000e+06</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>-0.839228</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>2.454730e-01</td>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.000000</td>\n",
       "      <td>4.837480e+06</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>-1.516370</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>2.297260e-01</td>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.000000</td>\n",
       "      <td>4.525550e+06</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>-2.421510</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>2.394300e-01</td>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.000000</td>\n",
       "      <td>4.717910e+06</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>-1.817100</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>2.458870e-01</td>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.000000</td>\n",
       "      <td>4.845680e+06</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>-1.524700</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>1.955900e-02</td>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.800000</td>\n",
       "      <td>3.857390e+05</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.193216</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x737b0c921f10>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename minor_AF to EAF\n",
    "mysumstats.data.rename(columns={\"minor_AF\": \"EAF\"}, inplace=True)\n",
    "\n",
    "# Now harmonize\n",
    "mysumstats.harmonize(basic_check=True,  # Perform basic check\n",
    "                     threads=3,         # Number of threads to use\n",
    "                     ref_seq=\"/mnt/hdd_1/ofgeha/galaxytool/Data/ref/hg19.fa\",  # Reference genome\n",
    "                     ref_infer=\"/mnt/hdd_1/ofgeha/galaxytool/Data/ref/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz\",  # Reference VCF file\n",
    "                     ref_alt_freq=\"AF\")  # Field for alternative allele frequency in the VCF\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f536d852",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.446880e-09</td>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>1.238950e-01</td>\n",
       "      <td>894.616000</td>\n",
       "      <td>1204.870000</td>\n",
       "      <td>0.742499</td>\n",
       "      <td>0.457786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.768250e-06</td>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>1.035150e+02</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>-1.150590</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>1.878960e-04</td>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.278000</td>\n",
       "      <td>3.644560e+03</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>-1.143630</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.681100e-06</td>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>1.018400e+02</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>-1.145510</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1.105900e-01</td>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.100000</td>\n",
       "      <td>2.179000e+06</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>-0.839228</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>2.454730e-01</td>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.000000</td>\n",
       "      <td>4.837480e+06</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>-1.516370</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>2.297260e-01</td>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.000000</td>\n",
       "      <td>4.525550e+06</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>-2.421510</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>2.394300e-01</td>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.000000</td>\n",
       "      <td>4.717910e+06</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>-1.817100</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>2.458870e-01</td>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.000000</td>\n",
       "      <td>4.845680e+06</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>-1.524700</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>1.955900e-02</td>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.800000</td>\n",
       "      <td>3.857390e+05</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.193216</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS           EAF  \\\n",
       "0               1:15791:C:T    1      15791  T    C  9960090  5.446880e-09   \n",
       "1               1:69487:G:A    1      69487  A    G  9960090  5.768250e-06   \n",
       "2               1:69569:T:C    1      69569  C    T  9960090  1.878960e-04   \n",
       "3              1:139853:C:T    1     139853  T    C  9960090  5.681100e-06   \n",
       "4             1:692794:CA:C    1     692794  C   CA  9960399  1.105900e-01   \n",
       "...                     ...  ...        ... ..  ...      ...           ...   \n",
       "13791462    X:154929412:C:T   23  154929412  T    C  9960090  2.454730e-01   \n",
       "13791463   X:154929637:CT:C   23  154929637  C   CT  9960399  2.297260e-01   \n",
       "13791464  X:154929952:CAA:C   23  154929952  C  CAA  9960399  2.394300e-01   \n",
       "13791465    X:154930230:A:G   23  154930230  G    A  9960090  2.458870e-01   \n",
       "13791466    X:154930487:T:A   23  154930487  A    T  9960099  1.955900e-02   \n",
       "\n",
       "                    variant minor_allele  low_confidence_variant  \\\n",
       "0               1:15791:C:T            T                    True   \n",
       "1               1:69487:G:A            A                    True   \n",
       "2               1:69569:T:C            C                    True   \n",
       "3              1:139853:C:T            T                    True   \n",
       "4             1:692794:CA:C            C                   False   \n",
       "...                     ...          ...                     ...   \n",
       "13791462    X:154929412:C:T            T                   False   \n",
       "13791463   X:154929637:CT:C            C                   False   \n",
       "13791464  X:154929952:CAA:C            C                   False   \n",
       "13791465    X:154930230:A:G            G                   False   \n",
       "13791466    X:154930487:T:A            A                   False   \n",
       "\n",
       "          n_complete_samples             AC           ytx        beta  \\\n",
       "0                     359983       0.003922  1.238950e-01  894.616000   \n",
       "1                     359983       4.152940  1.035150e+02   -2.715450   \n",
       "2                     359983     135.278000  3.644560e+03   -0.484284   \n",
       "3                     359983       4.090200  1.018400e+02   -2.703560   \n",
       "4                     359983   79621.100000  2.179000e+06   -0.016436   \n",
       "...                      ...            ...           ...         ...   \n",
       "13791462              359983  176732.000000  4.837480e+06   -0.016260   \n",
       "13791463              359983  165395.000000  4.525550e+06   -0.027098   \n",
       "13791464              359983  172381.000000  4.717910e+06   -0.020494   \n",
       "13791465              359983  177030.000000  4.845680e+06   -0.016347   \n",
       "13791466              359983   14081.800000  3.857390e+05    0.007107   \n",
       "\n",
       "                   se     tstat      pval  \n",
       "0         1204.870000  0.742499  0.457786  \n",
       "1            2.360060 -1.150590  0.249902  \n",
       "2            0.423462 -1.143630  0.252778  \n",
       "3            2.360130 -1.145510  0.251997  \n",
       "4            0.019585 -0.839228  0.401342  \n",
       "...               ...       ...       ...  \n",
       "13791462     0.010723 -1.516370  0.129427  \n",
       "13791463     0.011190 -2.421510  0.015456  \n",
       "13791464     0.011278 -1.817100  0.069202  \n",
       "13791465     0.010721 -1.524700  0.127334  \n",
       "13791466     0.036783  0.193216  0.846790  \n",
       "\n",
       "[13791467 rows x 17 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "2fc41d6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/03 04:23:02  -Number of threads/cores to use: 1\n",
      "2026/01/03 04:23:02 Start to assign rsID using reference file ...(v4.0.2)\n",
      "2026/01/03 04:23:03  -13791467 rsID could be possibly fixed...\n",
      "2026/01/03 04:23:11  -Setting block size:  5000000\n",
      "2026/01/03 04:23:11  -Loading block: 0   1   2   3   4   5   6   7   8   9   10   11   12   13   14   15   \n",
      "2026/01/03 04:30:31  -rsID annotation for 974144 variants needed to be fixed!\n",
      "2026/01/03 04:30:31  -Annotated 12817323 rsID successfully!\n",
      "2026/01/03 04:30:31  -Current Dataframe shape : 13791467 x 18 ; Memory usage: 1688.82 MB\n",
      "2026/01/03 04:30:31 Finished assign rsID using reference file.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>variant</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>tstat</th>\n",
       "      <th>pval</th>\n",
       "      <th>rsID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>15791</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.446880e-09</td>\n",
       "      <td>1:15791:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>0.003922</td>\n",
       "      <td>1.238950e-01</td>\n",
       "      <td>894.616000</td>\n",
       "      <td>1204.870000</td>\n",
       "      <td>0.742499</td>\n",
       "      <td>0.457786</td>\n",
       "      <td>rs547522712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.768250e-06</td>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.152940</td>\n",
       "      <td>1.035150e+02</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>-1.150590</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>rs568226429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>1.878960e-04</td>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.278000</td>\n",
       "      <td>3.644560e+03</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>-1.143630</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>rs2531267</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>5.681100e-06</td>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.090200</td>\n",
       "      <td>1.018400e+02</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>-1.145510</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>rs533633326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>1.105900e-01</td>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.100000</td>\n",
       "      <td>2.179000e+06</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>-0.839228</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>rs1268544509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791462</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>2.454730e-01</td>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.000000</td>\n",
       "      <td>4.837480e+06</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>-1.516370</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791463</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>2.297260e-01</td>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.000000</td>\n",
       "      <td>4.525550e+06</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>-2.421510</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791464</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>2.394300e-01</td>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.000000</td>\n",
       "      <td>4.717910e+06</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>-1.817100</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791465</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>2.458870e-01</td>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.000000</td>\n",
       "      <td>4.845680e+06</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>-1.524700</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13791466</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>1.955900e-02</td>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.800000</td>\n",
       "      <td>3.857390e+05</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.193216</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13791467 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x737b0c921f10>"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.assign_rsid(ref_rsid_tsv=gl.get_path(\"1kg_dbsnp151_hg19_auto\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a018b003",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'overview': {'Row_num': '13791467',\n",
       "  'Column_num': '17',\n",
       "  'Column_names': 'SNPID,CHR,POS,EA,NEA,STATUS,EAF,variant,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,beta,se,tstat,pval',\n",
       "  'Last_checked_time': 'Sat Jan  3 04:08:51 2026',\n",
       "  'QC and Harmonization': {'basic_check': {'performed': True,\n",
       "    'last_executed': '2026-01-03 03:45:27',\n",
       "    'parameters_used': {'basic_check': True,\n",
       "     'ref_seq': '/mnt/hdd_1/ofgeha/galaxytool/Data/ref/hg19.fa',\n",
       "     'ref_rsid_tsv': None,\n",
       "     'ref_rsid_vcf': None,\n",
       "     'ref_infer': '/mnt/hdd_1/ofgeha/galaxytool/Data/ref/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz',\n",
       "     'ref_alt_freq': 'AF',\n",
       "     'ref_maf_threshold': 0.4,\n",
       "     'maf_threshold': 0.4,\n",
       "     'threads': 3,\n",
       "     'remove': False,\n",
       "     'check_ref_kwargs': {},\n",
       "     'remove_dup_kwargs': {},\n",
       "     'assign_rsid_kwargs': {},\n",
       "     'infer_strand_kwargs': {},\n",
       "     'flip_allele_stats_kwargs': {},\n",
       "     'liftover_kwargs': {},\n",
       "     'fix_id_kwargs': {},\n",
       "     'fix_chr_kwargs': {},\n",
       "     'fix_pos_kwargs': {},\n",
       "     'fix_allele_kwargs': {},\n",
       "     'sanity_check_stats_kwargs': {},\n",
       "     'normalize_allele_kwargs': {},\n",
       "     'verbose': True,\n",
       "     'sweep_mode': False}},\n",
       "   'harmonize': {'performed': True,\n",
       "    'last_executed': '2026-01-03 03:47:18',\n",
       "    'parameters_used': {'basic_check': True,\n",
       "     'ref_seq': '/mnt/hdd_1/ofgeha/galaxytool/Data/ref/hg19.fa',\n",
       "     'ref_rsid_tsv': None,\n",
       "     'ref_rsid_vcf': None,\n",
       "     'ref_infer': '/mnt/hdd_1/ofgeha/galaxytool/Data/ref/EAS.ALL.split_norm_af.1kgp3v5.hg19.vcf.gz',\n",
       "     'ref_alt_freq': 'AF',\n",
       "     'ref_maf_threshold': 0.4,\n",
       "     'maf_threshold': 0.4,\n",
       "     'threads': 3,\n",
       "     'remove': False,\n",
       "     'check_ref_kwargs': {},\n",
       "     'remove_dup_kwargs': {},\n",
       "     'assign_rsid_kwargs': {},\n",
       "     'infer_strand_kwargs': {},\n",
       "     'flip_allele_stats_kwargs': {},\n",
       "     'liftover_kwargs': {},\n",
       "     'fix_id_kwargs': {},\n",
       "     'fix_chr_kwargs': {},\n",
       "     'fix_pos_kwargs': {},\n",
       "     'fix_allele_kwargs': {},\n",
       "     'sanity_check_stats_kwargs': {},\n",
       "     'normalize_allele_kwargs': {},\n",
       "     'verbose': True,\n",
       "     'sweep_mode': False}},\n",
       "   'qc_and_harmonization_status': {'qc': {'id': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:45:27',\n",
       "      'parameters_used': {}},\n",
       "     'chr': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:45:27',\n",
       "      'parameters_used': {}},\n",
       "     'pos': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:45:27',\n",
       "      'parameters_used': {}},\n",
       "     'allele': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:45:27',\n",
       "      'parameters_used': {}},\n",
       "     'sanity': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:45:27',\n",
       "      'parameters_used': {}},\n",
       "     'consistency': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:45:27',\n",
       "      'parameters_used': {}},\n",
       "     'normalize': {'performed': False,\n",
       "      'last_executed': '2026-01-03 03:22:49',\n",
       "      'parameters_used': {}},\n",
       "     'remove_dup': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}},\n",
       "     'sort_coord': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:47:10',\n",
       "      'parameters_used': {'chrom': 'CHR', 'pos': 'POS', 'reindex': True}},\n",
       "     'sort_column': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:47:18',\n",
       "      'parameters_used': {'order': ['SNPID',\n",
       "        'rsID',\n",
       "        'CHR',\n",
       "        'POS',\n",
       "        'EA',\n",
       "        'NEA',\n",
       "        'STATUS',\n",
       "        'REF',\n",
       "        'ALT',\n",
       "        'EAF',\n",
       "        'RAF',\n",
       "        'NEAF',\n",
       "        'MAF',\n",
       "        'DAF',\n",
       "        'INFO',\n",
       "        'BETA',\n",
       "        'SE',\n",
       "        'BETA_95U',\n",
       "        'BETA_95L',\n",
       "        'OR',\n",
       "        'OR_95U',\n",
       "        'OR_95L',\n",
       "        'HR',\n",
       "        'HR_95U',\n",
       "        'HR_95L',\n",
       "        'CHISQ',\n",
       "        'Z',\n",
       "        'T',\n",
       "        'F',\n",
       "        'P',\n",
       "        'P_MANTISSA',\n",
       "        'P_EXPONENT',\n",
       "        'MLOG10P',\n",
       "        'SNPR2',\n",
       "        'DOF',\n",
       "        'P_HET',\n",
       "        'I2_HET',\n",
       "        'I2',\n",
       "        'DIRECTION',\n",
       "        'DENSITY',\n",
       "        'N',\n",
       "        'N_CASE',\n",
       "        'N_CONTROL',\n",
       "        'N_EFF',\n",
       "        'GENENAME',\n",
       "        'CIS/TRANS',\n",
       "        'DISTANCE_TO_KNOWN',\n",
       "        'LOCATION_OF_KNOWN',\n",
       "        'KNOWN_ID',\n",
       "        'KNOWN_PUBMED_ID',\n",
       "        'KNOWN_AUTHOR',\n",
       "        'KNOWN_SET_VARIANT',\n",
       "        'KNOWN_VARIANT',\n",
       "        'KNOWN_SET',\n",
       "        'NOVEL',\n",
       "        'PIP',\n",
       "        'CREDIBLE_SET_INDEX',\n",
       "        'N_SNP',\n",
       "        'LOCUS',\n",
       "        'STUDY',\n",
       "        'TEST',\n",
       "        'BETA_RANDOM',\n",
       "        'SE_RANDOM',\n",
       "        'Z_RANDOM',\n",
       "        'P_RANDOM']}}},\n",
       "    'harmonize': {'check_ref': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:47:18',\n",
       "      'parameters_used': {}},\n",
       "     'flip_allele_stats': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:47:18',\n",
       "      'parameters_used': {}},\n",
       "     'infer_strand': {'performed': True,\n",
       "      'last_executed': '2026-01-03 03:47:18',\n",
       "      'parameters_used': {}},\n",
       "     'assign_rsid': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}},\n",
       "     'liftover': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}}}}}},\n",
       " 'chromosomes': {'Chromosomes_notations': [1,\n",
       "   2,\n",
       "   3,\n",
       "   4,\n",
       "   5,\n",
       "   6,\n",
       "   7,\n",
       "   8,\n",
       "   9,\n",
       "   10,\n",
       "   11,\n",
       "   12,\n",
       "   13,\n",
       "   14,\n",
       "   15,\n",
       "   16,\n",
       "   17,\n",
       "   18,\n",
       "   19,\n",
       "   20,\n",
       "   21,\n",
       "   22,\n",
       "   23],\n",
       "  'Chromosomes_numbers': 23,\n",
       "  'chr1': 1045646,\n",
       "  'chr2': 1133896,\n",
       "  'chr3': 957670,\n",
       "  'chr4': 968009,\n",
       "  'chr5': 872864,\n",
       "  'chr6': 892038,\n",
       "  'chr7': 783273,\n",
       "  'chr8': 742846,\n",
       "  'chr9': 582304,\n",
       "  'chr10': 677958,\n",
       "  'chr11': 668174,\n",
       "  'chr12': 633728,\n",
       "  'chr13': 479581,\n",
       "  'chr14': 430810,\n",
       "  'chr15': 378044,\n",
       "  'chr16': 417320,\n",
       "  'chr17': 363449,\n",
       "  'chr18': 374023,\n",
       "  'chr19': 308988,\n",
       "  'chr20': 295761,\n",
       "  'chr21': 178155,\n",
       "  'chr22': 179766,\n",
       "  'chr23': 427164,\n",
       "  'Chromosomes_numbers percentage': 1.6676978598433364e-06,\n",
       "  'chr1 percentage': 0.0758183302762498,\n",
       "  'chr2 percentage': 0.08221721445586608,\n",
       "  'chr3 percentage': 0.06943931345374643,\n",
       "  'chr4 percentage': 0.07018897989604732,\n",
       "  'chr5 percentage': 0.06329014890149104,\n",
       "  'chr6 percentage': 0.06468042884777957,\n",
       "  'chr7 percentage': 0.05679403068578564,\n",
       "  'chr8 percentage': 0.053862725408399266,\n",
       "  'chr9 percentage': 0.042222049329487574,\n",
       "  'chr10 percentage': 0.049157787202768206,\n",
       "  'chr11 percentage': 0.04844836303491137,\n",
       "  'chr12 percentage': 0.045950731709686866,\n",
       "  'chr13 percentage': 0.034773748144414224,\n",
       "  'chr14 percentage': 0.03123743108691773,\n",
       "  'chr15 percentage': 0.02741144216202671,\n",
       "  'chr16 percentage': 0.030259290037818312,\n",
       "  'chr17 percentage': 0.026353179107052208,\n",
       "  'chr18 percentage': 0.027119885070964533,\n",
       "  'chr19 percentage': 0.022404288100750992,\n",
       "  'chr20 percentage': 0.02144521681413587,\n",
       "  'chr21 percentage': 0.012917770096538678,\n",
       "  'chr22 percentage': 0.013034581455330313,\n",
       "  'chr23 percentage': 0.03097306472183126},\n",
       " 'missing_values': {'Missing_total': 0, 'Missing_total percentage': 0.0},\n",
       " 'MAF': {'Common (MAF>=0.05)': 6949329,\n",
       "  'Low_frequency (0.01<MAF<=0.05)': 2715132,\n",
       "  'Rare (0.001<MAF<=0.01)': 3922114,\n",
       "  'Ultra Rare (MAF<=0.001)': 204895,\n",
       "  'Common (MAF>=0.05) percentage': 0.5038861348107493,\n",
       "  'Low_frequency (0.01<MAF<=0.05) percentage': 0.19687042719965903,\n",
       "  'Rare (0.001<MAF<=0.01) percentage': 0.28438700538528644,\n",
       "  'Ultra Rare (MAF<=0.001) percentage': 0.014856650130113062},\n",
       " 'variant_status': {'9960090': {'count': 10608049,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Not_palindromic_SNPs'}},\n",
       "  '9960099': {'count': 1772352,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Unchecked'}},\n",
       "  '9960399': {'count': 1254193,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized & normalized indel',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Unchecked'}},\n",
       "  '9960097': {'count': 156873,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Indistinguishable'}}},\n",
       " 'variants': {'variant_number': 13791467, 'min_minor_allele_freq': 0.0}}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1e7c7521",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>rsID&amp;SNPID</th>\n",
       "      <th>CHR&amp;POS</th>\n",
       "      <th>Stadardize&amp;Normalize</th>\n",
       "      <th>Align</th>\n",
       "      <th>Panlidromic_SNP&amp;Indel</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percentage(%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9960090</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>Not_palindromic_SNPs</td>\n",
       "      <td>10608049</td>\n",
       "      <td>76.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960097</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>Indistinguishable</td>\n",
       "      <td>156873</td>\n",
       "      <td>1.14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960099</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>1772352</td>\n",
       "      <td>12.85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960399</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized &amp; normalized indel</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>1254193</td>\n",
       "      <td>9.09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Genome_Build                  rsID&SNPID                CHR&POS  \\\n",
       "9960090    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960097    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960099    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960399    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "\n",
       "                    Stadardize&Normalize      Align Panlidromic_SNP&Indel  \\\n",
       "9960090                 standardized SNP  Unchecked  Not_palindromic_SNPs   \n",
       "9960097                 standardized SNP  Unchecked     Indistinguishable   \n",
       "9960099                 standardized SNP  Unchecked             Unchecked   \n",
       "9960399  standardized & normalized indel  Unchecked             Unchecked   \n",
       "\n",
       "            Count Percentage(%)  \n",
       "9960090  10608049         76.92  \n",
       "9960097    156873          1.14  \n",
       "9960099   1772352         12.85  \n",
       "9960399   1254193          9.09  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.lookup_status()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "df9da092",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1: 'NC_000001.10',\n",
       " 2: 'NC_000002.11',\n",
       " 3: 'NC_000003.11',\n",
       " 4: 'NC_000004.11',\n",
       " 5: 'NC_000005.9',\n",
       " 6: 'NC_000006.11',\n",
       " 7: 'NC_000007.13',\n",
       " 8: 'NC_000008.10',\n",
       " 9: 'NC_000009.11',\n",
       " 10: 'NC_000010.10',\n",
       " 11: 'NC_000011.9',\n",
       " 12: 'NC_000012.11',\n",
       " 13: 'NC_000013.10',\n",
       " 14: 'NC_000014.8',\n",
       " 15: 'NC_000015.9',\n",
       " 16: 'NC_000016.9',\n",
       " 17: 'NC_000017.10',\n",
       " 18: 'NC_000018.9',\n",
       " 19: 'NC_000019.9',\n",
       " 20: 'NC_000020.10',\n",
       " 21: 'NC_000021.8',\n",
       " 22: 'NC_000022.10',\n",
       " 23: 'NC_000023.10',\n",
       " 24: 'NC_000024.9',\n",
       " 25: 'NC_012920.1'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.get_number_to_NC(build=\"19\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
