{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c40ea94b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c8695a92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:09:08 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/14 08:09:08 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/14 08:09:08 Python version: 3.12.3 (main, Jan  8 2026, 11:30:50) [GCC 13.3.0]\n",
      "2026/01/14 08:09:08 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\n",
      "2026/01/14 08:09:33  -Reading columns          : BETA,P,CHR,POS,SNPID,EA,NEA,EAF,SE\n",
      "2026/01/14 08:09:33  -Renaming columns to      : BETA,P,CHR,POS,SNPID,EA,NEA,EAF,SE\n",
      "2026/01/14 08:09:33  -Current Dataframe shape : 13779885  x  9\n",
      "2026/01/14 08:09:34  -Initiating a status column: STATUS ...\n",
      "2026/01/14 08:09:34 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/14 08:09:37 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 08:09:37  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 08:09:37 Finished reordering the columns.\n",
      "2026/01/14 08:09:38  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2026/01/14 08:09:39  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P      \n",
      "2026/01/14 08:09:39  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64\n",
      "2026/01/14 08:09:39  -Verified: T      T     T     T        T        T      T       T       T       T      \n",
      "2026/01/14 08:09:39  -Current Dataframe memory usage: 963.05 MB\n",
      "2026/01/14 08:09:40 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\",\n",
    "    snpid=\"SNPID\",\n",
    "    chrom=\"CHR\",\n",
    "    pos=\"POS\",\n",
    "    ea=\"EA\",            # In your data, it's now EA (was ALT)\n",
    "    nea=\"NEA\",          # In your data, it's now NEA (was REF)\n",
    "    eaf=\"EAF\",          # In your data, it's now EAF (was minor_AF/Frq)\n",
    "    beta=\"BETA\",\n",
    "    se=\"SE\",\n",
    "    p=\"P\",\n",
    "    # Note: 'Dir' and 'N' are not in your column list, \n",
    "    # so we should omit them or point to the correct ones.\n",
    "    # n=\"n_complete_samples\", # Optional: if you want to use sample size\n",
    "    verbose=True,\n",
    "    readargs={'sep': '\\t'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5f255f90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['SNPID', 'CHR', 'POS', 'EA', 'NEA', 'STATUS', 'EAF', 'BETA', 'SE', 'P']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "01eb78b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:11:19 Start to download  1kg_dbsnp151_hg38_auto  ...\n",
      "2026/01/14 08:11:19  -Downloading to: /home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz\n",
      "2026/01/14 08:11:19  -File /home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz exists.\n",
      "2026/01/14 08:11:19  - Updating record in config file...\n",
      "2026/01/14 08:11:19 Downloaded  1kg_dbsnp151_hg38_auto  successfully!\n"
     ]
    }
   ],
   "source": [
    "# import gwaslab as gl\n",
    "\n",
    "# GWASLab has a built-in helper to fetch some of these\n",
    "# This will download the auto-chromosome dbSNP reference for hg38 which includes frequencies\n",
    "gl.download_ref(\"1kg_dbsnp151_hg38_auto\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b5e7ca1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:13:52 Start to check if NEA is aligned with reference sequence ...(v4.0.4)\n",
      "2026/01/14 08:13:52  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\n",
      "2026/01/14 08:14:06    -Loading and building numpy fasta records:\n",
      "2026/01/14 08:14:14  -Variants allele on given reference sequence :  0\n",
      "2026/01/14 08:14:14  -Variants flipped :  0\n",
      "2026/01/14 08:14:14   -Raw Matching rate :  0.00%\n",
      "2026/01/14 08:14:14 #WARNING! Matching rate is low, please check if the right reference genome is used.\n",
      "2026/01/14 08:14:14  -Variants inferred reverse_complement :  0\n",
      "2026/01/14 08:14:14  -Variants inferred reverse_complement_flipped :  0\n",
      "2026/01/14 08:14:14  -Both allele on genome + unable to distinguish :  0\n",
      "2026/01/14 08:14:14  -Variants not on given reference sequence :  0\n",
      "2026/01/14 08:14:14 Finished checking if NEA is aligned with reference sequence.\n",
      "2026/01/14 08:14:15 #WARNING! Reference file not found: /mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz. Using default format.\n",
      "2026/01/14 08:14:17 #WARNING! No palindromic variants available for checking.\n",
      "2026/01/14 08:14:24 #WARNING! No indistinguishable indels available for checking.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7243ba757d70>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "ref_fasta = \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\"\n",
    "\n",
    "\n",
    "ref_vcf = \"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz\"\n",
    "\n",
    "# 3. Perform Harmonization\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=ref_fasta,     \n",
    "    ref_infer=ref_vcf,      \n",
    "    ref_alt_freq=\"AF\",      \n",
    "    threads=6,             \n",
    "    verbose=False,          \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "26e84088",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      SNPID  CHR        POS EA  NEA   STATUS       EAF  \\\n",
       "0               1:69487:G:A    1      69487  A    G  9960090  0.999994   \n",
       "1               1:69569:T:C    1      69569  C    T  9960090  0.999812   \n",
       "2              1:139853:C:T    1     139853  T    C  9960090  0.999994   \n",
       "3             1:692794:CA:C    1     692794  C   CA  9960399  0.889410   \n",
       "4              1:693731:A:G    1     693731  G    A  9960090  0.884233   \n",
       "...                     ...  ...        ... ..  ...      ...       ...   \n",
       "13779880    X:154929412:C:T   23  154929412  T    C  9960090  0.754527   \n",
       "13779881   X:154929637:CT:C   23  154929637  C   CT  9960399  0.770274   \n",
       "13779882  X:154929952:CAA:C   23  154929952  C  CAA  9960399  0.760570   \n",
       "13779883    X:154930230:A:G   23  154930230  G    A  9960090  0.754113   \n",
       "13779884    X:154930487:T:A   23  154930487  A    T  9960099  0.980441   \n",
       "\n",
       "              BETA        SE         P  \n",
       "0        -2.715450  2.360060  0.249902  \n",
       "1        -0.484284  0.423462  0.252778  \n",
       "2        -2.703560  2.360130  0.251997  \n",
       "3        -0.016436  0.019585  0.401342  \n",
       "4        -0.004255  0.018507  0.818155  \n",
       "...            ...       ...       ...  \n",
       "13779880 -0.016260  0.010723  0.129427  \n",
       "13779881 -0.027098  0.011190  0.015456  \n",
       "13779882 -0.020494  0.011278  0.069202  \n",
       "13779883 -0.016347  0.010721  0.127334  \n",
       "13779884  0.007107  0.036783  0.846790  \n",
       "\n",
       "[13779885 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e926429",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "69c39476",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:19:18 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/14 08:19:18  -Number of threads/cores to use: 1\n",
      "2026/01/14 08:19:18  -No available variants to normalize..\n",
      "2026/01/14 08:19:18 Finished normalizing indels.\n",
      "2026/01/14 08:19:18 Start to remove duplicated/multiallelic variants ...(v4.0.4)\n",
      "2026/01/14 08:19:18  -Removing mode:dm\n",
      "2026/01/14 08:19:18 Start to sort the sumstats using P ...\n",
      "2026/01/14 08:19:25 Start to remove duplicated variants based on snpid ...(v4.0.4)\n",
      "2026/01/14 08:19:25  -Which variant to keep:  first\n",
      "2026/01/14 08:19:47  -Removed variants based on SNPID: 0\n",
      "2026/01/14 08:19:47 Start to remove duplicated variants based on CHR,POS,EA and NEA ...\n",
      "2026/01/14 08:19:47  -Which variant to keep:  first\n",
      "2026/01/14 08:20:03  -Removed variants based on CHR,POS,EA and NEA: 978\n",
      "2026/01/14 08:20:03 Start to remove multiallelic variants based on chr:pos ...\n",
      "2026/01/14 08:20:03  -Current Dataframe shape : 13778907 x 10 ; Memory usage: 1043.31 MB\n",
      "2026/01/14 08:20:03  -Which variant to keep:  first\n",
      "2026/01/14 08:20:19  -Removed variants multiallelic variants: 17251\n",
      "2026/01/14 08:20:19  -Removed variants in total: 18229\n",
      "2026/01/14 08:20:19  -Sort the coordinates based on CHR and POS...\n",
      "2026/01/14 08:20:39 Finished removing duplicated/multiallelic variants.\n",
      "2026/01/14 08:20:39 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/14 08:20:39  -Current Dataframe shape : 13761656 x 10 ; Memory usage: 937.02 MB\n",
      "2026/01/14 08:20:39  -Checking SNPID data type...\n",
      "2026/01/14 08:20:39  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/14 08:20:40  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/14 08:20:50 Finished checking SNPID/rsID.\n",
      "2026/01/14 08:20:50 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/14 08:20:50  -Checking CHR data type...\n",
      "2026/01/14 08:20:55  -Variants with standardized chromosome notation: 13761656\n",
      "2026/01/14 08:20:55  -All CHR are already fixed...\n",
      "2026/01/14 08:20:58 Finished fixing chromosome notation (CHR).\n",
      "2026/01/14 08:20:58 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/14 08:20:58  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/14 08:20:59  -Position bound:(0 , 250,000,000)\n",
      "2026/01/14 08:20:59  -No outlier variants were removed.\n",
      "2026/01/14 08:21:00  -Removed variants with bad positions: 0\n",
      "2026/01/14 08:21:00 Finished fixing basepair positions (POS).\n",
      "2026/01/14 08:21:01 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/14 08:21:01  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/14 08:21:05  -Variants with bad EA: 0\n",
      "2026/01/14 08:21:05  -Variants with bad NEA: 0\n",
      "2026/01/14 08:21:05  -Variants with NA for EA or NEA: 0\n",
      "2026/01/14 08:21:05  -Variants with same EA and NEA: 1948473\n",
      "2026/01/14 08:21:20 Finished fixing alleles (EA and NEA).\n",
      "2026/01/14 08:21:20 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/14 08:21:20  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/14 08:21:20  -Checking if any columns are empty...\n",
      "2026/01/14 08:21:21  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/14 08:21:22  -Checking if -1e-07 < P < 1.0000001 : removed 0 variants.\n",
      "2026/01/14 08:21:23  -Checking if -100.0000001 < BETA < 100.0000001 : removed 0 variants.\n",
      "2026/01/14 08:21:24  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/14 08:21:24  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/14 08:21:24  -No variants were removed with bad statistics in total.\n",
      "2026/01/14 08:21:24 Finished sanity check for statistics.\n",
      "2026/01/14 08:21:24 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/14 08:21:24  -Number of threads/cores to use: 6\n",
      "2026/01/14 08:21:24  -No available variants to normalize..\n",
      "2026/01/14 08:21:24 Finished normalizing indels.\n",
      "2026/01/14 08:21:24 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 08:21:24  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 08:21:25 Finished reordering the columns.\n",
      "2026/01/14 08:21:25 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/14 08:21:39 Finished sorting coordinates.\n",
      "2026/01/14 08:21:39 Start to check if NEA is aligned with reference sequence ...(v4.0.4)\n",
      "2026/01/14 08:21:39  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\n",
      "2026/01/14 08:21:52    -Loading and building numpy fasta records:\n",
      "2026/01/14 08:22:01  -Variants allele on given reference sequence :  0\n",
      "2026/01/14 08:22:01  -Variants flipped :  0\n",
      "2026/01/14 08:22:01   -Raw Matching rate :  0.00%\n",
      "2026/01/14 08:22:01 #WARNING! Matching rate is low, please check if the right reference genome is used.\n",
      "2026/01/14 08:22:01  -Variants inferred reverse_complement :  0\n",
      "2026/01/14 08:22:01  -Variants inferred reverse_complement_flipped :  0\n",
      "2026/01/14 08:22:01  -Both allele on genome + unable to distinguish :  0\n",
      "2026/01/14 08:22:01  -Variants not on given reference sequence :  0\n",
      "2026/01/14 08:22:01 Finished checking if NEA is aligned with reference sequence.\n",
      "2026/01/14 08:22:01 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/14 08:22:02  -No statistics have been changed.\n",
      "2026/01/14 08:22:02 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/14 08:22:02 Start to infer strand for palindromic SNPs/align indistinguishable indels ...(v4.0.4)\n",
      "2026/01/14 08:22:02  -Number of threads/cores to use: 6\n",
      "2026/01/14 08:22:02  -Field for alternative allele frequency in VCF INFO: AF\n",
      "2026/01/14 08:22:03  -Identified  1658027  palindromic SNPs...\n",
      "2026/01/14 08:22:04  -After filtering by MAF<= 0.45 , 0 palindromic SNPs with unknown strand will be inferred...\n",
      "2026/01/14 08:22:04 #WARNING! No palindromic variants available for checking.\n",
      "2026/01/14 08:22:05   -Non-palindromic :  9162782\n",
      "2026/01/14 08:22:06   -Palindromic SNPs on + strand:  0\n",
      "2026/01/14 08:22:08   -Palindromic SNPs on - strand and needed to be flipped: 0\n",
      "2026/01/14 08:22:09   -Palindromic SNPs with MAF not available to infer :  86036\n",
      "2026/01/14 08:22:10   -Palindromic SNPs lacking reference matches, or with reference ALT unavailable or above MAF threshold for inference :  0\n",
      "2026/01/14 08:22:11  -Identified  0  indistinguishable Indels...\n",
      "2026/01/14 08:22:11 #WARNING! No indistinguishable indels available for checking.\n",
      "2026/01/14 08:22:11 Finished inferring strand for palindromic SNPs/align indistinguishable indels.\n",
      "2026/01/14 08:22:11 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/14 08:22:12  -No statistics have been changed.\n",
      "2026/01/14 08:22:12 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/14 08:22:12 Start to remove duplicated/multiallelic variants ...(v4.0.4)\n",
      "2026/01/14 08:22:12  -Removing mode:dm\n",
      "2026/01/14 08:22:12 Start to sort the sumstats using P ...\n",
      "2026/01/14 08:22:19 Start to remove duplicated variants based on snpid ...(v4.0.4)\n",
      "2026/01/14 08:22:19  -Which variant to keep:  first\n",
      "2026/01/14 08:22:40  -Removed variants based on SNPID: 0\n",
      "2026/01/14 08:22:40 Start to remove duplicated variants based on CHR,POS,EA and NEA ...\n",
      "2026/01/14 08:22:40  -Which variant to keep:  first\n",
      "2026/01/14 08:22:56  -Removed variants based on CHR,POS,EA and NEA: 0\n",
      "2026/01/14 08:22:56 Start to remove multiallelic variants based on chr:pos ...\n",
      "2026/01/14 08:22:56  -Which variant to keep:  first\n",
      "2026/01/14 08:23:11  -Removed variants multiallelic variants: 0\n",
      "2026/01/14 08:23:11  -Removed variants in total: 0\n",
      "2026/01/14 08:23:11  -Sort the coordinates based on CHR and POS...\n",
      "2026/01/14 08:23:32 Finished removing duplicated/multiallelic variants.\n",
      "2026/01/14 08:23:32 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/14 08:23:45 Finished sorting coordinates.\n",
      "2026/01/14 08:23:45 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 08:23:45  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 08:23:45 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761651</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761652</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761653</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761654</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761655</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13761656 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7243ba757d70>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Normalize Alleles (Fixes the 'invalid allele notation' for the 1.9M variants)\n",
    "mysumstats.normalize_allele()\n",
    "\n",
    "# 2. Remove Duplicates (Essential before publication)\n",
    "mysumstats.remove_dup()\n",
    "\n",
    "# 3. Final Harmonization Sweep\n",
    "# We use a slightly more relaxed MAF threshold to capture more variants\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=\"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\",\n",
    "    ref_infer=\"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz\",\n",
    "    ref_alt_freq=\"AF\",\n",
    "    maf_threshold=0.45, \n",
    "    threads=6,\n",
    "    remove=True  # This will remove variants that fundamentally mismatch hg38\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e50fa02a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 09:18:45 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/14 09:18:45  -Current Dataframe shape : 12629916 x 10 ; Memory usage: 956.74 MB\n",
      "2026/01/14 09:18:45  -Checking SNPID data type...\n",
      "2026/01/14 09:18:45  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/14 09:18:46  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/14 09:18:55 Finished checking SNPID/rsID.\n",
      "2026/01/14 09:18:55 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/14 09:18:55  -Checking CHR data type...\n",
      "2026/01/14 09:18:59  -Variants with standardized chromosome notation: 12629916\n",
      "2026/01/14 09:18:59  -All CHR are already fixed...\n",
      "2026/01/14 09:19:02 Finished fixing chromosome notation (CHR).\n",
      "2026/01/14 09:19:02 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/14 09:19:02  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/14 09:19:03  -Position bound:(0 , 250,000,000)\n",
      "2026/01/14 09:19:03  -No outlier variants were removed.\n",
      "2026/01/14 09:19:04  -Removed variants with bad positions: 0\n",
      "2026/01/14 09:19:04 Finished fixing basepair positions (POS).\n",
      "2026/01/14 09:19:04 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/14 09:19:04  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/14 09:19:09  -Variants with bad EA: 0\n",
      "2026/01/14 09:19:09  -Variants with bad NEA: 0\n",
      "2026/01/14 09:19:09  -Variants with NA for EA or NEA: 0\n",
      "2026/01/14 09:19:09  -Variants with same EA and NEA: 1809107\n",
      "2026/01/14 09:19:22 Finished fixing alleles (EA and NEA).\n",
      "2026/01/14 09:19:22 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/14 09:19:22  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/14 09:19:23  -Checking if any columns are empty...\n",
      "2026/01/14 09:19:23  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/14 09:19:24  -Checking if -1e-07 < P < 1.0000001 : removed 0 variants.\n",
      "2026/01/14 09:19:25  -Checking if -100.0000001 < BETA < 100.0000001 : removed 0 variants.\n",
      "2026/01/14 09:19:26  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/14 09:19:26  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/14 09:19:26  -No variants were removed with bad statistics in total.\n",
      "2026/01/14 09:19:26 Finished sanity check for statistics.\n",
      "2026/01/14 09:19:26 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/14 09:19:26  -Number of threads/cores to use: 6\n",
      "2026/01/14 09:19:27  -No available variants to normalize..\n",
      "2026/01/14 09:19:27 Finished normalizing indels.\n",
      "2026/01/14 09:19:27 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 09:19:27  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 09:19:27 Finished reordering the columns.\n",
      "2026/01/14 09:19:27 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/14 09:19:38 Finished sorting coordinates.\n",
      "2026/01/14 09:19:38 Start to check if NEA is aligned with reference sequence ...(v4.0.4)\n",
      "2026/01/14 09:19:38  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\n",
      "2026/01/14 09:19:49    -Loading and building numpy fasta records:\n",
      "2026/01/14 09:19:56  -Variants allele on given reference sequence :  0\n",
      "2026/01/14 09:19:56  -Variants flipped :  0\n",
      "2026/01/14 09:19:56   -Raw Matching rate :  0.00%\n",
      "2026/01/14 09:19:56 #WARNING! Matching rate is low, please check if the right reference genome is used.\n",
      "2026/01/14 09:19:56  -Variants inferred reverse_complement :  0\n",
      "2026/01/14 09:19:56  -Variants inferred reverse_complement_flipped :  0\n",
      "2026/01/14 09:19:56  -Both allele on genome + unable to distinguish :  0\n",
      "2026/01/14 09:19:56  -Variants not on given reference sequence :  0\n",
      "2026/01/14 09:19:56 Finished checking if NEA is aligned with reference sequence.\n",
      "2026/01/14 09:19:56 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/14 09:19:57  -No statistics have been changed.\n",
      "2026/01/14 09:19:57 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/14 09:19:57 Start to infer strand for palindromic SNPs/align indistinguishable indels ...(v4.0.4)\n",
      "2026/01/14 09:19:57  -Number of threads/cores to use: 6\n",
      "2026/01/14 09:19:57  -Field for alternative allele frequency in VCF INFO: AF\n",
      "2026/01/14 09:19:58  -Identified  1658027  palindromic SNPs...\n",
      "2026/01/14 09:19:59  -After filtering by MAF<= 0.4 , 0 palindromic SNPs with unknown strand will be inferred...\n",
      "2026/01/14 09:19:59 #WARNING! No palindromic variants available for checking.\n",
      "2026/01/14 09:20:00   -Non-palindromic :  9162782\n",
      "2026/01/14 09:20:01   -Palindromic SNPs on + strand:  0\n",
      "2026/01/14 09:20:02   -Palindromic SNPs on - strand and needed to be flipped: 0\n",
      "2026/01/14 09:20:04   -Palindromic SNPs with MAF not available to infer :  86036\n",
      "2026/01/14 09:20:05   -Palindromic SNPs lacking reference matches, or with reference ALT unavailable or above MAF threshold for inference :  0\n",
      "2026/01/14 09:20:05  -Identified  0  indistinguishable Indels...\n",
      "2026/01/14 09:20:05 #WARNING! No indistinguishable indels available for checking.\n",
      "2026/01/14 09:20:05 Finished inferring strand for palindromic SNPs/align indistinguishable indels.\n",
      "2026/01/14 09:20:05 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/14 09:20:06  -No statistics have been changed.\n",
      "2026/01/14 09:20:06 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/14 09:20:06 Start to remove duplicated/multiallelic variants ...(v4.0.4)\n",
      "2026/01/14 09:20:06  -Removing mode:dm\n",
      "2026/01/14 09:20:06 Start to sort the sumstats using P ...\n",
      "2026/01/14 09:20:12 Start to remove duplicated variants based on snpid ...(v4.0.4)\n",
      "2026/01/14 09:20:12  -Which variant to keep:  first\n",
      "2026/01/14 09:20:29  -Removed variants based on SNPID: 0\n",
      "2026/01/14 09:20:29 Start to remove duplicated variants based on CHR,POS,EA and NEA ...\n",
      "2026/01/14 09:20:29  -Which variant to keep:  first\n",
      "2026/01/14 09:20:41  -Removed variants based on CHR,POS,EA and NEA: 0\n",
      "2026/01/14 09:20:41 Start to remove multiallelic variants based on chr:pos ...\n",
      "2026/01/14 09:20:41  -Which variant to keep:  first\n",
      "2026/01/14 09:20:53  -Removed variants multiallelic variants: 0\n",
      "2026/01/14 09:20:53  -Removed variants in total: 0\n",
      "2026/01/14 09:20:53  -Sort the coordinates based on CHR and POS...\n",
      "2026/01/14 09:21:10 Finished removing duplicated/multiallelic variants.\n",
      "2026/01/14 09:21:10 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/14 09:21:20 Finished sorting coordinates.\n",
      "2026/01/14 09:21:20 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 09:21:20  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 09:21:21 Finished reordering the columns.\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'Sumstats' object has no attribute 'check_row'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/g_Sumstats.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self, name)\u001b[39m\n\u001b[32m    514\u001b[39m             \u001b[38;5;28;01mreturn\u001b[39;00m getattr(data, name)\n\u001b[32m    515\u001b[39m         \u001b[38;5;28;01mexcept\u001b[39;00m AttributeError:\n\u001b[32m--> \u001b[39m\u001b[32m516\u001b[39m             \u001b[38;5;28;01mraise\u001b[39;00m AttributeError(f\"'{type(self).__name__}' object has no attribute '{name}'\")\n",
      "\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/core/generic.py\u001b[39m in \u001b[36m?\u001b[39m\u001b[34m(self, name)\u001b[39m\n\u001b[32m   6319\u001b[39m         ):\n\u001b[32m   6320\u001b[39m             \u001b[38;5;28;01mreturn\u001b[39;00m self[name]\n\u001b[32m-> \u001b[39m\u001b[32m6321\u001b[39m         \u001b[38;5;28;01mreturn\u001b[39;00m object.__getattribute__(self, name)\n",
      "\u001b[31mAttributeError\u001b[39m: 'DataFrame' object has no attribute 'check_row'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[23]\u001b[39m\u001b[32m, line 23\u001b[39m\n\u001b[32m     14\u001b[39m mysumstats.harmonize(\n\u001b[32m     15\u001b[39m     ref_seq=\u001b[33m\"\u001b[39m\u001b[33m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     16\u001b[39m     ref_infer=\u001b[33m\"\u001b[39m\u001b[33m/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr\u001b[39m\u001b[38;5;132;01m{chrom}\u001b[39;00m\u001b[33m.vcf.gz\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m   (...)\u001b[39m\u001b[32m     19\u001b[39m     remove=\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[32m     20\u001b[39m )\n\u001b[32m     22\u001b[39m \u001b[38;5;66;03m# 4. Update the internal status tracker\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m23\u001b[39m \u001b[43mmysumstats\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcheck_row\u001b[49m()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/g_Sumstats.py:516\u001b[39m, in \u001b[36mSumstats.__getattr__\u001b[39m\u001b[34m(self, name)\u001b[39m\n\u001b[32m    514\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(data, name)\n\u001b[32m    515\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m516\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mAttributeError\u001b[39;00m(\u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(\u001b[38;5;28mself\u001b[39m).\u001b[34m__name__\u001b[39m\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m object has no attribute \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m\"\u001b[39m)\n",
      "\u001b[31mAttributeError\u001b[39m: 'Sumstats' object has no attribute 'check_row'"
     ]
    }
   ],
   "source": [
    "# 1. Manually clean and filter the dataframe\n",
    "# This targets the 1.9M variants directly\n",
    "valid_bases = {'A', 'T', 'C', 'G'}\n",
    "mysumstats.data = mysumstats.data[\n",
    "    mysumstats.data[\"EA\"].isin(valid_bases) & \n",
    "    mysumstats.data[\"NEA\"].isin(valid_bases)\n",
    "].copy()\n",
    "\n",
    "# 2. Hard-set the build metadata so GWASLab stops saying \"Unchecked\"\n",
    "mysumstats.data.attrs[\"build\"] = \"38\"\n",
    "\n",
    "# 3. Re-run harmonization with 'remove=True'\n",
    "# This will align the remaining clean SNPs to your hg38 reference VCFs\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=\"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\",\n",
    "    ref_infer=\"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz\",\n",
    "    ref_alt_freq=\"AF\",\n",
    "    threads=6,\n",
    "    remove=True\n",
    ")\n",
    "\n",
    "# 4. Update the internal status tracker\n",
    "mysumstats.check_row()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85fa4d5e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad801b57",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99dd956d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "1cd3e48e",
   "metadata": {},
   "source": [
    "### Check summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "742ac24e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'overview': {'Row_num': '13761656',\n",
       "  'Column_num': '10',\n",
       "  'Column_names': 'SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P',\n",
       "  'Last_checked_time': 'Wed Jan 14 09:17:42 2026',\n",
       "  'QC and Harmonization': {'basic_check': {'performed': True,\n",
       "    'last_executed': '2026-01-14 08:54:08',\n",
       "    'parameters_used': {'basic_check': True,\n",
       "     'ref_seq': '/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa',\n",
       "     'ref_rsid_tsv': None,\n",
       "     'ref_rsid_vcf': None,\n",
       "     'ref_infer': '/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz',\n",
       "     'ref_alt_freq': 'AF',\n",
       "     'ref_maf_threshold': 0.4,\n",
       "     'maf_threshold': 0.4,\n",
       "     'threads': 6,\n",
       "     'remove': True,\n",
       "     'check_ref_kwargs': {},\n",
       "     'remove_dup_kwargs': {},\n",
       "     'assign_rsid_kwargs': {},\n",
       "     'infer_strand_kwargs': {},\n",
       "     'flip_allele_stats_kwargs': {},\n",
       "     'liftover_kwargs': {},\n",
       "     'fix_id_kwargs': {},\n",
       "     'fix_chr_kwargs': {},\n",
       "     'fix_pos_kwargs': {},\n",
       "     'fix_allele_kwargs': {},\n",
       "     'sanity_check_stats_kwargs': {},\n",
       "     'normalize_allele_kwargs': {},\n",
       "     'verbose': True,\n",
       "     'sweep_mode': False}},\n",
       "   'harmonize': {'performed': True,\n",
       "    'last_executed': '2026-01-14 08:56:15',\n",
       "    'parameters_used': {'basic_check': True,\n",
       "     'ref_seq': '/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa',\n",
       "     'ref_rsid_tsv': None,\n",
       "     'ref_rsid_vcf': None,\n",
       "     'ref_infer': '/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz',\n",
       "     'ref_alt_freq': 'AF',\n",
       "     'ref_maf_threshold': 0.4,\n",
       "     'maf_threshold': 0.4,\n",
       "     'threads': 6,\n",
       "     'remove': True,\n",
       "     'check_ref_kwargs': {},\n",
       "     'remove_dup_kwargs': {},\n",
       "     'assign_rsid_kwargs': {},\n",
       "     'infer_strand_kwargs': {},\n",
       "     'flip_allele_stats_kwargs': {},\n",
       "     'liftover_kwargs': {},\n",
       "     'fix_id_kwargs': {},\n",
       "     'fix_chr_kwargs': {},\n",
       "     'fix_pos_kwargs': {},\n",
       "     'fix_allele_kwargs': {},\n",
       "     'sanity_check_stats_kwargs': {},\n",
       "     'normalize_allele_kwargs': {},\n",
       "     'verbose': True,\n",
       "     'sweep_mode': False}},\n",
       "   'qc_and_harmonization_status': {'qc': {'id': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:54:08',\n",
       "      'parameters_used': {}},\n",
       "     'chr': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:54:08',\n",
       "      'parameters_used': {}},\n",
       "     'pos': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:54:08',\n",
       "      'parameters_used': {}},\n",
       "     'allele': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:54:08',\n",
       "      'parameters_used': {}},\n",
       "     'sanity': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:54:08',\n",
       "      'parameters_used': {}},\n",
       "     'consistency': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:54:08',\n",
       "      'parameters_used': {}},\n",
       "     'normalize': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}},\n",
       "     'remove_dup': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:56:00',\n",
       "      'parameters_used': {'mode': 'dm',\n",
       "       'chrom': 'CHR',\n",
       "       'pos': 'POS',\n",
       "       'snpid': 'SNPID',\n",
       "       'ea': 'EA',\n",
       "       'nea': 'NEA',\n",
       "       'rsid': 'rsID',\n",
       "       'keep': 'first',\n",
       "       'keep_col': 'P',\n",
       "       'remove_na': False,\n",
       "       'keep_ascend': True}},\n",
       "     'sort_coord': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:56:14',\n",
       "      'parameters_used': {'chrom': 'CHR', 'pos': 'POS', 'reindex': True}},\n",
       "     'sort_column': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:56:14',\n",
       "      'parameters_used': {'order': ['SNPID',\n",
       "        'rsID',\n",
       "        'CHR',\n",
       "        'POS',\n",
       "        'EA',\n",
       "        'NEA',\n",
       "        'STATUS',\n",
       "        'REF',\n",
       "        'ALT',\n",
       "        'EAF',\n",
       "        'RAF',\n",
       "        'NEAF',\n",
       "        'MAF',\n",
       "        'DAF',\n",
       "        'INFO',\n",
       "        'BETA',\n",
       "        'SE',\n",
       "        'BETA_95U',\n",
       "        'BETA_95L',\n",
       "        'OR',\n",
       "        'OR_95U',\n",
       "        'OR_95L',\n",
       "        'HR',\n",
       "        'HR_95U',\n",
       "        'HR_95L',\n",
       "        'CHISQ',\n",
       "        'Z',\n",
       "        'T',\n",
       "        'F',\n",
       "        'P',\n",
       "        'P_MANTISSA',\n",
       "        'P_EXPONENT',\n",
       "        'MLOG10P',\n",
       "        'SNPR2',\n",
       "        'DOF',\n",
       "        'P_HET',\n",
       "        'I2_HET',\n",
       "        'I2',\n",
       "        'DIRECTION',\n",
       "        'DENSITY',\n",
       "        'N',\n",
       "        'N_CASE',\n",
       "        'N_CONTROL',\n",
       "        'N_EFF',\n",
       "        'GENENAME',\n",
       "        'CIS/TRANS',\n",
       "        'DISTANCE_TO_KNOWN',\n",
       "        'LOCATION_OF_KNOWN',\n",
       "        'KNOWN_ID',\n",
       "        'KNOWN_PUBMED_ID',\n",
       "        'KNOWN_AUTHOR',\n",
       "        'KNOWN_SET_VARIANT',\n",
       "        'KNOWN_VARIANT',\n",
       "        'KNOWN_SET',\n",
       "        'NOVEL',\n",
       "        'PIP',\n",
       "        'CREDIBLE_SET_INDEX',\n",
       "        'N_SNP',\n",
       "        'LOCUS',\n",
       "        'STUDY',\n",
       "        'TEST',\n",
       "        'BETA_RANDOM',\n",
       "        'SE_RANDOM',\n",
       "        'Z_RANDOM',\n",
       "        'P_RANDOM']}}},\n",
       "    'harmonize': {'check_ref': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:56:15',\n",
       "      'parameters_used': {}},\n",
       "     'flip_allele_stats': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:56:15',\n",
       "      'parameters_used': {}},\n",
       "     'infer_strand': {'performed': True,\n",
       "      'last_executed': '2026-01-14 08:56:15',\n",
       "      'parameters_used': {}},\n",
       "     'assign_rsid': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}},\n",
       "     'liftover': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}}}}}},\n",
       " 'chromosomes': {'Chromosomes_notations': [1,\n",
       "   2,\n",
       "   3,\n",
       "   4,\n",
       "   5,\n",
       "   6,\n",
       "   7,\n",
       "   8,\n",
       "   9,\n",
       "   10,\n",
       "   11,\n",
       "   12,\n",
       "   13,\n",
       "   14,\n",
       "   15,\n",
       "   16,\n",
       "   17,\n",
       "   18,\n",
       "   19,\n",
       "   20,\n",
       "   21,\n",
       "   22,\n",
       "   23],\n",
       "  'Chromosomes_numbers': 23,\n",
       "  'chr1': 1043179,\n",
       "  'chr2': 1131789,\n",
       "  'chr3': 955880,\n",
       "  'chr4': 966616,\n",
       "  'chr5': 871384,\n",
       "  'chr6': 890235,\n",
       "  'chr7': 781804,\n",
       "  'chr8': 741629,\n",
       "  'chr9': 580661,\n",
       "  'chr10': 676289,\n",
       "  'chr11': 666091,\n",
       "  'chr12': 632100,\n",
       "  'chr13': 478652,\n",
       "  'chr14': 429726,\n",
       "  'chr15': 376996,\n",
       "  'chr16': 416237,\n",
       "  'chr17': 362194,\n",
       "  'chr18': 373482,\n",
       "  'chr19': 307896,\n",
       "  'chr20': 295118,\n",
       "  'chr21': 177826,\n",
       "  'chr22': 179247,\n",
       "  'chr23': 426625,\n",
       "  'Chromosomes_numbers percentage': 1.6713104876331743e-06,\n",
       "  'chr1 percentage': 0.07580330448602988,\n",
       "  'chr2 percentage': 0.08224220980382012,\n",
       "  'chr3 percentage': 0.06945966386603472,\n",
       "  'chr4 percentage': 0.07023980253539254,\n",
       "  'chr5 percentage': 0.06331970512851069,\n",
       "  'chr6 percentage': 0.06468952573730952,\n",
       "  'chr7 percentage': 0.05681031410754636,\n",
       "  'chr8 percentage': 0.05389097067969145,\n",
       "  'chr9 percentage': 0.042194122567807246,\n",
       "  'chr10 percentage': 0.04914299558134573,\n",
       "  'chr11 percentage': 0.04840195104426386,\n",
       "  'chr12 percentage': 0.045931972140562154,\n",
       "  'chr13 percentage': 0.03478156989246062,\n",
       "  'chr14 percentage': 0.031226329156897978,\n",
       "  'chr15 percentage': 0.027394668199815488,\n",
       "  'chr16 percentage': 0.03024614188873781,\n",
       "  'chr17 percentage': 0.026319070902513477,\n",
       "  'chr18 percentage': 0.027139321023574487,\n",
       "  'chr19 percentage': 0.022373470169578427,\n",
       "  'chr20 percentage': 0.02144494819518814,\n",
       "  'chr21 percentage': 0.012921846033645951,\n",
       "  'chr22 percentage': 0.01302510395551233,\n",
       "  'chr23 percentage': 0.031000992903761},\n",
       " 'missing_values': {'Missing_total': 0, 'Missing_total percentage': 0.0},\n",
       " 'MAF': {'Common (MAF>=0.05)': 6939563,\n",
       "  'Low_frequency (0.01<MAF<=0.05)': 2711246,\n",
       "  'Rare (0.001<MAF<=0.01)': 3917773,\n",
       "  'Ultra Rare (MAF<=0.001)': 193074,\n",
       "  'Common (MAF>=0.05) percentage': 0.5042680183257015,\n",
       "  'Low_frequency (0.01<MAF<=0.05) percentage': 0.19701451627623884,\n",
       "  'Rare (0.001<MAF<=0.01) percentage': 0.28468761317678626,\n",
       "  'Ultra Rare (MAF<=0.001) percentage': 0.01402985222127337},\n",
       " 'p_values': {'Minimum': 5.4652e-220,\n",
       "  'P<5e-8': 51762,\n",
       "  'P<5e-6': 114723,\n",
       "  'P<5e-8 percentage': 0.0037613205852551467,\n",
       "  'P<5e-6 percentage': 0.008336424046640898},\n",
       " 'beta_for_significant_variants': {'P<5e-8 with ABS(BETA)>10': 11,\n",
       "  'P<5e-8 with ABS(BETA)>3': 11,\n",
       "  'P<5e-8 with ABS(BETA)>1': 11,\n",
       "  'P<5e-8 with ABS(BETA)>0.5': 70,\n",
       "  'P<5e-8 with ABS(BETA)>0.3': 269,\n",
       "  'P<5e-8 with ABS(BETA)>0.2': 825,\n",
       "  'P<5e-8 with ABS(BETA)>0.1': 11262,\n",
       "  'P<5e-8 with ABS(BETA)>0.05': 51762,\n",
       "  'P<5e-8 with ABS(BETA)>10 percentage': 7.993224071289095e-07,\n",
       "  'P<5e-8 with ABS(BETA)>3 percentage': 7.993224071289095e-07,\n",
       "  'P<5e-8 with ABS(BETA)>1 percentage': 7.993224071289095e-07,\n",
       "  'P<5e-8 with ABS(BETA)>0.5 percentage': 5.086597136274879e-06,\n",
       "  'P<5e-8 with ABS(BETA)>0.3 percentage': 1.9547066137970603e-05,\n",
       "  'P<5e-8 with ABS(BETA)>0.2 percentage': 5.9949180534668213e-05,\n",
       "  'P<5e-8 with ABS(BETA)>0.1 percentage': 0.0008183608135532526,\n",
       "  'P<5e-8 with ABS(BETA)>0.05 percentage': 0.0037613205852551467},\n",
       " 'variant_status': {'9960090': {'count': 9162782,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Not_palindromic_SNPs'}},\n",
       "  '9960699': {'count': 1948473,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'invalid allele notation',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Unchecked'}},\n",
       "  '9960099': {'count': 1571991,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Unchecked'}},\n",
       "  '9960399': {'count': 992374,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized & normalized indel',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Unchecked'}},\n",
       "  '9960097': {'count': 86036,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Indistinguishable'}}},\n",
       " 'variants': {'variant_number': 13761656,\n",
       "  'min_P': 5.4652e-220,\n",
       "  'min_minor_allele_freq': 0.0}}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "4081dbcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:26:15 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/14 08:26:15  -Number of threads/cores to use: 1\n",
      "2026/01/14 08:26:15  -No available variants to normalize..\n",
      "2026/01/14 08:26:15 Finished normalizing indels.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761651</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761652</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761653</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761654</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761655</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13761656 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7243ba757d70>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Force allele standardization\n",
    "mysumstats.normalize_allele()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3147a107",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 08:26:33 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/14 08:26:33  -Checking SNPID data type...\n",
      "2026/01/14 08:26:33  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/14 08:26:34  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/14 08:26:44 Finished checking SNPID/rsID.\n",
      "2026/01/14 08:26:44 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/14 08:26:44  -Checking CHR data type...\n",
      "2026/01/14 08:26:49  -Variants with standardized chromosome notation: 13761656\n",
      "2026/01/14 08:26:49  -All CHR are already fixed...\n",
      "2026/01/14 08:26:52 Finished fixing chromosome notation (CHR).\n",
      "2026/01/14 08:26:52 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/14 08:26:52  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/14 08:26:53  -Position bound:(0 , 250,000,000)\n",
      "2026/01/14 08:26:53  -No outlier variants were removed.\n",
      "2026/01/14 08:26:54  -Removed variants with bad positions: 0\n",
      "2026/01/14 08:26:54 Finished fixing basepair positions (POS).\n",
      "2026/01/14 08:26:54 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/14 08:26:54  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/14 08:26:59  -Variants with bad EA: 0\n",
      "2026/01/14 08:26:59  -Variants with bad NEA: 0\n",
      "2026/01/14 08:26:59  -Variants with NA for EA or NEA: 0\n",
      "2026/01/14 08:26:59  -Variants with same EA and NEA: 1948473\n",
      "2026/01/14 08:27:14 Finished fixing alleles (EA and NEA).\n",
      "2026/01/14 08:27:14 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/14 08:27:14  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/14 08:27:14  -Checking if any columns are empty...\n",
      "2026/01/14 08:27:15  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/14 08:27:16  -Checking if -1e-07 < P < 1.0000001 : removed 0 variants.\n",
      "2026/01/14 08:27:17  -Checking if -100.0000001 < BETA < 100.0000001 : removed 0 variants.\n",
      "2026/01/14 08:27:18  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/14 08:27:18  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/14 08:27:18  -No variants were removed with bad statistics in total.\n",
      "2026/01/14 08:27:18 Finished sanity check for statistics.\n",
      "2026/01/14 08:27:18 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/14 08:27:18  -Number of threads/cores to use: 6\n",
      "2026/01/14 08:27:18  -No available variants to normalize..\n",
      "2026/01/14 08:27:18 Finished normalizing indels.\n",
      "2026/01/14 08:27:18 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 08:27:18  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 08:27:19 Finished reordering the columns.\n",
      "2026/01/14 08:27:19 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/14 08:27:32 Finished sorting coordinates.\n",
      "2026/01/14 08:27:32 Start to check if NEA is aligned with reference sequence ...(v4.0.4)\n",
      "2026/01/14 08:27:33  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\n",
      "2026/01/14 08:27:46    -Loading and building numpy fasta records:\n",
      "2026/01/14 08:27:54  -Variants allele on given reference sequence :  0\n",
      "2026/01/14 08:27:54  -Variants flipped :  0\n",
      "2026/01/14 08:27:54   -Raw Matching rate :  0.00%\n",
      "2026/01/14 08:27:54 #WARNING! Matching rate is low, please check if the right reference genome is used.\n",
      "2026/01/14 08:27:54  -Variants inferred reverse_complement :  0\n",
      "2026/01/14 08:27:54  -Variants inferred reverse_complement_flipped :  0\n",
      "2026/01/14 08:27:54  -Both allele on genome + unable to distinguish :  0\n",
      "2026/01/14 08:27:54  -Variants not on given reference sequence :  0\n",
      "2026/01/14 08:27:55 Finished checking if NEA is aligned with reference sequence.\n",
      "2026/01/14 08:27:55 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/14 08:27:56  -No statistics have been changed.\n",
      "2026/01/14 08:27:56 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/14 08:27:56 Start to infer strand for palindromic SNPs/align indistinguishable indels ...(v4.0.4)\n",
      "2026/01/14 08:27:56  -Number of threads/cores to use: 6\n",
      "2026/01/14 08:27:56  -Field for alternative allele frequency in VCF INFO: AF\n",
      "2026/01/14 08:27:57  -Identified  1658027  palindromic SNPs...\n",
      "2026/01/14 08:27:57  -After filtering by MAF<= 0.4 , 0 palindromic SNPs with unknown strand will be inferred...\n",
      "2026/01/14 08:27:57 #WARNING! No palindromic variants available for checking.\n",
      "2026/01/14 08:27:59   -Non-palindromic :  9162782\n",
      "2026/01/14 08:28:00   -Palindromic SNPs on + strand:  0\n",
      "2026/01/14 08:28:01   -Palindromic SNPs on - strand and needed to be flipped: 0\n",
      "2026/01/14 08:28:03   -Palindromic SNPs with MAF not available to infer :  86036\n",
      "2026/01/14 08:28:04   -Palindromic SNPs lacking reference matches, or with reference ALT unavailable or above MAF threshold for inference :  0\n",
      "2026/01/14 08:28:04  -Identified  0  indistinguishable Indels...\n",
      "2026/01/14 08:28:04 #WARNING! No indistinguishable indels available for checking.\n",
      "2026/01/14 08:28:04 Finished inferring strand for palindromic SNPs/align indistinguishable indels.\n",
      "2026/01/14 08:28:04 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/14 08:28:05  -No statistics have been changed.\n",
      "2026/01/14 08:28:05 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/14 08:28:05 Start to remove duplicated/multiallelic variants ...(v4.0.4)\n",
      "2026/01/14 08:28:05  -Removing mode:dm\n",
      "2026/01/14 08:28:05 Start to sort the sumstats using P ...\n",
      "2026/01/14 08:28:12 Start to remove duplicated variants based on snpid ...(v4.0.4)\n",
      "2026/01/14 08:28:12  -Which variant to keep:  first\n",
      "2026/01/14 08:28:34  -Removed variants based on SNPID: 0\n",
      "2026/01/14 08:28:34 Start to remove duplicated variants based on CHR,POS,EA and NEA ...\n",
      "2026/01/14 08:28:34  -Which variant to keep:  first\n",
      "2026/01/14 08:28:49  -Removed variants based on CHR,POS,EA and NEA: 0\n",
      "2026/01/14 08:28:49 Start to remove multiallelic variants based on chr:pos ...\n",
      "2026/01/14 08:28:49  -Which variant to keep:  first\n",
      "2026/01/14 08:29:05  -Removed variants multiallelic variants: 0\n",
      "2026/01/14 08:29:05  -Removed variants in total: 0\n",
      "2026/01/14 08:29:05  -Sort the coordinates based on CHR and POS...\n",
      "2026/01/14 08:29:26 Finished removing duplicated/multiallelic variants.\n",
      "2026/01/14 08:29:26 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/14 08:29:39 Finished sorting coordinates.\n",
      "2026/01/14 08:29:39 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 08:29:39  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 08:29:40 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.889410</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761651</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761652</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.770274</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761653</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>9960399</td>\n",
       "      <td>0.760570</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761654</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13761655</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13761656 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7243ba757d70>"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 1. Explicitly set the build metadata so the harmonizer knows the goal\n",
    "mysumstats.data.attrs[\"build\"] = \"38\"\n",
    "\n",
    "# 2. Run the final harmonization with 'remove=True' \n",
    "# This will drop the variants that are truly broken or incompatible with hg38\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=\"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\",\n",
    "    ref_infer=\"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz\",\n",
    "    ref_alt_freq=\"AF\",\n",
    "    threads=6,\n",
    "    remove=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f6022d4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "946f78c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use ref_rsid_tsv because the file is a text table, not a VCF\n",
    "mysumstats.assign_rsid(\n",
    "    ref_rsid_tsv=\"/home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz\",\n",
    "    threads=4\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6efb77ac",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8d9fb93",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A. Manually fix Chromosomes to avoid the \"target\" error\n",
    "mysumstats.data['CHR'] = mysumstats.data['CHR'].astype(str).str.replace('chr', '', case=False)\n",
    "mysumstats.data['CHR'] = mysumstats.data['CHR'].replace({'X': '23', 'Y': '24', 'MT': '25'})\n",
    "\n",
    "# B. Set the build globally for this object\n",
    "mysumstats.set_build(\"38\")\n",
    "\n",
    "# C. Run harmonization (removing the 'build' keyword that caused the error)\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=\"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/hg38.fa\",\n",
    "    basic_check=False,\n",
    "    threads=4\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b300662b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
