{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b47056e",
   "metadata": {},
   "source": [
    "#### Harmonization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c40ea94b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "911c97b1",
   "metadata": {},
   "source": [
    "### Load test data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8695a92",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 13:09:52 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/14 13:09:52 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/14 13:09:52 Python version: 3.12.3 (main, Jan  8 2026, 11:30:50) [GCC 13.3.0]\n",
      "2026/01/14 13:09:52 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 13:10:18  -Reading columns          : CHR,P,BETA,NEA,POS,SE,EA,EAF,SNPID\n",
      "2026/01/14 13:10:18  -Renaming columns to      : CHR,P,BETA,NEA,POS,SE,EA,EAF,SNPID\n",
      "2026/01/14 13:10:18  -Current Dataframe shape : 13779885  x  9\n",
      "2026/01/14 13:10:18  -Initiating a status column: STATUS ...\n",
      "2026/01/14 13:10:18 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/14 13:10:22 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 13:10:22  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 13:10:23 Finished reordering the columns.\n",
      "2026/01/14 13:10:23  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2026/01/14 13:10:25  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P      \n",
      "2026/01/14 13:10:25  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64\n",
      "2026/01/14 13:10:25  -Verified: T      T     T     T        T        T      T       T       T       T      \n",
      "2026/01/14 13:10:25  -Current Dataframe memory usage: 963.05 MB\n",
      "2026/01/14 13:10:25 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover_rsid.tsv.gz\",\n",
    "    snpid=\"SNPID\",\n",
    "    chrom=\"CHR\",\n",
    "    pos=\"POS\",\n",
    "    ea=\"EA\",            \n",
    "    nea=\"NEA\",          \n",
    "    eaf=\"EAF\",          \n",
    "    beta=\"BETA\",\n",
    "    se=\"SE\",\n",
    "    p=\"P\",\n",
    "    verbose=True,\n",
    "    readargs={'sep': '\\t'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f8803a36",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>707522</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.902745</td>\n",
       "      <td>-0.010428</td>\n",
       "      <td>0.020804</td>\n",
       "      <td>0.616190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631866</th>\n",
       "      <td>X:154925895:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154925895</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.758067</td>\n",
       "      <td>-0.019349</td>\n",
       "      <td>0.010966</td>\n",
       "      <td>0.077653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631867</th>\n",
       "      <td>X:154927581:G:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154927581</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.697822</td>\n",
       "      <td>-0.012985</td>\n",
       "      <td>0.010082</td>\n",
       "      <td>0.197734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631868</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631869</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631870</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12631871 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SNPID  CHR        POS EA NEA   STATUS       EAF      BETA  \\\n",
       "0             1:69487:G:A    1      69487  A   G  9960090  0.999994 -2.715450   \n",
       "1             1:69569:T:C    1      69569  C   T  9960090  0.999812 -0.484284   \n",
       "2            1:139853:C:T    1     139853  T   C  9960090  0.999994 -2.703560   \n",
       "3            1:693731:A:G    1     693731  G   A  9960090  0.884233 -0.004255   \n",
       "4            1:707522:G:C    1     707522  C   G  9960099  0.902745 -0.010428   \n",
       "...                   ...  ...        ... ..  ..      ...       ...       ...   \n",
       "12631866  X:154925895:C:T   23  154925895  T   C  9960090  0.758067 -0.019349   \n",
       "12631867  X:154927581:G:A   23  154927581  A   G  9960090  0.697822 -0.012985   \n",
       "12631868  X:154929412:C:T   23  154929412  T   C  9960090  0.754527 -0.016260   \n",
       "12631869  X:154930230:A:G   23  154930230  G   A  9960090  0.754113 -0.016347   \n",
       "12631870  X:154930487:T:A   23  154930487  A   T  9960099  0.980441  0.007107   \n",
       "\n",
       "                SE         P  \n",
       "0         2.360060  0.249902  \n",
       "1         0.423462  0.252778  \n",
       "2         2.360130  0.251997  \n",
       "3         0.018507  0.818155  \n",
       "4         0.020804  0.616190  \n",
       "...            ...       ...  \n",
       "12631866  0.010966  0.077653  \n",
       "12631867  0.010082  0.197734  \n",
       "12631868  0.010723  0.129427  \n",
       "12631869  0.010721  0.127334  \n",
       "12631870  0.036783  0.846790  \n",
       "\n",
       "[12631871 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3cf97a5",
   "metadata": {},
   "source": [
    "### Perform harmonization\n",
    " - ref_seq : reference genome fasta file for allele alignment\n",
    " - ref_infer : vcf file with allele frequency information for inferring strand and comparing allele frequency\n",
    " - ref_alt_freq : field in INFO of vcf file for alternative allele frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa7b57e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/14 13:41:05 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/14 13:41:05  -Checking SNPID data type...\n",
      "2026/01/14 13:41:05  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/14 13:41:06  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/14 13:41:19 Finished checking SNPID/rsID.\n",
      "2026/01/14 13:41:19 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/14 13:41:19  -Checking CHR data type...\n",
      "2026/01/14 13:41:23  -Variants with standardized chromosome notation: 12631871\n",
      "2026/01/14 13:41:23  -All CHR are already fixed...\n",
      "2026/01/14 13:41:28 Finished fixing chromosome notation (CHR).\n",
      "2026/01/14 13:41:28 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/14 13:41:28  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/14 13:41:32  -Position bound:(0 , 250,000,000)\n",
      "2026/01/14 13:41:33  -No outlier variants were removed.\n",
      "2026/01/14 13:41:41  -Removed variants with bad positions: 0\n",
      "2026/01/14 13:41:41 Finished fixing basepair positions (POS).\n",
      "2026/01/14 13:41:41 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/14 13:41:41  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/14 13:41:45  -Variants with bad EA: 0\n",
      "2026/01/14 13:41:45  -Variants with bad NEA: 0\n",
      "2026/01/14 13:41:45  -Variants with NA for EA or NEA: 0\n",
      "2026/01/14 13:41:45  -Variants with same EA and NEA: 1809229\n",
      "2026/01/14 13:41:58 Finished fixing alleles (EA and NEA).\n",
      "2026/01/14 13:41:58 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/14 13:41:58  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/14 13:41:59  -Checking if any columns are empty...\n",
      "2026/01/14 13:42:00  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/14 13:42:03  -Checking if -1e-07 < P < 1.0000001 : removed 0 variants.\n",
      "2026/01/14 13:42:04  -Checking if -100.0000001 < BETA < 100.0000001 : removed 0 variants.\n",
      "2026/01/14 13:42:05  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/14 13:42:06  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/14 13:42:06  -No variants were removed with bad statistics in total.\n",
      "2026/01/14 13:42:06 Finished sanity check for statistics.\n",
      "2026/01/14 13:42:06 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/14 13:42:06  -Number of threads/cores to use: 6\n",
      "2026/01/14 13:42:06  -No available variants to normalize..\n",
      "2026/01/14 13:42:06 Finished normalizing indels.\n",
      "2026/01/14 13:42:06 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 13:42:06  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 13:42:06 Finished reordering the columns.\n",
      "2026/01/14 13:42:06 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/14 13:42:18 Finished sorting coordinates.\n",
      "2026/01/14 13:42:19 Start to check if NEA is aligned with reference sequence ...(v4.0.4)\n",
      "2026/01/14 13:42:19  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\n",
      "2026/01/14 13:42:28    -Loading and building numpy fasta records:\n",
      "2026/01/14 13:42:36  -Variants allele on given reference sequence :  0\n",
      "2026/01/14 13:42:36  -Variants flipped :  0\n",
      "2026/01/14 13:42:36   -Raw Matching rate :  0.00%\n",
      "2026/01/14 13:42:36 #WARNING! Matching rate is low, please check if the right reference genome is used.\n",
      "2026/01/14 13:42:36  -Variants inferred reverse_complement :  0\n",
      "2026/01/14 13:42:36  -Variants inferred reverse_complement_flipped :  0\n",
      "2026/01/14 13:42:36  -Both allele on genome + unable to distinguish :  0\n",
      "2026/01/14 13:42:36  -Variants not on given reference sequence :  0\n",
      "2026/01/14 13:42:36 Finished checking if NEA is aligned with reference sequence.\n",
      "2026/01/14 13:42:36 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/14 13:42:37  -No statistics have been changed.\n",
      "2026/01/14 13:42:37 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/14 13:42:37 Start to infer strand for palindromic SNPs/align indistinguishable indels ...(v4.0.4)\n",
      "2026/01/14 13:42:37  -Number of threads/cores to use: 6\n",
      "2026/01/14 13:42:37  -Field for alternative allele frequency in VCF INFO: AF\n",
      "2026/01/14 13:42:38  -Identified  1658600  palindromic SNPs...\n",
      "2026/01/14 13:42:38  -After filtering by MAF<= 0.4 , 0 palindromic SNPs with unknown strand will be inferred...\n",
      "2026/01/14 13:42:38 #WARNING! No palindromic variants available for checking.\n",
      "2026/01/14 13:42:40   -Non-palindromic :  9164042\n",
      "2026/01/14 13:42:41   -Palindromic SNPs on + strand:  0\n",
      "2026/01/14 13:42:42   -Palindromic SNPs on - strand and needed to be flipped: 0\n",
      "2026/01/14 13:42:43   -Palindromic SNPs with MAF not available to infer :  86055\n",
      "2026/01/14 13:42:44   -Palindromic SNPs lacking reference matches, or with reference ALT unavailable or above MAF threshold for inference :  0\n",
      "2026/01/14 13:42:45  -Identified  0  indistinguishable Indels...\n",
      "2026/01/14 13:42:45 #WARNING! No indistinguishable indels available for checking.\n",
      "2026/01/14 13:42:45 Finished inferring strand for palindromic SNPs/align indistinguishable indels.\n",
      "2026/01/14 13:42:45 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/14 13:42:45  -No statistics have been changed.\n",
      "2026/01/14 13:42:45 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/14 13:42:45 Start to remove duplicated/multiallelic variants ...(v4.0.4)\n",
      "2026/01/14 13:42:45  -Removing mode:dm\n",
      "2026/01/14 13:42:46 Start to sort the sumstats using P ...\n",
      "2026/01/14 13:42:51 Start to remove duplicated variants based on snpid ...(v4.0.4)\n",
      "2026/01/14 13:42:51  -Which variant to keep:  first\n",
      "2026/01/14 13:43:07  -Removed variants based on SNPID: 0\n",
      "2026/01/14 13:43:07 Start to remove duplicated variants based on CHR,POS,EA and NEA ...\n",
      "2026/01/14 13:43:07  -Which variant to keep:  first\n",
      "2026/01/14 13:43:19  -Removed variants based on CHR,POS,EA and NEA: 0\n",
      "2026/01/14 13:43:19 Start to remove multiallelic variants based on chr:pos ...\n",
      "2026/01/14 13:43:19  -Which variant to keep:  first\n",
      "2026/01/14 13:43:32  -Removed variants multiallelic variants: 0\n",
      "2026/01/14 13:43:32  -Removed variants in total: 0\n",
      "2026/01/14 13:43:32  -Sort the coordinates based on CHR and POS...\n",
      "2026/01/14 13:43:48 Finished removing duplicated/multiallelic variants.\n",
      "2026/01/14 13:43:48 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/14 13:43:58 Finished sorting coordinates.\n",
      "2026/01/14 13:43:59 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/14 13:43:59  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/14 13:43:59 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>707522</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.902745</td>\n",
       "      <td>-0.010428</td>\n",
       "      <td>0.020804</td>\n",
       "      <td>0.616190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631866</th>\n",
       "      <td>X:154925895:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154925895</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.758067</td>\n",
       "      <td>-0.019349</td>\n",
       "      <td>0.010966</td>\n",
       "      <td>0.077653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631867</th>\n",
       "      <td>X:154927581:G:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154927581</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.697822</td>\n",
       "      <td>-0.012985</td>\n",
       "      <td>0.010082</td>\n",
       "      <td>0.197734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631868</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631869</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631870</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12631871 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x71809af45940>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_bases = {'A', 'T', 'C', 'G'}\n",
    "mysumstats.data = mysumstats.data[\n",
    "    mysumstats.data[\"EA\"].isin(valid_bases) & \n",
    "    mysumstats.data[\"NEA\"].isin(valid_bases)\n",
    "].copy()\n",
    "\n",
    "mysumstats.data.attrs[\"build\"] = \"38\"\n",
    "\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=\"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa\",\n",
    "    ref_infer=\"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz\",\n",
    "    ref_alt_freq=\"AF\",\n",
    "    threads=6,\n",
    "    remove=True\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9fbdd12",
   "metadata": {},
   "source": [
    "\n",
    "All variants were checked and harmonized based on the reference datasets. The manipulations are reflected in STATUS column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a94ef127",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999812</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.999994</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.884233</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:707522:G:C</td>\n",
       "      <td>1</td>\n",
       "      <td>707522</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.902745</td>\n",
       "      <td>-0.010428</td>\n",
       "      <td>0.020804</td>\n",
       "      <td>0.616190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631866</th>\n",
       "      <td>X:154925895:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154925895</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.758067</td>\n",
       "      <td>-0.019349</td>\n",
       "      <td>0.010966</td>\n",
       "      <td>0.077653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631867</th>\n",
       "      <td>X:154927581:G:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154927581</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.697822</td>\n",
       "      <td>-0.012985</td>\n",
       "      <td>0.010082</td>\n",
       "      <td>0.197734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631868</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754527</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631869</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>9960090</td>\n",
       "      <td>0.754113</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12631870</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>9960099</td>\n",
       "      <td>0.980441</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12631871 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SNPID  CHR        POS EA NEA   STATUS       EAF      BETA  \\\n",
       "0             1:69487:G:A    1      69487  A   G  9960090  0.999994 -2.715450   \n",
       "1             1:69569:T:C    1      69569  C   T  9960090  0.999812 -0.484284   \n",
       "2            1:139853:C:T    1     139853  T   C  9960090  0.999994 -2.703560   \n",
       "3            1:693731:A:G    1     693731  G   A  9960090  0.884233 -0.004255   \n",
       "4            1:707522:G:C    1     707522  C   G  9960099  0.902745 -0.010428   \n",
       "...                   ...  ...        ... ..  ..      ...       ...       ...   \n",
       "12631866  X:154925895:C:T   23  154925895  T   C  9960090  0.758067 -0.019349   \n",
       "12631867  X:154927581:G:A   23  154927581  A   G  9960090  0.697822 -0.012985   \n",
       "12631868  X:154929412:C:T   23  154929412  T   C  9960090  0.754527 -0.016260   \n",
       "12631869  X:154930230:A:G   23  154930230  G   A  9960090  0.754113 -0.016347   \n",
       "12631870  X:154930487:T:A   23  154930487  A   T  9960099  0.980441  0.007107   \n",
       "\n",
       "                SE         P  \n",
       "0         2.360060  0.249902  \n",
       "1         0.423462  0.252778  \n",
       "2         2.360130  0.251997  \n",
       "3         0.018507  0.818155  \n",
       "4         0.020804  0.616190  \n",
       "...            ...       ...  \n",
       "12631866  0.010966  0.077653  \n",
       "12631867  0.010082  0.197734  \n",
       "12631868  0.010723  0.129427  \n",
       "12631869  0.010721  0.127334  \n",
       "12631870  0.036783  0.846790  \n",
       "\n",
       "[12631871 rows x 10 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c67b2342",
   "metadata": {},
   "source": [
    "### Check summary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a8c4eb0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'overview': {'Row_num': '12631871',\n",
       "  'Column_num': '10',\n",
       "  'Column_names': 'SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P',\n",
       "  'Last_checked_time': 'Wed Jan 14 13:46:14 2026',\n",
       "  'QC and Harmonization': {'basic_check': {'performed': True,\n",
       "    'last_executed': '2026-01-14 13:42:18',\n",
       "    'parameters_used': {'basic_check': True,\n",
       "     'ref_seq': '/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa',\n",
       "     'ref_rsid_tsv': None,\n",
       "     'ref_rsid_vcf': None,\n",
       "     'ref_infer': '/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz',\n",
       "     'ref_alt_freq': 'AF',\n",
       "     'ref_maf_threshold': 0.4,\n",
       "     'maf_threshold': 0.4,\n",
       "     'threads': 6,\n",
       "     'remove': True,\n",
       "     'check_ref_kwargs': {},\n",
       "     'remove_dup_kwargs': {},\n",
       "     'assign_rsid_kwargs': {},\n",
       "     'infer_strand_kwargs': {},\n",
       "     'flip_allele_stats_kwargs': {},\n",
       "     'liftover_kwargs': {},\n",
       "     'fix_id_kwargs': {},\n",
       "     'fix_chr_kwargs': {},\n",
       "     'fix_pos_kwargs': {},\n",
       "     'fix_allele_kwargs': {},\n",
       "     'sanity_check_stats_kwargs': {},\n",
       "     'normalize_allele_kwargs': {},\n",
       "     'verbose': True,\n",
       "     'sweep_mode': False}},\n",
       "   'harmonize': {'performed': True,\n",
       "    'last_executed': '2026-01-14 13:43:59',\n",
       "    'parameters_used': {'basic_check': True,\n",
       "     'ref_seq': '/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/hg38.fa',\n",
       "     'ref_rsid_tsv': None,\n",
       "     'ref_rsid_vcf': None,\n",
       "     'ref_infer': '/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/homo_sapiens-chr{chrom}.vcf.gz',\n",
       "     'ref_alt_freq': 'AF',\n",
       "     'ref_maf_threshold': 0.4,\n",
       "     'maf_threshold': 0.4,\n",
       "     'threads': 6,\n",
       "     'remove': True,\n",
       "     'check_ref_kwargs': {},\n",
       "     'remove_dup_kwargs': {},\n",
       "     'assign_rsid_kwargs': {},\n",
       "     'infer_strand_kwargs': {},\n",
       "     'flip_allele_stats_kwargs': {},\n",
       "     'liftover_kwargs': {},\n",
       "     'fix_id_kwargs': {},\n",
       "     'fix_chr_kwargs': {},\n",
       "     'fix_pos_kwargs': {},\n",
       "     'fix_allele_kwargs': {},\n",
       "     'sanity_check_stats_kwargs': {},\n",
       "     'normalize_allele_kwargs': {},\n",
       "     'verbose': True,\n",
       "     'sweep_mode': False}},\n",
       "   'qc_and_harmonization_status': {'qc': {'id': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:42:18',\n",
       "      'parameters_used': {}},\n",
       "     'chr': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:42:18',\n",
       "      'parameters_used': {}},\n",
       "     'pos': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:42:18',\n",
       "      'parameters_used': {}},\n",
       "     'allele': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:42:18',\n",
       "      'parameters_used': {}},\n",
       "     'sanity': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:42:18',\n",
       "      'parameters_used': {}},\n",
       "     'consistency': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:42:18',\n",
       "      'parameters_used': {}},\n",
       "     'normalize': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}},\n",
       "     'remove_dup': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:43:48',\n",
       "      'parameters_used': {'mode': 'dm',\n",
       "       'chrom': 'CHR',\n",
       "       'pos': 'POS',\n",
       "       'snpid': 'SNPID',\n",
       "       'ea': 'EA',\n",
       "       'nea': 'NEA',\n",
       "       'rsid': 'rsID',\n",
       "       'keep': 'first',\n",
       "       'keep_col': 'P',\n",
       "       'remove_na': False,\n",
       "       'keep_ascend': True}},\n",
       "     'sort_coord': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:43:58',\n",
       "      'parameters_used': {'chrom': 'CHR', 'pos': 'POS', 'reindex': True}},\n",
       "     'sort_column': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:43:59',\n",
       "      'parameters_used': {'order': ['SNPID',\n",
       "        'rsID',\n",
       "        'CHR',\n",
       "        'POS',\n",
       "        'EA',\n",
       "        'NEA',\n",
       "        'STATUS',\n",
       "        'REF',\n",
       "        'ALT',\n",
       "        'EAF',\n",
       "        'RAF',\n",
       "        'NEAF',\n",
       "        'MAF',\n",
       "        'DAF',\n",
       "        'INFO',\n",
       "        'BETA',\n",
       "        'SE',\n",
       "        'BETA_95U',\n",
       "        'BETA_95L',\n",
       "        'OR',\n",
       "        'OR_95U',\n",
       "        'OR_95L',\n",
       "        'HR',\n",
       "        'HR_95U',\n",
       "        'HR_95L',\n",
       "        'CHISQ',\n",
       "        'Z',\n",
       "        'T',\n",
       "        'F',\n",
       "        'P',\n",
       "        'P_MANTISSA',\n",
       "        'P_EXPONENT',\n",
       "        'MLOG10P',\n",
       "        'SNPR2',\n",
       "        'DOF',\n",
       "        'P_HET',\n",
       "        'I2_HET',\n",
       "        'I2',\n",
       "        'DIRECTION',\n",
       "        'DENSITY',\n",
       "        'N',\n",
       "        'N_CASE',\n",
       "        'N_CONTROL',\n",
       "        'N_EFF',\n",
       "        'GENENAME',\n",
       "        'CIS/TRANS',\n",
       "        'DISTANCE_TO_KNOWN',\n",
       "        'LOCATION_OF_KNOWN',\n",
       "        'KNOWN_ID',\n",
       "        'KNOWN_PUBMED_ID',\n",
       "        'KNOWN_AUTHOR',\n",
       "        'KNOWN_SET_VARIANT',\n",
       "        'KNOWN_VARIANT',\n",
       "        'KNOWN_SET',\n",
       "        'NOVEL',\n",
       "        'PIP',\n",
       "        'CREDIBLE_SET_INDEX',\n",
       "        'N_SNP',\n",
       "        'LOCUS',\n",
       "        'STUDY',\n",
       "        'TEST',\n",
       "        'BETA_RANDOM',\n",
       "        'SE_RANDOM',\n",
       "        'Z_RANDOM',\n",
       "        'P_RANDOM']}}},\n",
       "    'harmonize': {'check_ref': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:43:59',\n",
       "      'parameters_used': {}},\n",
       "     'flip_allele_stats': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:43:59',\n",
       "      'parameters_used': {}},\n",
       "     'infer_strand': {'performed': True,\n",
       "      'last_executed': '2026-01-14 13:43:59',\n",
       "      'parameters_used': {}},\n",
       "     'assign_rsid': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}},\n",
       "     'liftover': {'performed': False,\n",
       "      'last_executed': '',\n",
       "      'parameters_used': {}}}}}},\n",
       " 'chromosomes': {'Chromosomes_notations': [1,\n",
       "   2,\n",
       "   3,\n",
       "   4,\n",
       "   5,\n",
       "   6,\n",
       "   7,\n",
       "   8,\n",
       "   9,\n",
       "   10,\n",
       "   11,\n",
       "   12,\n",
       "   13,\n",
       "   14,\n",
       "   15,\n",
       "   16,\n",
       "   17,\n",
       "   18,\n",
       "   19,\n",
       "   20,\n",
       "   21,\n",
       "   22,\n",
       "   23],\n",
       "  'Chromosomes_numbers': 23,\n",
       "  'chr1': 956723,\n",
       "  'chr2': 1039127,\n",
       "  'chr3': 877890,\n",
       "  'chr4': 886991,\n",
       "  'chr5': 799667,\n",
       "  'chr6': 814966,\n",
       "  'chr7': 717584,\n",
       "  'chr8': 685433,\n",
       "  'chr9': 534786,\n",
       "  'chr10': 621526,\n",
       "  'chr11': 613862,\n",
       "  'chr12': 577765,\n",
       "  'chr13': 437372,\n",
       "  'chr14': 392787,\n",
       "  'chr15': 345667,\n",
       "  'chr16': 385609,\n",
       "  'chr17': 330467,\n",
       "  'chr18': 342514,\n",
       "  'chr19': 280649,\n",
       "  'chr20': 271457,\n",
       "  'chr21': 163170,\n",
       "  'chr22': 164185,\n",
       "  'chr23': 391674,\n",
       "  'Chromosomes_numbers percentage': 1.8207912351226513e-06,\n",
       "  'chr1 percentage': 0.07573881968870645,\n",
       "  'chr2 percentage': 0.08226231885996936,\n",
       "  'chr3 percentage': 0.0694980181479054,\n",
       "  'chr4 percentage': 0.07021849732315981,\n",
       "  'chr5 percentage': 0.06330550715725326,\n",
       "  'chr6 percentage': 0.06451664998795507,\n",
       "  'chr7 percentage': 0.05680741989844577,\n",
       "  'chr8 percentage': 0.05426219124625323,\n",
       "  'chr9 percentage': 0.042336246150708794,\n",
       "  'chr10 percentage': 0.049203004052210474,\n",
       "  'chr11 percentage': 0.04859628474673308,\n",
       "  'chr12 percentage': 0.045738671650462547,\n",
       "  'chr13 percentage': 0.03462448278643757,\n",
       "  'chr14 percentage': 0.03109491855957047,\n",
       "  'chr15 percentage': 0.027364671472658326,\n",
       "  'chr16 percentage': 0.030526673364539583,\n",
       "  'chr17 percentage': 0.026161365960751182,\n",
       "  'chr18 percentage': 0.027115064743773905,\n",
       "  'chr19 percentage': 0.02221753214547552,\n",
       "  'chr20 percentage': 0.021489848970116937,\n",
       "  'chr21 percentage': 0.012917326340650565,\n",
       "  'chr22 percentage': 0.012997678649504891,\n",
       "  'chr23 percentage': 0.031006808096757797},\n",
       " 'missing_values': {'Missing_total': 0, 'Missing_total percentage': 0.0},\n",
       " 'MAF': {'Common (MAF>=0.05)': 6212666,\n",
       "  'Low_frequency (0.01<MAF<=0.05)': 2497962,\n",
       "  'Rare (0.001<MAF<=0.01)': 3731173,\n",
       "  'Ultra Rare (MAF<=0.001)': 190070,\n",
       "  'Common (MAF>=0.05) percentage': 0.49182468693671744,\n",
       "  'Low_frequency (0.01<MAF<=0.05) percentage': 0.1977507528378021,\n",
       "  'Rare (0.001<MAF<=0.01) percentage': 0.29537769978809947,\n",
       "  'Ultra Rare (MAF<=0.001) percentage': 0.01504686043738097},\n",
       " 'p_values': {'Minimum': 5.4652e-220,\n",
       "  'P<5e-8': 46291,\n",
       "  'P<5e-6': 102733,\n",
       "  'P<5e-8 percentage': 0.0036646194376114194,\n",
       "  'P<5e-6 percentage': 0.008132841128602407},\n",
       " 'beta_for_significant_variants': {'P<5e-8 with ABS(BETA)>10': 11,\n",
       "  'P<5e-8 with ABS(BETA)>3': 11,\n",
       "  'P<5e-8 with ABS(BETA)>1': 11,\n",
       "  'P<5e-8 with ABS(BETA)>0.5': 64,\n",
       "  'P<5e-8 with ABS(BETA)>0.3': 245,\n",
       "  'P<5e-8 with ABS(BETA)>0.2': 754,\n",
       "  'P<5e-8 with ABS(BETA)>0.1': 9970,\n",
       "  'P<5e-8 with ABS(BETA)>0.05': 46291,\n",
       "  'P<5e-8 with ABS(BETA)>10 percentage': 8.708131994064854e-07,\n",
       "  'P<5e-8 with ABS(BETA)>3 percentage': 8.708131994064854e-07,\n",
       "  'P<5e-8 with ABS(BETA)>1 percentage': 8.708131994064854e-07,\n",
       "  'P<5e-8 with ABS(BETA)>0.5 percentage': 5.066549523819552e-06,\n",
       "  'P<5e-8 with ABS(BETA)>0.3 percentage': 1.939538489587172e-05,\n",
       "  'P<5e-8 with ABS(BETA)>0.2 percentage': 5.969028657749909e-05,\n",
       "  'P<5e-8 with ABS(BETA)>0.1 percentage': 0.0007892734180075145,\n",
       "  'P<5e-8 with ABS(BETA)>0.05 percentage': 0.0036646194376114194},\n",
       " 'variant_status': {'9960090': {'count': 9164042,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Not_palindromic_SNPs'}},\n",
       "  '9960699': {'count': 1809229,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'invalid allele notation',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Unchecked'}},\n",
       "  '9960099': {'count': 1572545,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Unchecked'}},\n",
       "  '9960097': {'count': 86055,\n",
       "   'explanation': {'Genome_Build': 'Unchecked',\n",
       "    'rsID&SNPID': 'rsid unknown & SNPID valid',\n",
       "    'CHR&POS': 'CHR valid & POS valid',\n",
       "    'Stadardize&Normalize': 'standardized SNP',\n",
       "    'Align': 'Unchecked',\n",
       "    'Panlidromic_SNP&Indel': 'Indistinguishable'}}},\n",
       " 'variants': {'variant_number': 12631871,\n",
       "  'min_P': 5.4652e-220,\n",
       "  'min_minor_allele_freq': 0.0}}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f8a6bc09",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome_Build</th>\n",
       "      <th>rsID&amp;SNPID</th>\n",
       "      <th>CHR&amp;POS</th>\n",
       "      <th>Stadardize&amp;Normalize</th>\n",
       "      <th>Align</th>\n",
       "      <th>Panlidromic_SNP&amp;Indel</th>\n",
       "      <th>Count</th>\n",
       "      <th>Percentage(%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>9960090</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>Not_palindromic_SNPs</td>\n",
       "      <td>9164042</td>\n",
       "      <td>72.55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960097</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>Indistinguishable</td>\n",
       "      <td>86055</td>\n",
       "      <td>0.68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960099</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>standardized SNP</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>1572545</td>\n",
       "      <td>12.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9960699</th>\n",
       "      <td>Unchecked</td>\n",
       "      <td>rsid unknown &amp; SNPID valid</td>\n",
       "      <td>CHR valid &amp; POS valid</td>\n",
       "      <td>invalid allele notation</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>Unchecked</td>\n",
       "      <td>1809229</td>\n",
       "      <td>14.32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Genome_Build                  rsID&SNPID                CHR&POS  \\\n",
       "9960090    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960097    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960099    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "9960699    Unchecked  rsid unknown & SNPID valid  CHR valid & POS valid   \n",
       "\n",
       "            Stadardize&Normalize      Align Panlidromic_SNP&Indel    Count  \\\n",
       "9960090         standardized SNP  Unchecked  Not_palindromic_SNPs  9164042   \n",
       "9960097         standardized SNP  Unchecked     Indistinguishable    86055   \n",
       "9960099         standardized SNP  Unchecked             Unchecked  1572545   \n",
       "9960699  invalid allele notation  Unchecked             Unchecked  1809229   \n",
       "\n",
       "        Percentage(%)  \n",
       "9960090         72.55  \n",
       "9960097          0.68  \n",
       "9960099         12.45  \n",
       "9960699         14.32  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mysumstats.lookup_status()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
