{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7d48475f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0,\"/home/yunye/work/gwaslab/src\")\n",
    "import gwaslab as gl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "623ca292",
   "metadata": {},
   "source": [
    "### loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1217ef70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 13:31:45 Start to check downloaded reference files...\n",
      "2026/01/10 13:31:45  -Checking the config file:/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/data/config.json\n",
      "2026/01/10 13:31:45  -Config file exists.\n",
      "2026/01/10 13:31:45  -Updating config.json...\n",
      "2026/01/10 13:31:45  - Local file keywords:  1kg_dbsnp151_hg38_auto\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'1kg_dbsnp151_hg38_auto': {'local_path': '/home/ofgeha/.gwaslab/1kg_dbsnp151_hg38_auto.txt.gz',\n",
       "  'description': '1KG dbSNP151 hg38 auto variants table containing SNP ID, chromosome, position, effect allele, non-effect allele, and rsID mappings for autosomal variants.',\n",
       "  'suggested_use': 'SNP annotation and rsID mapping'}}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gl.check_downloaded_ref()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d56b7c03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 13:40:56 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/10 13:40:56 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/10 13:40:56 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/10 13:40:56 Start to initialize gl.Sumstats from file :/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover.tsv.gz\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 13:41:26  -Reading columns          : SNPID,EAF,CHR,POS,EA,P,NEA,SE,BETA\n",
      "2026/01/10 13:41:26  -Renaming columns to      : SNPID,EAF,CHR,POS,EA,P,NEA,SE,BETA\n",
      "2026/01/10 13:41:26  -Current Dataframe shape : 13640038  x  9\n",
      "2026/01/10 13:41:27  -Initiating a status column: STATUS ...\n",
      "2026/01/10 13:41:27 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/10 13:41:30 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 13:41:30  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P\n",
      "2026/01/10 13:41:31 Finished reordering the columns.\n",
      "2026/01/10 13:41:31  -Trying to convert datatype for CHR: string -> Int64...Success\n",
      "2026/01/10 13:41:32  -Column  : SNPID  CHR   POS   EA       NEA      STATUS EAF     BETA    SE      P      \n",
      "2026/01/10 13:41:32  -DType   : object Int64 int64 category category int64  float64 float64 float64 float64\n",
      "2026/01/10 13:41:32  -Verified: T      T     T     T        T        T      T       T       T       T      \n",
      "2026/01/10 13:41:33  -Current Dataframe memory usage: 953.30 MB\n",
      "2026/01/10 13:41:33 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "mysumstats = gl.Sumstats(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_T2T_liftover.tsv.gz\",\n",
    "    snpid=\"SNPID\",\n",
    "    chrom=\"CHR\",\n",
    "    pos=\"POS\",\n",
    "    ea=\"EA\",            \n",
    "    nea=\"NEA\",          \n",
    "    eaf=\"EAF\",          \n",
    "    beta=\"BETA\",\n",
    "    se=\"SE\",\n",
    "    p=\"P\",\n",
    "    # Note: 'Dir' and 'N' are not in your column list, \n",
    "    # so we should omit them or point to the correct ones.\n",
    "    # n=\"n_complete_samples\", # Optional: if you want to use sample size\n",
    "    verbose=True,\n",
    "    readargs={'sep': '\\t'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "402d4f9e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
