{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd99acb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "796daaf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Load your original UKBB dataset\n",
    "# Nealelab UKBB v3 is in build 19\n",
    "df = pd.read_csv(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "254d195c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split variant into CHR, POS, REF, ALT\n",
    "df[['CHR', 'POS', 'REF', 'ALT']] = df['variant'].str.split(\":\", expand=True)\n",
    "df['POS'] = df['POS'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f37ec073",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename to GWASLab standards\n",
    "df_standard = df.rename(columns={\n",
    "    'variant': 'SNPID',\n",
    "    'ALT': 'EA',\n",
    "    'REF': 'NEA',\n",
    "    'minor_AF': 'EAF',\n",
    "    'beta': 'BETA',\n",
    "    'se': 'SE',\n",
    "    'pval': 'P'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8ae53e22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 15:45:01 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/10 15:45:01 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/10 15:45:01 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/10 15:45:01 Start to initialize gl.Sumstats from pandas DataFrame ...\n",
      "2026/01/10 15:45:02  -Reading columns          : \n",
      "2026/01/10 15:45:02  -Renaming columns to      : \n",
      "2026/01/10 15:45:02  -Current Dataframe shape : 13791467  x  15\n",
      "2026/01/10 15:45:02  -Initiating a status column: STATUS ...\n",
      "2026/01/10 15:45:02  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2026/01/10 15:45:05 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 15:45:05  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 15:45:05 Finished reordering the columns.\n",
      "2026/01/10 15:45:06  -Trying to convert datatype for CHR: object -> Int64...Failed...\n",
      "2026/01/10 15:45:09  -Column  : SNPID  CHR    POS   EA       NEA      STATUS EAF     BETA    SE      P       minor_allele low_confidence_variant n_complete_samples AC      ytx     tstat  \n",
      "2026/01/10 15:45:09  -DType   : object object int64 category category int64  float64 float64 float64 float64 object       bool                   int64              float64 float64 float64\n",
      "2026/01/10 15:45:09  -Verified: T      F      T     T        T        T      T       T       T       T       NA           NA                     NA                 NA      NA      NA     \n",
      "2026/01/10 15:45:09 #WARNING! Columns with possibly incompatible dtypes: CHR\n",
      "2026/01/10 15:45:09 #WARNING! Consider using Sumstats.fix_chr() to fix CHR dtype\n",
      "2026/01/10 15:45:09  -Current Dataframe memory usage: 1489.96 MB\n",
      "2026/01/10 15:45:09 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "# 2. Initialize as Build 19 (Essential!)\n",
    "mysumstats = gl.Sumstats(df_standard, build=\"19\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6c61d2fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 15:45:50 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/10 15:45:50  -Current Dataframe shape : 13791467 x 16 ; Memory usage: 1489.96 MB\n",
      "2026/01/10 15:45:50 #WARNING! Data types were not fixed for : CHR \n",
      "2026/01/10 15:45:50 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 15:45:50  -Checking CHR data type...\n",
      "2026/01/10 15:45:52  -Variants with standardized chromosome notation: 13364303\n",
      "2026/01/10 15:45:53  -Variants with fixable chromosome notations: 427164\n",
      "2026/01/10 15:45:53  -No unrecognized chromosome notations...\n",
      "2026/01/10 15:45:53  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/10 15:45:53  -Identified 427164 variants on sex chromosomes...\n",
      "2026/01/10 15:45:53  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/10 15:45:56 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 15:45:56  -Dtype fixes applied successfully for requested columns.\n",
      "2026/01/10 15:45:56  -Using built-in chain file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/data/chains/hg19ToHg38.over.chain.gz\n",
      "2026/01/10 15:45:56  -No variants with status code xxx0xxx to liftover\n",
      "2026/01/10 15:45:56 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "# 3. Step One: Liftover from hg19 to hg38\n",
    "# GWASLab has built-in chains for 19 to 38\n",
    "mysumstats.liftover(from_build=\"19\", to_build=\"38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1a405193",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 15:47:03 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/10 15:47:03  -Checking SNPID data type...\n",
      "2026/01/10 15:47:03  -Converted datatype for SNPID: object -> string\n",
      "2026/01/10 15:47:04  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/10 15:47:05  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/10 15:47:15 Finished checking SNPID/rsID.\n",
      "2026/01/10 15:47:15 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 15:47:15  -Checking CHR data type...\n",
      "2026/01/10 15:47:19  -Variants with standardized chromosome notation: 13791467\n",
      "2026/01/10 15:47:19  -All CHR are already fixed...\n",
      "2026/01/10 15:47:21 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 15:47:21 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 15:47:21  -Trying to convert datatype for POS: int64 -> Int64...\n",
      "2026/01/10 15:47:22  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 15:47:22  -No outlier variants were removed.\n",
      "2026/01/10 15:47:23 Finished fixing basepair positions (POS).\n",
      "2026/01/10 15:47:24 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/10 15:47:24  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/10 15:47:29  -All variants have valid alleles (EA and NEA are A/C/T/G, not NA, and different).\n",
      "2026/01/10 15:47:43 Finished fixing alleles (EA and NEA).\n",
      "2026/01/10 15:47:43 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/10 15:47:43  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/10 15:47:43  -Checking if any columns are empty...\n",
      "2026/01/10 15:47:44  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/10 15:47:45   -Examples of invalid variants(SNPID): 1:776572:C:T,1:819974:G:C,1:871173:C:T,1:871198:C:T,1:874451:C:T ...\n",
      "2026/01/10 15:47:45   -Examples of invalid values (P): NA,NA,NA,NA,NA ...\n",
      "2026/01/10 15:47:46  -Checking if -1e-07 < P < 1.0000001 : removed 1676 variants.\n",
      "2026/01/10 15:47:46   -Examples of invalid variants(SNPID): 1:15791:C:T,1:877608:C:G,1:897032:A:G,1:899336:G:A,1:907676:A:T ...\n",
      "2026/01/10 15:47:46   -Examples of invalid values (BETA): 894.616,279.71,-108.064,270.068,116.14 ...\n",
      "2026/01/10 15:47:47  -Checking if -100.0000001 < BETA < 100.0000001 : removed 9906 variants.\n",
      "2026/01/10 15:47:49  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/10 15:47:49  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/10 15:47:49  -Removed 11582 variants with bad statistics in total.\n",
      "2026/01/10 15:47:49 Finished sanity check for statistics.\n",
      "2026/01/10 15:47:49 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/10 15:47:49  -Number of threads/cores to use: 1\n",
      "2026/01/10 15:47:49  -Current Dataframe shape : 13779885 x 16 ; Memory usage: 1582.28 MB\n",
      "2026/01/10 15:47:49  -Number of variants to check:728\n",
      "2026/01/10 15:47:49  -Chunk size:3000000\n",
      "2026/01/10 15:47:49  -Processing in chunks:0 \n",
      "2026/01/10 15:47:51 Finished normalizing indels.\n",
      "2026/01/10 15:47:51 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 15:47:51  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 15:47:52 Finished reordering the columns.\n",
      "2026/01/10 15:47:52 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/10 15:48:06 Finished sorting coordinates.\n",
      "2026/01/10 15:48:06 Start to check if NEA is aligned with reference sequence ...(v4.0.4)\n",
      "2026/01/10 15:48:06  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/hg38.fa\n",
      "2026/01/10 15:48:20    -Loading and building numpy fasta records:\n",
      "2026/01/10 15:48:28  -Variants allele on given reference sequence :  0\n",
      "2026/01/10 15:48:28  -Variants flipped :  0\n",
      "2026/01/10 15:48:28   -Raw Matching rate :  0.00%\n",
      "2026/01/10 15:48:28 #WARNING! Matching rate is low, please check if the right reference genome is used.\n",
      "2026/01/10 15:48:28  -Variants inferred reverse_complement :  0\n",
      "2026/01/10 15:48:28  -Variants inferred reverse_complement_flipped :  0\n",
      "2026/01/10 15:48:28  -Both allele on genome + unable to distinguish :  0\n",
      "2026/01/10 15:48:28  -Variants not on given reference sequence :  0\n",
      "2026/01/10 15:48:28 Finished checking if NEA is aligned with reference sequence.\n",
      "2026/01/10 15:48:28 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/10 15:48:29  -No statistics have been changed.\n",
      "2026/01/10 15:48:29 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/10 15:48:29 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/10 15:48:42 Finished sorting coordinates.\n",
      "2026/01/10 15:48:42 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 15:48:42  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 15:48:43 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.15294</td>\n",
       "      <td>103.515</td>\n",
       "      <td>-1.150590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.27800</td>\n",
       "      <td>3644.560</td>\n",
       "      <td>-1.143630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.09020</td>\n",
       "      <td>101.840</td>\n",
       "      <td>-1.145510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10000</td>\n",
       "      <td>2179000.000</td>\n",
       "      <td>-0.839228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00000</td>\n",
       "      <td>2281760.000</td>\n",
       "      <td>-0.229918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00000</td>\n",
       "      <td>4837480.000</td>\n",
       "      <td>-1.516370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00000</td>\n",
       "      <td>4525550.000</td>\n",
       "      <td>-2.421510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00000</td>\n",
       "      <td>4717910.000</td>\n",
       "      <td>-1.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00000</td>\n",
       "      <td>4845680.000</td>\n",
       "      <td>-1.524700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80000</td>\n",
       "      <td>385739.000</td>\n",
       "      <td>0.193216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7c4a750bdf10>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 4. Step Two: Harmonize with hg38 FASTA \n",
    "# This aligns alleles to the hg38 reference genome\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=\"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/hg38.fa\",\n",
    "    # n_jobs=4\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "54aeaaa7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 15:50:16 Start to assign rsID using reference file ...(v4.0.4)\n",
      "2026/01/10 15:50:16  -Number of threads/cores to use: 1\n",
      "2026/01/10 15:50:17  -13779885 rsID could be possibly fixed...\n",
      "2026/01/10 15:50:17  -Setting block size:  5000000\n",
      "2026/01/10 15:50:17  -Loading block: 0   1   2   3   4   5   6   7   8   9   10   11   12   13   \n",
      "2026/01/10 15:57:23  -rsID annotation for 13638696 variants needed to be fixed!\n",
      "2026/01/10 15:57:23  -Annotated 141189 rsID successfully!\n",
      "2026/01/10 15:57:24  -Current Dataframe shape : 13779885 x 17 ; Memory usage: 1582.28 MB\n",
      "2026/01/10 15:57:24 Finished assign rsID using reference file.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "      <th>rsID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.15294</td>\n",
       "      <td>103.515</td>\n",
       "      <td>-1.150590</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.27800</td>\n",
       "      <td>3644.560</td>\n",
       "      <td>-1.143630</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.09020</td>\n",
       "      <td>101.840</td>\n",
       "      <td>-1.145510</td>\n",
       "      <td>rs533633326</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10000</td>\n",
       "      <td>2179000.000</td>\n",
       "      <td>-0.839228</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00000</td>\n",
       "      <td>2281760.000</td>\n",
       "      <td>-0.229918</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00000</td>\n",
       "      <td>4837480.000</td>\n",
       "      <td>-1.516370</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00000</td>\n",
       "      <td>4525550.000</td>\n",
       "      <td>-2.421510</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00000</td>\n",
       "      <td>4717910.000</td>\n",
       "      <td>-1.817100</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00000</td>\n",
       "      <td>4845680.000</td>\n",
       "      <td>-1.524700</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80000</td>\n",
       "      <td>385739.000</td>\n",
       "      <td>0.193216</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7c4a750bdf10>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 5. Step Three: Assign rsIDs (Now positions and build match!)\n",
    "mysumstats.assign_rsid(\n",
    "    ref_rsid_tsv=gl.get_path(\"1kg_dbsnp151_hg38_auto\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "33ff6f99",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 15:58:25 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/10 15:58:25  -Using provided chain file: ./grch38-chm13v2.chain\n",
      "2026/01/10 15:58:27  -Converting variants with status code xxx0xxx: 13,779,885\n",
      "2026/01/10 15:58:27  -Target build: 13\n",
      "2026/01/10 15:58:27  -Input positions are 1-based\n",
      "2026/01/10 15:58:27  -Output positions will be 1-based\n",
      "2026/01/10 15:59:56  -Mapped: 13640038 variants\n",
      "2026/01/10 15:59:56  -Unmapped: 139847 variants\n",
      "2026/01/10 15:59:56  -Examples of unmapped variants:\n",
      "2026/01/10 15:59:56    SNPID=1:69487:G:A | CHR=1 | POS=69487 | STATUS=1960099\n",
      "2026/01/10 15:59:56    SNPID=1:69569:T:C | CHR=1 | POS=69569 | STATUS=1960099\n",
      "2026/01/10 15:59:56    SNPID=1:139853:C:T | CHR=1 | POS=139853 | STATUS=1960099\n",
      "2026/01/10 15:59:56    SNPID=1:905017:T:C | CHR=1 | POS=905017 | STATUS=1960099\n",
      "2026/01/10 15:59:56    SNPID=1:909221:T:C | CHR=1 | POS=909221 | STATUS=1960099\n",
      "2026/01/10 16:00:07  -Removed 139847 unmapped variants\n",
      "2026/01/10 16:00:07 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 16:00:07  -Current Dataframe shape : 13640038 x 17 ; Memory usage: 1657.33 MB\n",
      "2026/01/10 16:00:07  -Checking CHR data type...\n",
      "2026/01/10 16:00:09  -Variants with standardized chromosome notation: 13217569\n",
      "2026/01/10 16:00:10  -Variants with fixable chromosome notations: 422469\n",
      "2026/01/10 16:00:10  -No unrecognized chromosome notations...\n",
      "2026/01/10 16:00:11  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/10 16:00:11  -Identified 422469 variants on sex chromosomes...\n",
      "2026/01/10 16:00:11  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/10 16:00:14 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 16:00:14 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 16:00:14  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/10 16:00:15  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 16:00:15  -No outlier variants were removed.\n",
      "2026/01/10 16:00:20  -Removed variants with bad positions: 0\n",
      "2026/01/10 16:00:20 Finished fixing basepair positions (POS).\n",
      "2026/01/10 16:00:21 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "# 6. Step Four: Final Liftover to T2T (CHM13)\n",
    "# Use your downloaded chain file\n",
    "mysumstats.liftover(\n",
    "    from_build=\"38\", \n",
    "    to_build=\"13\", \n",
    "    chain_path=\"./grch38-chm13v2.chain\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9b9acc27",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
