{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd99acb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "796daaf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Load your original UKBB dataset\n",
    "# Nealelab UKBB v3 is in build 19\n",
    "df = pd.read_csv(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "254d195c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split variant into CHR, POS, REF, ALT\n",
    "df[['CHR', 'POS', 'REF', 'ALT']] = df['variant'].str.split(\":\", expand=True)\n",
    "df['POS'] = df['POS'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f37ec073",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename to GWASLab standards\n",
    "df_standard = df.rename(columns={\n",
    "    'variant': 'SNPID',\n",
    "    'ALT': 'EA',\n",
    "    'REF': 'NEA',\n",
    "    'minor_AF': 'EAF',\n",
    "    'beta': 'BETA',\n",
    "    'se': 'SE',\n",
    "    'pval': 'P'\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8ae53e22",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 15:45:01 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/10 15:45:01 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/10 15:45:01 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/10 15:45:01 Start to initialize gl.Sumstats from pandas DataFrame ...\n",
      "2026/01/10 15:45:02  -Reading columns          : \n",
      "2026/01/10 15:45:02  -Renaming columns to      : \n",
      "2026/01/10 15:45:02  -Current Dataframe shape : 13791467  x  15\n",
      "2026/01/10 15:45:02  -Initiating a status column: STATUS ...\n",
      "2026/01/10 15:45:02  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2026/01/10 15:45:05 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 15:45:05  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 15:45:05 Finished reordering the columns.\n",
      "2026/01/10 15:45:06  -Trying to convert datatype for CHR: object -> Int64...Failed...\n",
      "2026/01/10 15:45:09  -Column  : SNPID  CHR    POS   EA       NEA      STATUS EAF     BETA    SE      P       minor_allele low_confidence_variant n_complete_samples AC      ytx     tstat  \n",
      "2026/01/10 15:45:09  -DType   : object object int64 category category int64  float64 float64 float64 float64 object       bool                   int64              float64 float64 float64\n",
      "2026/01/10 15:45:09  -Verified: T      F      T     T        T        T      T       T       T       T       NA           NA                     NA                 NA      NA      NA     \n",
      "2026/01/10 15:45:09 #WARNING! Columns with possibly incompatible dtypes: CHR\n",
      "2026/01/10 15:45:09 #WARNING! Consider using Sumstats.fix_chr() to fix CHR dtype\n",
      "2026/01/10 15:45:09  -Current Dataframe memory usage: 1489.96 MB\n",
      "2026/01/10 15:45:09 Finished loading data successfully!\n"
     ]
    }
   ],
   "source": [
    "# 2. Initialize as Build 19 (Essential!)\n",
    "mysumstats = gl.Sumstats(df_standard, build=\"19\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c61d2fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 15:45:50 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/10 15:45:50  -Current Dataframe shape : 13791467 x 16 ; Memory usage: 1489.96 MB\n",
      "2026/01/10 15:45:50 #WARNING! Data types were not fixed for : CHR \n",
      "2026/01/10 15:45:50 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 15:45:50  -Checking CHR data type...\n",
      "2026/01/10 15:45:52  -Variants with standardized chromosome notation: 13364303\n",
      "2026/01/10 15:45:53  -Variants with fixable chromosome notations: 427164\n",
      "2026/01/10 15:45:53  -No unrecognized chromosome notations...\n",
      "2026/01/10 15:45:53  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/10 15:45:53  -Identified 427164 variants on sex chromosomes...\n",
      "2026/01/10 15:45:53  -Standardizing first sex chromosome chromosome notations: X to 23...\n"
     ]
    }
   ],
   "source": [
    "# 3. Step One: Liftover from hg19 to hg38\n",
    "# GWASLab has built-in chains for 19 to 38\n",
    "mysumstats.liftover(from_build=\"19\", to_build=\"38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a405193",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
