{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cd99acb1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "796daaf4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Load your original UKBB dataset (hg19)\n",
    "df = pd.read_csv(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\")\n",
    "\n",
    "df[['CHR', 'POS', 'REF', 'ALT']] = df['variant'].str.split(\":\", expand=True)\n",
    "df['POS'] = df['POS'].astype(int)\n",
    "\n",
    "df_standard = df.rename(columns={\n",
    "    'variant': 'SNPID', 'ALT': 'EA', 'REF': 'NEA',\n",
    "    'minor_AF': 'EAF', 'beta': 'BETA', 'se': 'SE', 'pval': 'P'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "2b899eb5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 16:12:58 GWASLab v4.0.4 https://cloufield.github.io/gwaslab/\n",
      "2026/01/10 16:12:58 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/10 16:12:58 Python version: 3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0]\n",
      "2026/01/10 16:12:58 Start to initialize gl.Sumstats from pandas DataFrame ...\n",
      "2026/01/10 16:12:59  -Reading columns          : \n",
      "2026/01/10 16:12:59  -Renaming columns to      : \n",
      "2026/01/10 16:12:59  -Current Dataframe shape : 13791467  x  15\n",
      "2026/01/10 16:13:00  -Initiating a status column: STATUS ...\n",
      "2026/01/10 16:13:00  -Genomic coordinates are based on GRCh37/hg19...\n",
      "2026/01/10 16:13:02 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 16:13:02  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 16:13:03 Finished reordering the columns.\n",
      "2026/01/10 16:13:03  -Trying to convert datatype for CHR: object -> Int64...Failed...\n",
      "2026/01/10 16:13:06  -Column  : SNPID  CHR    POS   EA       NEA      STATUS EAF     BETA    SE      P       minor_allele low_confidence_variant n_complete_samples AC      ytx     tstat  \n",
      "2026/01/10 16:13:06  -DType   : object object int64 category category int64  float64 float64 float64 float64 object       bool                   int64              float64 float64 float64\n",
      "2026/01/10 16:13:06  -Verified: T      F      T     T        T        T      T       T       T       T       NA           NA                     NA                 NA      NA      NA     \n",
      "2026/01/10 16:13:06 #WARNING! Columns with possibly incompatible dtypes: CHR\n",
      "2026/01/10 16:13:06 #WARNING! Consider using Sumstats.fix_chr() to fix CHR dtype\n",
      "2026/01/10 16:13:06  -Current Dataframe memory usage: 1489.96 MB\n",
      "2026/01/10 16:13:07 Finished loading data successfully!\n",
      "2026/01/10 16:13:08 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/10 16:13:08  -Current Dataframe shape : 13791467 x 16 ; Memory usage: 1489.96 MB\n",
      "2026/01/10 16:13:08  -Checking SNPID data type...\n",
      "2026/01/10 16:13:08  -Converted datatype for SNPID: object -> string\n",
      "2026/01/10 16:13:08  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/10 16:13:09  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/10 16:13:19 Finished checking SNPID/rsID.\n",
      "2026/01/10 16:13:19 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 16:13:19  -Checking CHR data type...\n",
      "2026/01/10 16:13:21  -Variants with standardized chromosome notation: 13364303\n",
      "2026/01/10 16:13:22  -Variants with fixable chromosome notations: 427164\n",
      "2026/01/10 16:13:22  -No unrecognized chromosome notations...\n",
      "2026/01/10 16:13:22  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/10 16:13:22  -Identified 427164 variants on sex chromosomes...\n",
      "2026/01/10 16:13:22  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/10 16:13:25 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 16:13:25 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 16:13:25  -Trying to convert datatype for POS: int64 -> Int64...\n",
      "2026/01/10 16:13:26  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 16:13:26  -No outlier variants were removed.\n",
      "2026/01/10 16:13:27 Finished fixing basepair positions (POS).\n",
      "2026/01/10 16:13:28 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/10 16:13:28  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/10 16:13:33  -All variants have valid alleles (EA and NEA are A/C/T/G, not NA, and different).\n",
      "2026/01/10 16:13:47 Finished fixing alleles (EA and NEA).\n",
      "2026/01/10 16:13:47 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/10 16:13:47  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/10 16:13:47  -Checking if any columns are empty...\n",
      "2026/01/10 16:13:48  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/10 16:13:49   -Examples of invalid variants(SNPID): 1:776572:C:T,1:819974:G:C,1:871173:C:T,1:871198:C:T,1:874451:C:T ...\n",
      "2026/01/10 16:13:49   -Examples of invalid values (P): NA,NA,NA,NA,NA ...\n",
      "2026/01/10 16:13:50  -Checking if -1e-07 < P < 1.0000001 : removed 1676 variants.\n",
      "2026/01/10 16:13:50   -Examples of invalid variants(SNPID): 1:15791:C:T,1:877608:C:G,1:897032:A:G,1:899336:G:A,1:907676:A:T ...\n",
      "2026/01/10 16:13:50   -Examples of invalid values (BETA): 894.616,279.71,-108.064,270.068,116.14 ...\n",
      "2026/01/10 16:13:51  -Checking if -100.0000001 < BETA < 100.0000001 : removed 9906 variants.\n",
      "2026/01/10 16:13:53  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/10 16:13:53  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/10 16:13:53  -Removed 11582 variants with bad statistics in total.\n",
      "2026/01/10 16:13:53 Finished sanity check for statistics.\n",
      "2026/01/10 16:13:53 Start to check data consistency across columns ...(v4.0.4)\n",
      "2026/01/10 16:13:53  -Current Dataframe shape : 13779885 x 16 ; Memory usage: 1582.28 MB\n",
      "2026/01/10 16:13:53  -Tolerance: 0.001 (Relative) and 0.001 (Absolute)\n",
      "2026/01/10 16:13:53  -Checking if BETA/SE-derived-P is consistent with P...\n",
      "2026/01/10 16:13:55   -Variants with inconsistent values were not detected.\n",
      "2026/01/10 16:13:55   -Note: Minor differences are typically due to rounding.\n",
      "2026/01/10 16:13:55    -If the max difference is greater than expected, please check your original sumstats.\n",
      "2026/01/10 16:13:55 Finished checking data consistency across columns.\n",
      "2026/01/10 16:13:55 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/10 16:13:55  -Number of threads/cores to use: 1\n",
      "2026/01/10 16:13:55  -Number of variants to check:728\n",
      "2026/01/10 16:13:55  -Chunk size:3000000\n",
      "2026/01/10 16:13:55  -Processing in chunks:0 \n",
      "2026/01/10 16:13:57 Finished normalizing indels.\n",
      "2026/01/10 16:13:57 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/10 16:14:11 Finished sorting coordinates.\n",
      "2026/01/10 16:14:11 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 16:14:11  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 16:14:12 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.15294</td>\n",
       "      <td>103.515</td>\n",
       "      <td>-1.150590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569:T:C</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.27800</td>\n",
       "      <td>3644.560</td>\n",
       "      <td>-1.143630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.09020</td>\n",
       "      <td>101.840</td>\n",
       "      <td>-1.145510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:692794:CA:C</td>\n",
       "      <td>1</td>\n",
       "      <td>692794</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10000</td>\n",
       "      <td>2179000.000</td>\n",
       "      <td>-0.839228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:693731:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>693731</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00000</td>\n",
       "      <td>2281760.000</td>\n",
       "      <td>-0.229918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779880</th>\n",
       "      <td>X:154929412:C:T</td>\n",
       "      <td>23</td>\n",
       "      <td>154929412</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00000</td>\n",
       "      <td>4837480.000</td>\n",
       "      <td>-1.516370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779881</th>\n",
       "      <td>X:154929637:CT:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929637</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00000</td>\n",
       "      <td>4525550.000</td>\n",
       "      <td>-2.421510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779882</th>\n",
       "      <td>X:154929952:CAA:C</td>\n",
       "      <td>23</td>\n",
       "      <td>154929952</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>1960399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00000</td>\n",
       "      <td>4717910.000</td>\n",
       "      <td>-1.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779883</th>\n",
       "      <td>X:154930230:A:G</td>\n",
       "      <td>23</td>\n",
       "      <td>154930230</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00000</td>\n",
       "      <td>4845680.000</td>\n",
       "      <td>-1.524700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13779884</th>\n",
       "      <td>X:154930487:T:A</td>\n",
       "      <td>23</td>\n",
       "      <td>154930487</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1960099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80000</td>\n",
       "      <td>385739.000</td>\n",
       "      <td>0.193216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13779885 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7c4b0f1d02c0>"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 2. Initialize and Fix\n",
    "mysumstats = gl.Sumstats(df_standard, build=\"19\")\n",
    "mysumstats.basic_check()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c1b65310",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 16:14:39 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/10 16:14:39  -Using built-in chain file: /mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/data/chains/hg19ToHg38.over.chain.gz\n",
      "2026/01/10 16:14:41  -Converting variants with status code xxx0xxx: 13,779,885\n",
      "2026/01/10 16:14:41  -Target build: 38\n",
      "2026/01/10 16:14:41  -Input positions are 1-based\n",
      "2026/01/10 16:14:41  -Output positions will be 1-based\n",
      "2026/01/10 16:16:16  -Chromosome mismatches detected: 1896 variants (treated as unmapped)\n",
      "2026/01/10 16:16:16  -Examples of chromosome mismatches:\n",
      "2026/01/10 16:16:16    SNPID=1:13123419:T:A | CHR=1 | POS=13123419 | CHR_LIFT=1_KI270766v1_alt | POS_LIFT=1490 | STATUS=1960099\n",
      "2026/01/10 16:16:16    SNPID=1:13123431:A:C | CHR=1 | POS=13123431 | CHR_LIFT=1_KI270766v1_alt | POS_LIFT=1502 | STATUS=1960099\n",
      "2026/01/10 16:16:16    SNPID=1:13124908:G:A | CHR=1 | POS=13124908 | CHR_LIFT=1_KI270766v1_alt | POS_LIFT=2979 | STATUS=1960099\n",
      "2026/01/10 16:16:16    SNPID=1:13125084:A:T | CHR=1 | POS=13125084 | CHR_LIFT=1_KI270766v1_alt | POS_LIFT=3155 | STATUS=1960099\n",
      "2026/01/10 16:16:16    SNPID=1:13130561:T:C | CHR=1 | POS=13130561 | CHR_LIFT=1_KI270766v1_alt | POS_LIFT=8632 | STATUS=1960099\n",
      "2026/01/10 16:16:16  -Mapped: 13771094 variants\n",
      "2026/01/10 16:16:16  -Unmapped: 8791 variants\n",
      "2026/01/10 16:16:16  -Examples of unmapped variants:\n",
      "2026/01/10 16:16:16    SNPID=1:1581586:T:C | CHR=1 | POS=1581586 | STATUS=1960099\n",
      "2026/01/10 16:16:16    SNPID=1:1584102:G:A | CHR=1 | POS=1584102 | STATUS=1960099\n",
      "2026/01/10 16:16:16    SNPID=1:1584117:G:C | CHR=1 | POS=1584117 | STATUS=1960099\n",
      "2026/01/10 16:16:16    SNPID=1:1584218:T:C | CHR=1 | POS=1584218 | STATUS=1960099\n",
      "2026/01/10 16:16:16    SNPID=1:1584224:G:C | CHR=1 | POS=1584224 | STATUS=1960099\n",
      "2026/01/10 16:16:29  -Removed 8791 unmapped variants\n",
      "2026/01/10 16:16:29 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 16:16:29  -Current Dataframe shape : 13771094 x 16 ; Memory usage: 1568.14 MB\n",
      "2026/01/10 16:16:29  -Checking CHR data type...\n",
      "2026/01/10 16:16:30  -Variants with standardized chromosome notation: 13345276\n",
      "2026/01/10 16:16:32  -Variants with fixable chromosome notations: 425818\n",
      "2026/01/10 16:16:32  -No unrecognized chromosome notations...\n",
      "2026/01/10 16:16:33  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/10 16:16:33  -Identified 425818 variants on sex chromosomes...\n",
      "2026/01/10 16:16:33  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/10 16:16:36 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 16:16:36 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 16:16:36  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/10 16:16:37  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 16:16:37  -No outlier variants were removed.\n",
      "2026/01/10 16:16:42  -Removed variants with bad positions: 0\n",
      "2026/01/10 16:16:42 Finished fixing basepair positions (POS).\n",
      "2026/01/10 16:16:43 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "# 3. Liftover hg19 -> hg38\n",
    "mysumstats.liftover(from_build=\"19\", to_build=\"38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8a63a340",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 16:17:26 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/10 16:17:26  -Checking SNPID data type...\n",
      "2026/01/10 16:17:26  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/10 16:17:27  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/10 16:17:37 Finished checking SNPID/rsID.\n",
      "2026/01/10 16:17:37 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 16:17:37  -Checking CHR data type...\n",
      "2026/01/10 16:17:41  -Variants with standardized chromosome notation: 13771094\n",
      "2026/01/10 16:17:41  -All CHR are already fixed...\n",
      "2026/01/10 16:17:44 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 16:17:44 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 16:17:44  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/10 16:17:45  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 16:17:45  -No outlier variants were removed.\n",
      "2026/01/10 16:17:46 Finished fixing basepair positions (POS).\n",
      "2026/01/10 16:17:46 Start to fix alleles (EA and NEA) ...(v4.0.4)\n",
      "2026/01/10 16:17:46  -Converted all bases to string datatype and UPPERCASE\n",
      "2026/01/10 16:17:51  -All variants have valid alleles (EA and NEA are A/C/T/G, not NA, and different).\n",
      "2026/01/10 16:18:05 Finished fixing alleles (EA and NEA).\n",
      "2026/01/10 16:18:05 Start to perform sanity check for statistics ...(v4.0.4)\n",
      "2026/01/10 16:18:06  -Comparison tolerance for floats: 1e-07\n",
      "2026/01/10 16:18:06  -Checking if any columns are empty...\n",
      "2026/01/10 16:18:07  -Checking if -1e-07 < EAF < 1.0000001 : removed 0 variants.\n",
      "2026/01/10 16:18:09  -Checking if -1e-07 < P < 1.0000001 : removed 0 variants.\n",
      "2026/01/10 16:18:10  -Checking if -100.0000001 < BETA < 100.0000001 : removed 0 variants.\n",
      "2026/01/10 16:18:12  -Checking if -1e-07 < SE < inf : removed 0 variants.\n",
      "2026/01/10 16:18:12  -Checking STATUS and converting STATUS to Int64....\n",
      "2026/01/10 16:18:12  -No variants were removed with bad statistics in total.\n",
      "2026/01/10 16:18:12 Finished sanity check for statistics.\n",
      "2026/01/10 16:18:12 Start to normalize indels ...(v4.0.4)\n",
      "2026/01/10 16:18:12  -Number of threads/cores to use: 8\n",
      "2026/01/10 16:18:12  -No available variants to normalize..\n",
      "2026/01/10 16:18:12 Finished normalizing indels.\n",
      "2026/01/10 16:18:12 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 16:18:12  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 16:18:13 Finished reordering the columns.\n",
      "2026/01/10 16:18:13 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/10 16:18:27 Finished sorting coordinates.\n",
      "2026/01/10 16:18:28 Start to check if NEA is aligned with reference sequence ...(v4.0.4)\n",
      "2026/01/10 16:18:28  -Reference genome FASTA file: /mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/hg38.fa\n",
      "2026/01/10 16:18:42    -Loading and building numpy fasta records:\n",
      "2026/01/10 16:18:50  -Variants allele on given reference sequence :  0\n",
      "2026/01/10 16:18:50  -Variants flipped :  0\n",
      "2026/01/10 16:18:50   -Raw Matching rate :  0.00%\n",
      "2026/01/10 16:18:50 #WARNING! Matching rate is low, please check if the right reference genome is used.\n",
      "2026/01/10 16:18:50  -Variants inferred reverse_complement :  0\n",
      "2026/01/10 16:18:50  -Variants inferred reverse_complement_flipped :  0\n",
      "2026/01/10 16:18:50  -Both allele on genome + unable to distinguish :  0\n",
      "2026/01/10 16:18:50  -Variants not on given reference sequence :  0\n",
      "2026/01/10 16:18:51 Finished checking if NEA is aligned with reference sequence.\n",
      "2026/01/10 16:18:51 Start to adjust statistics based on STATUS code ...(v4.0.4)\n",
      "2026/01/10 16:18:52  -No statistics have been changed.\n",
      "2026/01/10 16:18:52 Finished adjusting statistics based on STATUS code.\n",
      "2026/01/10 16:18:52 Start to check SNPID/rsID ...(v4.0.4)\n",
      "2026/01/10 16:18:52  -Checking SNPID data type...\n",
      "2026/01/10 16:18:52  -Checking if SNPID contains NA strings :na,NA,Na,Nan,NaN,<NA>,null,NULL,#N/A,#VALUE!,N/A,n/a,missing,...\n",
      "2026/01/10 16:18:52  -Checking if SNPID is CHR:POS:NEA:EA...(separator: - ,: , _)\n",
      "2026/01/10 16:19:03  -Replacing separators in SNPID with \":\" ...\n",
      "2026/01/10 16:19:25     Filling SNPID from CHR:POS (using CHR, POS)...\n",
      "2026/01/10 16:19:25     Filling SNPID from CHR:POS:NEA:EA (using CHR, POS, EA, NEA)...\n",
      "2026/01/10 16:19:27  -Fixed 0 variants ID...\n",
      "2026/01/10 16:19:27 Finished checking SNPID/rsID.\n",
      "2026/01/10 16:19:27 Start to assign rsID using reference file ...(v4.0.4)\n",
      "2026/01/10 16:19:27  -Number of threads/cores to use: 8\n",
      "2026/01/10 16:19:28  -13771094 rsID could be possibly fixed...\n",
      "2026/01/10 16:19:28  -Setting block size:  5000000\n",
      "2026/01/10 16:19:28  -Loading block: 0   1   2   3   4   5   6   7   8   9   10   11   12   13   \n",
      "2026/01/10 16:26:15  -rsID annotation for 13771094 variants needed to be fixed!\n",
      "2026/01/10 16:26:15  -Annotated 0 rsID successfully!\n",
      "2026/01/10 16:26:15  -Current Dataframe shape : 13771094 x 17 ; Memory usage: 1581.27 MB\n",
      "2026/01/10 16:26:15 Finished assign rsID using reference file.\n",
      "2026/01/10 16:26:15 Start to sort the genome coordinates ...(v4.0.4)\n",
      "2026/01/10 16:26:29 Finished sorting coordinates.\n",
      "2026/01/10 16:26:29 Start to reorder the columns ...(v4.0.4)\n",
      "2026/01/10 16:26:29  -Reordering columns to    : SNPID,rsID,CHR,POS,EA,NEA,STATUS,EAF,BETA,SE,P,minor_allele,low_confidence_variant,n_complete_samples,AC,ytx,tstat\n",
      "2026/01/10 16:26:30 Finished reordering the columns.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNPID</th>\n",
       "      <th>rsID</th>\n",
       "      <th>CHR</th>\n",
       "      <th>POS</th>\n",
       "      <th>EA</th>\n",
       "      <th>NEA</th>\n",
       "      <th>STATUS</th>\n",
       "      <th>EAF</th>\n",
       "      <th>BETA</th>\n",
       "      <th>SE</th>\n",
       "      <th>P</th>\n",
       "      <th>minor_allele</th>\n",
       "      <th>low_confidence_variant</th>\n",
       "      <th>n_complete_samples</th>\n",
       "      <th>AC</th>\n",
       "      <th>ytx</th>\n",
       "      <th>tstat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1:69487</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>69487</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>3860099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.715450</td>\n",
       "      <td>2.360060</td>\n",
       "      <td>0.249902</td>\n",
       "      <td>A</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.15294</td>\n",
       "      <td>103.515</td>\n",
       "      <td>-1.150590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1:69569</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>69569</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>3860099</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>-0.484284</td>\n",
       "      <td>0.423462</td>\n",
       "      <td>0.252778</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>135.27800</td>\n",
       "      <td>3644.560</td>\n",
       "      <td>-1.143630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1:139853</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>139853</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3860099</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>-2.703560</td>\n",
       "      <td>2.360130</td>\n",
       "      <td>0.251997</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "      <td>359983</td>\n",
       "      <td>4.09020</td>\n",
       "      <td>101.840</td>\n",
       "      <td>-1.145510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1:757414</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>757414</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>3860399</td>\n",
       "      <td>0.110590</td>\n",
       "      <td>-0.016436</td>\n",
       "      <td>0.019585</td>\n",
       "      <td>0.401342</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>79621.10000</td>\n",
       "      <td>2179000.000</td>\n",
       "      <td>-0.839228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1:758351</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>1</td>\n",
       "      <td>758351</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3860099</td>\n",
       "      <td>0.115767</td>\n",
       "      <td>-0.004255</td>\n",
       "      <td>0.018507</td>\n",
       "      <td>0.818155</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>83348.00000</td>\n",
       "      <td>2281760.000</td>\n",
       "      <td>-0.229918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13771089</th>\n",
       "      <td>23:155699751</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>155699751</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>3860099</td>\n",
       "      <td>0.245473</td>\n",
       "      <td>-0.016260</td>\n",
       "      <td>0.010723</td>\n",
       "      <td>0.129427</td>\n",
       "      <td>T</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>176732.00000</td>\n",
       "      <td>4837480.000</td>\n",
       "      <td>-1.516370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13771090</th>\n",
       "      <td>23:155699976</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>155699976</td>\n",
       "      <td>C</td>\n",
       "      <td>CT</td>\n",
       "      <td>3860399</td>\n",
       "      <td>0.229726</td>\n",
       "      <td>-0.027098</td>\n",
       "      <td>0.011190</td>\n",
       "      <td>0.015456</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>165395.00000</td>\n",
       "      <td>4525550.000</td>\n",
       "      <td>-2.421510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13771091</th>\n",
       "      <td>23:155700291</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>155700291</td>\n",
       "      <td>C</td>\n",
       "      <td>CAA</td>\n",
       "      <td>3860399</td>\n",
       "      <td>0.239430</td>\n",
       "      <td>-0.020494</td>\n",
       "      <td>0.011278</td>\n",
       "      <td>0.069202</td>\n",
       "      <td>C</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>172381.00000</td>\n",
       "      <td>4717910.000</td>\n",
       "      <td>-1.817100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13771092</th>\n",
       "      <td>23:155700569</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>155700569</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>3860099</td>\n",
       "      <td>0.245887</td>\n",
       "      <td>-0.016347</td>\n",
       "      <td>0.010721</td>\n",
       "      <td>0.127334</td>\n",
       "      <td>G</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>177030.00000</td>\n",
       "      <td>4845680.000</td>\n",
       "      <td>-1.524700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13771093</th>\n",
       "      <td>23:155700826</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>23</td>\n",
       "      <td>155700826</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>3860099</td>\n",
       "      <td>0.019559</td>\n",
       "      <td>0.007107</td>\n",
       "      <td>0.036783</td>\n",
       "      <td>0.846790</td>\n",
       "      <td>A</td>\n",
       "      <td>False</td>\n",
       "      <td>359983</td>\n",
       "      <td>14081.80000</td>\n",
       "      <td>385739.000</td>\n",
       "      <td>0.193216</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13771094 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<gwaslab.g_Sumstats.Sumstats at 0x7c4b0f1d02c0>"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 4. Comprehensive Harmonization (The All-in-One way)\n",
    "# This handles check_ref, flip_allele_stats, and rsID assignment in one pass\n",
    "mysumstats.harmonize(\n",
    "    ref_seq=\"/mnt/hdd_1/ofgeha/gwas-sumstats-harmoniser/gwas-ref/hg38.fa\",\n",
    "    ref_rsid_tsv=gl.get_path(\"1kg_dbsnp151_hg38_auto\"),\n",
    "    sweep_mode=True,  # Critical for 13M+ variants\n",
    "    threads=8         # Use more cores if available\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8b8f48ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 16:26:42 No records in config file. Please download first.\n",
      "2026/01/10 16:26:42 Start to annotate and infer strand orientation using allele frequencies ...(v4.0.4)\n",
      "2026/01/10 16:26:42  -Number of threads/cores to use: 8\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'bool' object has no attribute 'lower'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[21]\u001b[39m\u001b[32m, line 3\u001b[39m\n\u001b[32m      1\u001b[39m \u001b[38;5;66;03m# 5. Optional: Infer Strand (Only if you have palindromic SNPs A/T or G/C)\u001b[39;00m\n\u001b[32m      2\u001b[39m \u001b[38;5;66;03m# Use infer_strand2 for speed on large datasets\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m3\u001b[39m \u001b[43mmysumstats\u001b[49m\u001b[43m.\u001b[49m\u001b[43minfer_strand2\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m      4\u001b[39m \u001b[43m    \u001b[49m\u001b[43mvcf_path\u001b[49m\u001b[43m=\u001b[49m\u001b[43mgl\u001b[49m\u001b[43m.\u001b[49m\u001b[43mget_path\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m1kg_eur_hg38\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;66;43;03m# Use appropriate population ref\u001b[39;49;00m\n\u001b[32m      5\u001b[39m \u001b[43m    \u001b[49m\u001b[43mmaf_threshold\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m0.40\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      6\u001b[39m \u001b[43m    \u001b[49m\u001b[43mthreads\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m8\u001b[39;49m\n\u001b[32m      7\u001b[39m \u001b[43m)\u001b[49m\n\u001b[32m      8\u001b[39m mysumstats.flip_allele_stats() \u001b[38;5;66;03m# Apply the strand flips\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/g_Sumstats.py:963\u001b[39m, in \u001b[36mSumstats.infer_strand2\u001b[39m\u001b[34m(self, **kwargs)\u001b[39m\n\u001b[32m    961\u001b[39m \u001b[38;5;129m@add_doc\u001b[39m(_infer_strand_with_annotation)\n\u001b[32m    962\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34minfer_strand2\u001b[39m(\u001b[38;5;28mself\u001b[39m,**kwargs):\n\u001b[32m--> \u001b[39m\u001b[32m963\u001b[39m     \u001b[38;5;28mself\u001b[39m.data = \u001b[43m_infer_strand_with_annotation\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/qc/qc_decorator.py:206\u001b[39m, in \u001b[36mwith_logging.<locals>.decorator.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    202\u001b[39m         \u001b[38;5;28;01mpass\u001b[39;00m\n\u001b[32m    204\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m is_enough_col==\u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[32m    205\u001b[39m     \u001b[38;5;66;03m# Execute the original function\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m206\u001b[39m     result = \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    207\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    208\u001b[39m     \u001b[38;5;66;03m# not enough cols, return sumstats\u001b[39;00m\n\u001b[32m    209\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m sumstats\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/hm/hm_infer_with_af.py:194\u001b[39m, in \u001b[36m_infer_strand_with_annotation\u001b[39m\u001b[34m(sumstats, path, vcf_path, tsv_path, assign_cols, mapper, threads, reuse_lookup, convert_to_bcf, strip_info, chrom, pos, ea, nea, eaf, raf, flipped_col, strand_col, maf_threshold, ref_maf_threshold, daf_tolerance, log, verbose)\u001b[39m\n\u001b[32m    192\u001b[39m \u001b[38;5;66;03m# Auto-detect reference file format for chromosome mapping\u001b[39;00m\n\u001b[32m    193\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m vcf_path \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m194\u001b[39m     \u001b[43mmapper\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdetect_reference_format\u001b[49m\u001b[43m(\u001b[49m\u001b[43mvcf_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    195\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m path \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m    196\u001b[39m     \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mgwaslab\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mhm\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mhm_assign_rsid\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m is_vcf_file\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/bd/bd_chromosome_mapper.py:645\u001b[39m, in \u001b[36mChromosomeMapper.detect_reference_format\u001b[39m\u001b[34m(self, reference_file)\u001b[39m\n\u001b[32m    642\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m format_type, prefix\n\u001b[32m    644\u001b[39m \u001b[38;5;66;03m# Detect file type and sample chromosomes\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m645\u001b[39m format_type, prefix = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_detect_reference_format_from_file\u001b[49m\u001b[43m(\u001b[49m\u001b[43mreference_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    647\u001b[39m \u001b[38;5;66;03m# Cache result\u001b[39;00m\n\u001b[32m    648\u001b[39m \u001b[38;5;28mself\u001b[39m._reference_cache[reference_file] = (format_type, prefix)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/gwaslab/bd/bd_chromosome_mapper.py:672\u001b[39m, in \u001b[36mChromosomeMapper._detect_reference_format_from_file\u001b[39m\u001b[34m(self, reference_file)\u001b[39m\n\u001b[32m    668\u001b[39m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mgwaslab\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mio\u001b[39;00m\u001b[34;01m.\u001b[39;00m\u001b[34;01mio_ucsc_bed\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m BED_SUFFIXES\n\u001b[32m    670\u001b[39m \u001b[38;5;66;03m# Sample chromosomes from file based on file type\u001b[39;00m\n\u001b[32m    671\u001b[39m \u001b[38;5;66;03m# Handle double extensions like .vcf.gz, .gtf.gz, etc.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m672\u001b[39m file_ext_lower = \u001b[43mreference_file\u001b[49m\u001b[43m.\u001b[49m\u001b[43mlower\u001b[49m()\n\u001b[32m    674\u001b[39m \u001b[38;5;66;03m# Optimized file type detection using tuple checks\u001b[39;00m\n\u001b[32m    675\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m file_ext_lower.endswith(VCF_BCF_SUFFIXES):\n",
      "\u001b[31mAttributeError\u001b[39m: 'bool' object has no attribute 'lower'"
     ]
    }
   ],
   "source": [
    "# 5. Optional: Infer Strand (Only if you have palindromic SNPs A/T or G/C)\n",
    "# Use infer_strand2 for speed on large datasets\n",
    "mysumstats.infer_strand2(\n",
    "    vcf_path=gl.get_path(\"1kg_eur_hg38\"), # Use appropriate population ref\n",
    "    maf_threshold=0.40,\n",
    "    threads=8\n",
    ")\n",
    "mysumstats.flip_allele_stats() # Apply the strand flips"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "2e584c05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/10 16:26:50 Start to perform liftover ...(v4.0.4)\n",
      "2026/01/10 16:26:50  -Using provided chain file: ./grch38-chm13v2.chain\n",
      "2026/01/10 16:26:52  -Converting variants with status code xxx0xxx: 13,771,094\n",
      "2026/01/10 16:26:52  -Target build: 13\n",
      "2026/01/10 16:26:52  -Input positions are 1-based\n",
      "2026/01/10 16:26:52  -Output positions will be 1-based\n",
      "2026/01/10 16:28:24  -Mapped: 13733999 variants\n",
      "2026/01/10 16:28:24  -Unmapped: 37095 variants\n",
      "2026/01/10 16:28:24  -Examples of unmapped variants:\n",
      "2026/01/10 16:28:24    SNPID=1:69487 | CHR=1 | POS=69487 | STATUS=3860099\n",
      "2026/01/10 16:28:24    SNPID=1:69569 | CHR=1 | POS=69569 | STATUS=3860099\n",
      "2026/01/10 16:28:24    SNPID=1:139853 | CHR=1 | POS=139853 | STATUS=3860099\n",
      "2026/01/10 16:28:24    SNPID=1:893791 | CHR=1 | POS=893791 | STATUS=3860399\n",
      "2026/01/10 16:28:24    SNPID=1:909096 | CHR=1 | POS=909096 | STATUS=3860099\n",
      "2026/01/10 16:28:33  -Removed 37095 unmapped variants\n",
      "2026/01/10 16:28:33 Start to fix chromosome notation (CHR) ...(v4.0.4)\n",
      "2026/01/10 16:28:33  -Current Dataframe shape : 13733999 x 17 ; Memory usage: 1668.71 MB\n",
      "2026/01/10 16:28:33  -Checking CHR data type...\n",
      "2026/01/10 16:28:34  -Variants with standardized chromosome notation: 13309342\n",
      "2026/01/10 16:28:35  -Variants with fixable chromosome notations: 424657\n",
      "2026/01/10 16:28:35  -No unrecognized chromosome notations...\n",
      "2026/01/10 16:28:36  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/10 16:28:36  -Identified 424657 variants on sex chromosomes...\n",
      "2026/01/10 16:28:37  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/10 16:28:40 Finished fixing chromosome notation (CHR).\n",
      "2026/01/10 16:28:40 Start to fix basepair positions (POS) ...(v4.0.4)\n",
      "2026/01/10 16:28:40  -Trying to convert datatype for POS: Int64 -> Int64...\n",
      "2026/01/10 16:28:41  -Position bound:(0 , 250,000,000)\n",
      "2026/01/10 16:28:41  -No outlier variants were removed.\n",
      "2026/01/10 16:28:44  -Removed variants with bad positions: 0\n",
      "2026/01/10 16:28:44 Finished fixing basepair positions (POS).\n",
      "2026/01/10 16:28:45 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "# 6. Final Liftover to T2T (CHM13)\n",
    "mysumstats.liftover(\n",
    "    from_build=\"38\", \n",
    "    to_build=\"13\", \n",
    "    chain_path=\"./grch38-chm13v2.chain\"\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
