{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9f832479",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "48e407e0",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[5]\u001b[39m\u001b[32m, line 5\u001b[39m\n\u001b[32m      2\u001b[39m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mgwaslab\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34;01mgl\u001b[39;00m\n\u001b[32m      4\u001b[39m \u001b[38;5;66;03m# Load gzipped GWAS summary stats\u001b[39;00m\n\u001b[32m----> \u001b[39m\u001b[32m5\u001b[39m df = \u001b[43mpd\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[32m      6\u001b[39m \u001b[43m    \u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/.21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz.xpRw4Z\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      7\u001b[39m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[33;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[32m      8\u001b[39m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mgzip\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\n\u001b[32m      9\u001b[39m \u001b[43m)\u001b[49m\n\u001b[32m     11\u001b[39m \u001b[38;5;66;03m# Split variant into CHR, POS, REF, ALT\u001b[39;00m\n\u001b[32m     12\u001b[39m df[[\u001b[33m'\u001b[39m\u001b[33mCHR\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mPOS\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mREF\u001b[39m\u001b[33m'\u001b[39m, \u001b[33m'\u001b[39m\u001b[33mALT\u001b[39m\u001b[33m'\u001b[39m]] = df[\u001b[33m'\u001b[39m\u001b[33mvariant\u001b[39m\u001b[33m'\u001b[39m].str.split(\u001b[33m\"\u001b[39m\u001b[33m:\u001b[39m\u001b[33m\"\u001b[39m, expand=\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1026\u001b[39m, in \u001b[36mread_csv\u001b[39m\u001b[34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[39m\n\u001b[32m   1013\u001b[39m kwds_defaults = _refine_defaults_read(\n\u001b[32m   1014\u001b[39m     dialect,\n\u001b[32m   1015\u001b[39m     delimiter,\n\u001b[32m   (...)\u001b[39m\u001b[32m   1022\u001b[39m     dtype_backend=dtype_backend,\n\u001b[32m   1023\u001b[39m )\n\u001b[32m   1024\u001b[39m kwds.update(kwds_defaults)\n\u001b[32m-> \u001b[39m\u001b[32m1026\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/io/parsers/readers.py:626\u001b[39m, in \u001b[36m_read\u001b[39m\u001b[34m(filepath_or_buffer, kwds)\u001b[39m\n\u001b[32m    623\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n\u001b[32m    625\u001b[39m \u001b[38;5;28;01mwith\u001b[39;00m parser:\n\u001b[32m--> \u001b[39m\u001b[32m626\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mparser\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/io/parsers/readers.py:1923\u001b[39m, in \u001b[36mTextFileReader.read\u001b[39m\u001b[34m(self, nrows)\u001b[39m\n\u001b[32m   1916\u001b[39m nrows = validate_integer(\u001b[33m\"\u001b[39m\u001b[33mnrows\u001b[39m\u001b[33m\"\u001b[39m, nrows)\n\u001b[32m   1917\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m   1918\u001b[39m     \u001b[38;5;66;03m# error: \"ParserBase\" has no attribute \"read\"\u001b[39;00m\n\u001b[32m   1919\u001b[39m     (\n\u001b[32m   1920\u001b[39m         index,\n\u001b[32m   1921\u001b[39m         columns,\n\u001b[32m   1922\u001b[39m         col_dict,\n\u001b[32m-> \u001b[39m\u001b[32m1923\u001b[39m     ) = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_engine\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread\u001b[49m\u001b[43m(\u001b[49m\u001b[43m  \u001b[49m\u001b[38;5;66;43;03m# type: ignore[attr-defined]\u001b[39;49;00m\n\u001b[32m   1924\u001b[39m \u001b[43m        \u001b[49m\u001b[43mnrows\u001b[49m\n\u001b[32m   1925\u001b[39m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   1926\u001b[39m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[32m   1927\u001b[39m     \u001b[38;5;28mself\u001b[39m.close()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/mnt/hdd_1/ofgeha/galaxy-gwas-tools/.venv/lib/python3.12/site-packages/pandas/io/parsers/c_parser_wrapper.py:234\u001b[39m, in \u001b[36mCParserWrapper.read\u001b[39m\u001b[34m(self, nrows)\u001b[39m\n\u001b[32m    232\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[32m    233\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.low_memory:\n\u001b[32m--> \u001b[39m\u001b[32m234\u001b[39m         chunks = \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_reader\u001b[49m\u001b[43m.\u001b[49m\u001b[43mread_low_memory\u001b[49m\u001b[43m(\u001b[49m\u001b[43mnrows\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    235\u001b[39m         \u001b[38;5;66;03m# destructive to chunks\u001b[39;00m\n\u001b[32m    236\u001b[39m         data = _concatenate_chunks(chunks)\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/parsers.pyx:838\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader.read_low_memory\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/parsers.pyx:905\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._read_rows\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/parsers.pyx:874\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._tokenize_rows\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/parsers.pyx:891\u001b[39m, in \u001b[36mpandas._libs.parsers.TextReader._check_tokenize_status\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32mpandas/_libs/parsers.pyx:2053\u001b[39m, in \u001b[36mpandas._libs.parsers.raise_parser_error\u001b[39m\u001b[34m()\u001b[39m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/usr/lib/python3.12/gzip.py:346\u001b[39m, in \u001b[36mGzipFile.closed\u001b[39m\u001b[34m(self)\u001b[39m\n\u001b[32m    343\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mOSError\u001b[39;00m(errno.EBADF, \u001b[33m\"\u001b[39m\u001b[33mpeek() on write-only GzipFile object\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m    344\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._buffer.peek(n)\n\u001b[32m--> \u001b[39m\u001b[32m346\u001b[39m \u001b[38;5;129m@property\u001b[39m\n\u001b[32m    347\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mclosed\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[32m    348\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m.fileobj \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[32m    350\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34mclose\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import gwaslab as gl\n",
    "\n",
    "# Load gzipped GWAS summary stats\n",
    "df = pd.read_csv(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxy-gwas-tools/Data/.21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz.xpRw4Z\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\"\n",
    ")\n",
    "\n",
    "# Split variant into CHR, POS, REF, ALT\n",
    "df[['CHR', 'POS', 'REF', 'ALT']] = df['variant'].str.split(\":\", expand=True)\n",
    "df['POS'] = df['POS'].astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "167b528a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/02 23:54:06 GWASLab v4.0.2 https://cloufield.github.io/gwaslab/\n",
      "2026/01/02 23:54:06 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/02 23:54:06 Python version: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]\n",
      "2026/01/02 23:54:06 Start to initialize gl.Sumstats from pandas DataFrame ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/02 23:54:20  -Reading columns          : CHR,beta,n_complete_samples,minor_AF,se,pval,variant,minor_allele,POS\n",
      "2026/01/02 23:54:20  -Renaming columns to      : CHR,BETA,N,MAF,SE,P,SNPID,EA,POS\n",
      "2026/01/02 23:54:20  -Current Dataframe shape : 13791467  x  15\n",
      "2026/01/02 23:54:28  -Initiating a status column: STATUS ...\n",
      "2026/01/02 23:54:28 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/02 23:54:29  NEA not available: assigning REF to NEA...\n",
      "2026/01/02 23:54:29  -EA,REF and ALT columns are available: assigning NEA...\n",
      "2026/01/02 23:54:31  -For variants with EA == ALT : assigning REF to NEA ...\n",
      "2026/01/02 23:54:35  -For variants with EA != ALT : assigning ALT to NEA ...\n",
      "2026/01/02 23:54:42 Start to reorder the columns ...(v4.0.2)\n",
      "2026/01/02 23:54:42  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,REF,ALT,MAF,BETA,SE,P,N,low_confidence_variant,AC,ytx,tstat\n",
      "2026/01/02 23:54:50 Finished reordering the columns.\n",
      "2026/01/02 23:54:51  -Trying to convert datatype for CHR: string -> Int64...Failed...\n",
      "2026/01/02 23:54:53  -Column  : SNPID  CHR    POS   EA       NEA      STATUS REF      ALT      MAF     BETA    SE      P       N     low_confidence_variant AC      ytx     tstat  \n",
      "2026/01/02 23:54:53  -DType   : object string int64 category category int64  category category float64 float64 float64 float64 int64 bool                   float64 float64 float64\n",
      "2026/01/02 23:54:53  -Verified: T      F      T     T        T        T      T        T        T       T       T       T       T     NA                     NA      NA      NA     \n",
      "2026/01/02 23:54:53 #WARNING! Columns with possibly incompatible dtypes: CHR\n",
      "2026/01/02 23:54:53 #WARNING! Consider using Sumstats.fix_chr() to fix CHR dtype\n",
      "2026/01/02 23:54:53  -Current Dataframe memory usage: 1493.68 MB\n",
      "2026/01/02 23:54:54 Finished loading data successfully!\n",
      "2026/01/02 23:54:54 Start to fix chromosome notation (CHR) ...(v4.0.2)\n",
      "2026/01/02 23:54:54  -Current Dataframe shape : 13791467 x 17 ; Memory usage: 1493.68 MB\n",
      "2026/01/02 23:54:54  -Checking CHR data type...\n",
      "2026/01/02 23:54:55  -Variants with standardized chromosome notation: 13364303\n",
      "2026/01/02 23:54:56  -Variants with fixable chromosome notations: 427164\n",
      "2026/01/02 23:54:56  -No unrecognized chromosome notations...\n",
      "2026/01/02 23:54:56  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/02 23:54:56  -Identified 427164 variants on sex chromosomes...\n",
      "2026/01/02 23:54:56  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/02 23:55:09 Finished fixing chromosome notation (CHR).\n",
      "2026/01/02 23:55:09 Start to perform liftover ...(v4.0.2)\n",
      "2026/01/02 23:55:09  -Using built-in chain file: /mnt/hdd_1/ofgeha/galaxytool/py312/lib/python3.12/site-packages/gwaslab/data/chains/hg19ToHg38.over.chain.gz\n",
      "2026/01/02 23:55:11  -No variants with status code xxx0xxx to liftover\n",
      "2026/01/02 23:55:11 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Create GWASLab Sumstats object and run liftover in one go\n",
    "my_sumstats = (\n",
    "    gl.Sumstats(\n",
    "        df,\n",
    "        snpid=\"variant\",\n",
    "        chrom=\"CHR\",\n",
    "        pos=\"POS\",\n",
    "        ea=\"minor_allele\",\n",
    "        beta=\"beta\",\n",
    "        se=\"se\",\n",
    "        p=\"pval\",\n",
    "        n=\"n_complete_samples\",\n",
    "        maf=\"minor_AF\",\n",
    "        verbose=True\n",
    "    )\n",
    "    .fix_chr()                    # fixes any CHR dtype issues automatically\n",
    "    .liftover(from_build=\"19\", to_build=\"38\")  # liftover to hg38\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d38bdc1e",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'data'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mmy_sumstats\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdata\u001b[49m\n",
      "\u001b[31mAttributeError\u001b[39m: 'NoneType' object has no attribute 'data'"
     ]
    }
   ],
   "source": [
    "my_sumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d190c30d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
