{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f832479",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gwaslab as gl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "48e407e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import gwaslab as gl\n",
    "\n",
    "# Load gzipped GWAS summary stats\n",
    "df = pd.read_csv(\n",
    "    \"/mnt/hdd_1/ofgeha/galaxytool/Data/21001_raw.gwas.imputed_v3.both_sexes.tsv.bgz\",\n",
    "    sep=\"\\t\",\n",
    "    compression=\"gzip\"\n",
    ")\n",
    "\n",
    "# Split variant into CHR, POS, REF, ALT\n",
    "df[['CHR', 'POS', 'REF', 'ALT']] = df['variant'].str.split(\":\", expand=True)\n",
    "df['POS'] = df['POS'].astype(int)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "167b528a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/02 23:54:06 GWASLab v4.0.2 https://cloufield.github.io/gwaslab/\n",
      "2026/01/02 23:54:06 (C) 2022-2026, Yunye He, Kamatani Lab, GPL-3.0 license, gwaslab@gmail.com\n",
      "2026/01/02 23:54:06 Python version: 3.12.12 (main, Oct 10 2025, 08:52:57) [GCC 11.4.0]\n",
      "2026/01/02 23:54:06 Start to initialize gl.Sumstats from pandas DataFrame ...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2026/01/02 23:54:20  -Reading columns          : CHR,beta,n_complete_samples,minor_AF,se,pval,variant,minor_allele,POS\n",
      "2026/01/02 23:54:20  -Renaming columns to      : CHR,BETA,N,MAF,SE,P,SNPID,EA,POS\n",
      "2026/01/02 23:54:20  -Current Dataframe shape : 13791467  x  15\n",
      "2026/01/02 23:54:28  -Initiating a status column: STATUS ...\n",
      "2026/01/02 23:54:28 #WARNING! Version of genomic coordinates is unknown for homo sapiens (build: 99)...\n",
      "2026/01/02 23:54:29  NEA not available: assigning REF to NEA...\n",
      "2026/01/02 23:54:29  -EA,REF and ALT columns are available: assigning NEA...\n",
      "2026/01/02 23:54:31  -For variants with EA == ALT : assigning REF to NEA ...\n",
      "2026/01/02 23:54:35  -For variants with EA != ALT : assigning ALT to NEA ...\n",
      "2026/01/02 23:54:42 Start to reorder the columns ...(v4.0.2)\n",
      "2026/01/02 23:54:42  -Reordering columns to    : SNPID,CHR,POS,EA,NEA,STATUS,REF,ALT,MAF,BETA,SE,P,N,low_confidence_variant,AC,ytx,tstat\n",
      "2026/01/02 23:54:50 Finished reordering the columns.\n",
      "2026/01/02 23:54:51  -Trying to convert datatype for CHR: string -> Int64...Failed...\n",
      "2026/01/02 23:54:53  -Column  : SNPID  CHR    POS   EA       NEA      STATUS REF      ALT      MAF     BETA    SE      P       N     low_confidence_variant AC      ytx     tstat  \n",
      "2026/01/02 23:54:53  -DType   : object string int64 category category int64  category category float64 float64 float64 float64 int64 bool                   float64 float64 float64\n",
      "2026/01/02 23:54:53  -Verified: T      F      T     T        T        T      T        T        T       T       T       T       T     NA                     NA      NA      NA     \n",
      "2026/01/02 23:54:53 #WARNING! Columns with possibly incompatible dtypes: CHR\n",
      "2026/01/02 23:54:53 #WARNING! Consider using Sumstats.fix_chr() to fix CHR dtype\n",
      "2026/01/02 23:54:53  -Current Dataframe memory usage: 1493.68 MB\n",
      "2026/01/02 23:54:54 Finished loading data successfully!\n",
      "2026/01/02 23:54:54 Start to fix chromosome notation (CHR) ...(v4.0.2)\n",
      "2026/01/02 23:54:54  -Current Dataframe shape : 13791467 x 17 ; Memory usage: 1493.68 MB\n",
      "2026/01/02 23:54:54  -Checking CHR data type...\n",
      "2026/01/02 23:54:55  -Variants with standardized chromosome notation: 13364303\n",
      "2026/01/02 23:54:56  -Variants with fixable chromosome notations: 427164\n",
      "2026/01/02 23:54:56  -No unrecognized chromosome notations...\n",
      "2026/01/02 23:54:56  -Identifying non-autosomal chromosomes : X, Y, and MT ...\n",
      "2026/01/02 23:54:56  -Identified 427164 variants on sex chromosomes...\n",
      "2026/01/02 23:54:56  -Standardizing first sex chromosome chromosome notations: X to 23...\n",
      "2026/01/02 23:55:09 Finished fixing chromosome notation (CHR).\n",
      "2026/01/02 23:55:09 Start to perform liftover ...(v4.0.2)\n",
      "2026/01/02 23:55:09  -Using built-in chain file: /mnt/hdd_1/ofgeha/galaxytool/py312/lib/python3.12/site-packages/gwaslab/data/chains/hg19ToHg38.over.chain.gz\n",
      "2026/01/02 23:55:11  -No variants with status code xxx0xxx to liftover\n",
      "2026/01/02 23:55:11 Finished liftover.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Create GWASLab Sumstats object and run liftover in one go\n",
    "my_sumstats = (\n",
    "    gl.Sumstats(\n",
    "        df,\n",
    "        snpid=\"variant\",\n",
    "        chrom=\"CHR\",\n",
    "        pos=\"POS\",\n",
    "        ea=\"minor_allele\",\n",
    "        beta=\"beta\",\n",
    "        se=\"se\",\n",
    "        p=\"pval\",\n",
    "        n=\"n_complete_samples\",\n",
    "        maf=\"minor_AF\",\n",
    "        verbose=True\n",
    "    )\n",
    "    .fix_chr()                    # fixes any CHR dtype issues automatically\n",
    "    .liftover(from_build=\"19\", to_build=\"38\")  # liftover to hg38\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d38bdc1e",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'NoneType' object has no attribute 'data'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 1\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m1\u001b[39m \u001b[43mmy_sumstats\u001b[49m\u001b[43m.\u001b[49m\u001b[43mdata\u001b[49m\n",
      "\u001b[31mAttributeError\u001b[39m: 'NoneType' object has no attribute 'data'"
     ]
    }
   ],
   "source": [
    "my_sumstats.data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d190c30d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py312",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
